{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82dc81c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import tqdm\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "from sklearn import preprocessing # LabelEncoder\n",
    "from sklearn.metrics import mean_squared_error # if squared=False; RMSE\n",
    "\n",
    "from EnvDL.core import *\n",
    "from EnvDL.dna import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef56ce07",
   "metadata": {},
   "outputs": [],
   "source": [
    "cache_path = '../nbs_artifacts/01.03_g2fc_prep_matrices/'\n",
    "ensure_dir_path_exists(dir_path = cache_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fb8bee0",
   "metadata": {},
   "outputs": [],
   "source": [
    "load_from = '../nbs_artifacts/01.02_g2fc_imputation/'\n",
    "\n",
    "meta = pd.read_csv(load_from+'meta.csv')\n",
    "\n",
    "phno = pd.read_csv(load_from+'phno.csv')\n",
    "soil = pd.read_csv(load_from+'soil.csv')\n",
    "wthr = pd.read_csv(load_from+'wthr.csv')\n",
    "\n",
    "cgmv = pd.read_csv(load_from+'cgmv.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6087d405",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = phno.groupby(['Env', 'Year', 'Hybrid', 'Replicate', 'Block', 'Plot']).count().reset_index()\n",
    "assert 1 == np.max(temp.Yield_Mg_ha)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5191eb37",
   "metadata": {},
   "outputs": [],
   "source": [
    "mask = ((phno.Yield_Mg_ha.notna())) # This used to allow for missing yield so long as they were in \n",
    "# 2022. Now that the 2022 data is available they should be excluded.\n",
    "phno = phno.loc[mask, :].reset_index(drop = True)\n",
    "phno = phno.loc[:, ['Env', 'Year', 'Hybrid', 'Replicate', 'Block', 'Plot', 'Yield_Mg_ha']]\n",
    "phno = phno.drop_duplicates().reset_index(drop = True).copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79ec008a",
   "metadata": {},
   "source": [
    "# Data Prep"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21b5b519",
   "metadata": {},
   "source": [
    "## Prep CVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a52bb0ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# YMat[phno.Year == 2021]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85189bcb",
   "metadata": {},
   "source": [
    "## Indexing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39b1becb",
   "metadata": {},
   "source": [
    "### Filtered Phenotype and lookup\n",
    "There are huge data savings to be had from keeping only deduplciated genomic data in memory in the GPU. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcb28457",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['M0028/LH82', 'PHN11_PHW65_0101/PHB47', 'Z035E0047/LH162']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniq_hybrids = list(set(phno.Hybrid))\n",
    "uniq_hybrids[0:3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85e1faa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check snp records and only keep phenotypes with snp data\n",
    "snps_found = [exists_geno((taxa_to_filename(taxa = ith_hybrid))) for ith_hybrid in  uniq_hybrids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f150a80d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hybrid</th>\n",
       "      <th>SNPS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GT603/PHB47</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>PHN11_PHW65_0032/PHT69</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>333</th>\n",
       "      <td>2369/LH123HT</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>GT603/PHZ51</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>LH195/PH463</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4639</th>\n",
       "      <td>PHN66/LH123HT</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4663</th>\n",
       "      <td>PHN11_PHW65_0407/PHT69</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4777</th>\n",
       "      <td>MO44_PHW65_0349/LH195</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4818</th>\n",
       "      <td>BGEM-0122-N/LH195</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4948</th>\n",
       "      <td>TX130_RP/LH198</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Hybrid   SNPS\n",
       "22               GT603/PHB47  False\n",
       "147   PHN11_PHW65_0032/PHT69  False\n",
       "333             2369/LH123HT  False\n",
       "380              GT603/PHZ51  False\n",
       "448              LH195/PH463  False\n",
       "...                      ...    ...\n",
       "4639           PHN66/LH123HT  False\n",
       "4663  PHN11_PHW65_0407/PHT69  False\n",
       "4777   MO44_PHW65_0349/LH195  False\n",
       "4818       BGEM-0122-N/LH195  False\n",
       "4948          TX130_RP/LH198  False\n",
       "\n",
       "[92 rows x 2 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = pd.DataFrame(zip(uniq_hybrids, snps_found), columns=['Hybrid', 'SNPS'])\n",
    "temp.loc[(temp.SNPS != True)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c8165e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genotypes flagged as missing that exist with different capitalization:\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "print('Genotypes flagged as missing that exist with different capitalization:')\n",
    "missing_hybrids = temp.loc[(temp.SNPS != True), ['Hybrid']]\n",
    "# quick check that some of these aren't just differing in capitalization\n",
    "missing_hybrids_list = [e.replace('/', '__').lower() for e in list(missing_hybrids.Hybrid)]\n",
    "print(\n",
    "    [e for e in [e.lower for e in os.listdir('../data/zma/g2fc/genotypes/snps/')] if e in missing_hybrids_list]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd50f3d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2736 observations (2.015%) missing genotype.\n"
     ]
    }
   ],
   "source": [
    "print(\n",
    "    str(missing_hybrids.merge(phno, how = 'left').shape[0]\n",
    "       )+' observations ('+str(round(100 * (missing_hybrids.merge(phno, how = 'left').shape[0]/phno.shape[0]), 3)\n",
    "       )+'%) missing genotype.'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5b9638b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hybrid</th>\n",
       "      <th>SNPS</th>\n",
       "      <th>Env</th>\n",
       "      <th>Year</th>\n",
       "      <th>Replicate</th>\n",
       "      <th>Block</th>\n",
       "      <th>Plot</th>\n",
       "      <th>Yield_Mg_ha</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M0028/LH82</td>\n",
       "      <td>True</td>\n",
       "      <td>ILH1_2015</td>\n",
       "      <td>2015</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>229.0</td>\n",
       "      <td>8.089230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M0028/LH82</td>\n",
       "      <td>True</td>\n",
       "      <td>MNH1_2015</td>\n",
       "      <td>2015</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>188.0</td>\n",
       "      <td>7.625307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M0028/LH82</td>\n",
       "      <td>True</td>\n",
       "      <td>NEH1_2015</td>\n",
       "      <td>2015</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>9.671387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M0028/LH82</td>\n",
       "      <td>True</td>\n",
       "      <td>NEH4_2015</td>\n",
       "      <td>2015</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>409.0</td>\n",
       "      <td>4.714949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M0028/LH82</td>\n",
       "      <td>True</td>\n",
       "      <td>NYH3_2015</td>\n",
       "      <td>2015</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>192.0</td>\n",
       "      <td>6.537947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133052</th>\n",
       "      <td>MOG_LH123HT-019-1-1-1-1-B/PB80</td>\n",
       "      <td>True</td>\n",
       "      <td>NCH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>106.0</td>\n",
       "      <td>8.801143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133053</th>\n",
       "      <td>MOG_LH123HT-019-1-1-1-1-B/PB80</td>\n",
       "      <td>True</td>\n",
       "      <td>TXH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>109.0</td>\n",
       "      <td>8.926460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133054</th>\n",
       "      <td>MOG_LH123HT-019-1-1-1-1-B/PB80</td>\n",
       "      <td>True</td>\n",
       "      <td>TXH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>8.942682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133055</th>\n",
       "      <td>MOG_LH123HT-019-1-1-1-1-B/PB80</td>\n",
       "      <td>True</td>\n",
       "      <td>TXH2_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>5.904733</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133056</th>\n",
       "      <td>MOG_LH123HT-019-1-1-1-1-B/PB80</td>\n",
       "      <td>True</td>\n",
       "      <td>TXH2_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>155.0</td>\n",
       "      <td>10.643744</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133057 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Hybrid  SNPS        Env  Year  Replicate  \\\n",
       "0                           M0028/LH82  True  ILH1_2015  2015        1.0   \n",
       "1                           M0028/LH82  True  MNH1_2015  2015        1.0   \n",
       "2                           M0028/LH82  True  NEH1_2015  2015        1.0   \n",
       "3                           M0028/LH82  True  NEH4_2015  2015        2.0   \n",
       "4                           M0028/LH82  True  NYH3_2015  2015        1.0   \n",
       "...                                ...   ...        ...   ...        ...   \n",
       "133052  MOG_LH123HT-019-1-1-1-1-B/PB80  True  NCH1_2014  2014        2.0   \n",
       "133053  MOG_LH123HT-019-1-1-1-1-B/PB80  True  TXH1_2014  2014        1.0   \n",
       "133054  MOG_LH123HT-019-1-1-1-1-B/PB80  True  TXH1_2014  2014        2.0   \n",
       "133055  MOG_LH123HT-019-1-1-1-1-B/PB80  True  TXH2_2014  2014        1.0   \n",
       "133056  MOG_LH123HT-019-1-1-1-1-B/PB80  True  TXH2_2014  2014        2.0   \n",
       "\n",
       "        Block   Plot  Yield_Mg_ha  \n",
       "0        10.0  229.0     8.089230  \n",
       "1         8.0  188.0     7.625307  \n",
       "2         4.0   85.0     9.671387  \n",
       "3         7.0  409.0     4.714949  \n",
       "4         8.0  192.0     6.537947  \n",
       "...       ...    ...          ...  \n",
       "133052    6.0  106.0     8.801143  \n",
       "133053    5.0  109.0     8.926460  \n",
       "133054    5.0  139.0     8.942682  \n",
       "133055    1.0   37.0     5.904733  \n",
       "133056    7.0  155.0    10.643744  \n",
       "\n",
       "[133057 rows x 8 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# limit to phno with snp data\n",
    "phno_geno = temp.loc[(temp.SNPS == True)].merge(phno, how = 'left')\n",
    "phno_geno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16112e06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Env</th>\n",
       "      <th>Year</th>\n",
       "      <th>Hybrid</th>\n",
       "      <th>Replicate</th>\n",
       "      <th>Block</th>\n",
       "      <th>Plot</th>\n",
       "      <th>Yield_Mg_ha</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0088/LH185</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.721725</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0143/LH185</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.338246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0003/LH185</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.540810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0035/LH185</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>10.366857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0052/LH185</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>10.908814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133052</th>\n",
       "      <td>WIH3_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>W10010_0337/LH244</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>11.975018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133053</th>\n",
       "      <td>WIH3_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>W10010_0346/LH244</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>12.971193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133054</th>\n",
       "      <td>WIH3_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>W10010_0358/LH244</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>13.499769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133055</th>\n",
       "      <td>WIH3_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>W10010_0381/LH244</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>10.831640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133056</th>\n",
       "      <td>WIH3_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>W10010_0452/LH244</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>9.526524</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133057 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Env  Year             Hybrid  Replicate  Block   Plot  \\\n",
       "0       DEH1_2014  2014        M0088/LH185        1.0    1.0    1.0   \n",
       "1       DEH1_2014  2014        M0143/LH185        1.0    1.0    2.0   \n",
       "2       DEH1_2014  2014        M0003/LH185        1.0    1.0    3.0   \n",
       "3       DEH1_2014  2014        M0035/LH185        1.0    1.0    4.0   \n",
       "4       DEH1_2014  2014        M0052/LH185        1.0    1.0    5.0   \n",
       "...           ...   ...                ...        ...    ...    ...   \n",
       "133052  WIH3_2022  2022  W10010_0337/LH244     -999.0 -999.0 -999.0   \n",
       "133053  WIH3_2022  2022  W10010_0346/LH244     -999.0 -999.0 -999.0   \n",
       "133054  WIH3_2022  2022  W10010_0358/LH244     -999.0 -999.0 -999.0   \n",
       "133055  WIH3_2022  2022  W10010_0381/LH244     -999.0 -999.0 -999.0   \n",
       "133056  WIH3_2022  2022  W10010_0452/LH244     -999.0 -999.0 -999.0   \n",
       "\n",
       "        Yield_Mg_ha  \n",
       "0          5.721725  \n",
       "1         11.338246  \n",
       "2          6.540810  \n",
       "3         10.366857  \n",
       "4         10.908814  \n",
       "...             ...  \n",
       "133052    11.975018  \n",
       "133053    12.971193  \n",
       "133054    13.499769  \n",
       "133055    10.831640  \n",
       "133056     9.526524  \n",
       "\n",
       "[133057 rows x 7 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# restrict to observations with snp data\n",
    "phno_geno = phno.merge(phno_geno.loc[:, ['Hybrid', 'SNPS']].drop_duplicates(), how  = 'left')\n",
    "phno_geno = phno_geno.loc[(phno_geno.SNPS == True)].drop(columns = ['SNPS']).reset_index(drop=True).copy()\n",
    "phno_geno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a146327",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert phno_geno.Yield_Mg_ha.isna().sum() == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a1bfab1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# phno_geno.to_csv(cache_path+'phno_geno_filter.csv', index=False)\n",
    "\n",
    "phno_geno = phno_geno.reset_index().rename(columns = {'index':'Phno_Idx'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1f05125",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e7280c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27f3638e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3bbd93f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f685775f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e1ff288",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e75ae5a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1ce2dd3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b44ef3cf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "977caba7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b73f14cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "Env_Idxs  = phno_geno.loc[:, ['Env']\n",
    "                    ].drop_duplicates(\n",
    "                    ).reset_index(drop = True\n",
    "                    ).reset_index(\n",
    "                    ).rename(columns = {'index':'Env_Idx'})\n",
    "Geno_Idxs = phno_geno.loc[:, ['Hybrid']\n",
    "                    ].drop_duplicates(\n",
    "                    ).reset_index(drop = True\n",
    "                    ).reset_index(\n",
    "                    ).rename(columns = {'index':'Geno_Idx'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a14c5285",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is key. I'm explictly adding an index column to fix frustrations jumping between sorted versions of the df.\n",
    "\n",
    "phno_geno = phno_geno.merge(Env_Idxs).merge(Geno_Idxs).sort_values('Phno_Idx').reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b824e6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Env</th>\n",
       "      <th>Year</th>\n",
       "      <th>Hybrid</th>\n",
       "      <th>Yield_Mg_ha</th>\n",
       "      <th>Replicate</th>\n",
       "      <th>Block</th>\n",
       "      <th>Plot</th>\n",
       "      <th>Phno_Idx</th>\n",
       "      <th>Env_Idx</th>\n",
       "      <th>Geno_Idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0088/LH185</td>\n",
       "      <td>5.721725</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0143/LH185</td>\n",
       "      <td>11.338246</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0003/LH185</td>\n",
       "      <td>6.540810</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0035/LH185</td>\n",
       "      <td>10.366857</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0052/LH185</td>\n",
       "      <td>10.908814</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133052</th>\n",
       "      <td>WIH3_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>W10010_0337/LH244</td>\n",
       "      <td>11.975018</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>133052</td>\n",
       "      <td>235</td>\n",
       "      <td>4871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133053</th>\n",
       "      <td>WIH3_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>W10010_0346/LH244</td>\n",
       "      <td>12.971193</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>133053</td>\n",
       "      <td>235</td>\n",
       "      <td>4872</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133054</th>\n",
       "      <td>WIH3_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>W10010_0358/LH244</td>\n",
       "      <td>13.499769</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>133054</td>\n",
       "      <td>235</td>\n",
       "      <td>4873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133055</th>\n",
       "      <td>WIH3_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>W10010_0381/LH244</td>\n",
       "      <td>10.831640</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>133055</td>\n",
       "      <td>235</td>\n",
       "      <td>4875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133056</th>\n",
       "      <td>WIH3_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>W10010_0452/LH244</td>\n",
       "      <td>9.526524</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>133056</td>\n",
       "      <td>235</td>\n",
       "      <td>4897</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133057 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Env  Year             Hybrid  Yield_Mg_ha  Replicate  Block  \\\n",
       "0       DEH1_2014  2014        M0088/LH185     5.721725        1.0    1.0   \n",
       "1       DEH1_2014  2014        M0143/LH185    11.338246        1.0    1.0   \n",
       "2       DEH1_2014  2014        M0003/LH185     6.540810        1.0    1.0   \n",
       "3       DEH1_2014  2014        M0035/LH185    10.366857        1.0    1.0   \n",
       "4       DEH1_2014  2014        M0052/LH185    10.908814        1.0    1.0   \n",
       "...           ...   ...                ...          ...        ...    ...   \n",
       "133052  WIH3_2022  2022  W10010_0337/LH244    11.975018     -999.0 -999.0   \n",
       "133053  WIH3_2022  2022  W10010_0346/LH244    12.971193     -999.0 -999.0   \n",
       "133054  WIH3_2022  2022  W10010_0358/LH244    13.499769     -999.0 -999.0   \n",
       "133055  WIH3_2022  2022  W10010_0381/LH244    10.831640     -999.0 -999.0   \n",
       "133056  WIH3_2022  2022  W10010_0452/LH244     9.526524     -999.0 -999.0   \n",
       "\n",
       "         Plot  Phno_Idx  Env_Idx  Geno_Idx  \n",
       "0         1.0         0        0         0  \n",
       "1         2.0         1        0         1  \n",
       "2         3.0         2        0         2  \n",
       "3         4.0         3        0         3  \n",
       "4         5.0         4        0         4  \n",
       "...       ...       ...      ...       ...  \n",
       "133052 -999.0    133052      235      4871  \n",
       "133053 -999.0    133053      235      4872  \n",
       "133054 -999.0    133054      235      4873  \n",
       "133055 -999.0    133055      235      4875  \n",
       "133056 -999.0    133056      235      4897  \n",
       "\n",
       "[133057 rows x 10 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phno_geno = phno_geno.loc[:, [ 'Env', 'Year', 'Hybrid', 'Yield_Mg_ha', \n",
    "                              'Replicate', 'Block', 'Plot', \n",
    "                              'Phno_Idx', 'Env_Idx', 'Geno_Idx']]\n",
    "\n",
    "phno_geno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b796f2e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "phno_geno.to_csv(cache_path+'phno_geno.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d16621c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# and overwrite phno to prevent incomplete data from creeping in.\n",
    "phno = phno_geno.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "afe83436",
   "metadata": {},
   "source": [
    "### Observation to Planting/Harvest DOY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bb618cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DOY_Planted</th>\n",
       "      <th>DOY_Harvested</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>125</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>125</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>125</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>125</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>125</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133052</th>\n",
       "      <td>136</td>\n",
       "      <td>303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133053</th>\n",
       "      <td>136</td>\n",
       "      <td>303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133054</th>\n",
       "      <td>136</td>\n",
       "      <td>303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133055</th>\n",
       "      <td>136</td>\n",
       "      <td>303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133056</th>\n",
       "      <td>136</td>\n",
       "      <td>303</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133057 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        DOY_Planted  DOY_Harvested\n",
       "0               125            272\n",
       "1               125            272\n",
       "2               125            272\n",
       "3               125            272\n",
       "4               125            272\n",
       "...             ...            ...\n",
       "133052          136            303\n",
       "133053          136            303\n",
       "133054          136            303\n",
       "133055          136            303\n",
       "133056          136            303\n",
       "\n",
       "[133057 rows x 2 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# use phno_index as lookup\n",
    "\n",
    "mgmt_time = phno.merge(\n",
    "    meta.loc[:, ['Env', 'Hybrid', 'Replicate', 'Block', 'Plot', 'Date_Planted','Date_Harvested']\n",
    "       ].drop_duplicates(), how = 'left')\n",
    "\n",
    "mgmt_time['DOY_Planted'] = round(mgmt_time['Date_Planted']).astype(int)\n",
    "mgmt_time['DOY_Harvested'] = round(mgmt_time['Date_Harvested']).astype(int)\n",
    "\n",
    "mgmt_time = mgmt_time.loc[:, ['DOY_Planted','DOY_Harvested']]\n",
    "mgmt_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e87037bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "PlantHarvest = np.array(mgmt_time.loc[:, ])\n",
    "np.save(cache_path+'PlantHarvestNames.npy', list(mgmt_time))\n",
    "np.save(cache_path+'PlantHarvest.npy', PlantHarvest)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41d18c52",
   "metadata": {},
   "source": [
    "### Observation to Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9f7817e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Env</th>\n",
       "      <th>Env_Idx</th>\n",
       "      <th>Pounds_Needed_Soil_Moisture</th>\n",
       "      <th>Cover_beet</th>\n",
       "      <th>Cover_corn</th>\n",
       "      <th>Cover_cotton</th>\n",
       "      <th>Cover_fallow</th>\n",
       "      <th>Cover_lima</th>\n",
       "      <th>Cover_peanut</th>\n",
       "      <th>Cover_pumpkin</th>\n",
       "      <th>Cover_rye</th>\n",
       "      <th>Cover_sorghum</th>\n",
       "      <th>Cover_soy</th>\n",
       "      <th>Cover_wheat</th>\n",
       "      <th>Pre_Chisel</th>\n",
       "      <th>Pre_Cult</th>\n",
       "      <th>Pre_Disc</th>\n",
       "      <th>Pre_MinTill</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>0</td>\n",
       "      <td>4.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GAH1_2014</td>\n",
       "      <td>1</td>\n",
       "      <td>4.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IAH1a_2014</td>\n",
       "      <td>2</td>\n",
       "      <td>4.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>IAH1b_2014</td>\n",
       "      <td>3</td>\n",
       "      <td>4.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IAH1c_2014</td>\n",
       "      <td>4</td>\n",
       "      <td>4.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>TXH2_2022</td>\n",
       "      <td>231</td>\n",
       "      <td>4.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>TXH3_2022</td>\n",
       "      <td>232</td>\n",
       "      <td>4.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>233</th>\n",
       "      <td>WIH1_2022</td>\n",
       "      <td>233</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>WIH2_2022</td>\n",
       "      <td>234</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>WIH3_2022</td>\n",
       "      <td>235</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>236 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Env  Env_Idx  Pounds_Needed_Soil_Moisture  Cover_beet  Cover_corn  \\\n",
       "0     DEH1_2014        0                          4.5         0.0         0.0   \n",
       "1     GAH1_2014        1                          4.5         0.0         0.0   \n",
       "2    IAH1a_2014        2                          4.5         0.0         0.0   \n",
       "3    IAH1b_2014        3                          4.5         0.0         0.0   \n",
       "4    IAH1c_2014        4                          4.5         0.0         0.0   \n",
       "..          ...      ...                          ...         ...         ...   \n",
       "231   TXH2_2022      231                          4.5         0.0         1.0   \n",
       "232   TXH3_2022      232                          4.5         0.0         1.0   \n",
       "233   WIH1_2022      233                          5.0         0.0         0.0   \n",
       "234   WIH2_2022      234                          5.0         0.0         0.0   \n",
       "235   WIH3_2022      235                          5.0         0.0         0.0   \n",
       "\n",
       "     Cover_cotton  Cover_fallow  Cover_lima  Cover_peanut  Cover_pumpkin  \\\n",
       "0             0.0           0.0         0.0           0.0            0.0   \n",
       "1             1.0           0.0         0.0           0.0            0.0   \n",
       "2             0.0           0.0         0.0           0.0            0.0   \n",
       "3             0.0           0.0         0.0           0.0            0.0   \n",
       "4             0.0           0.0         0.0           0.0            0.0   \n",
       "..            ...           ...         ...           ...            ...   \n",
       "231           0.0           0.0         0.0           0.0            0.0   \n",
       "232           0.0           0.0         0.0           0.0            0.0   \n",
       "233           0.0           0.0         0.0           0.0            0.0   \n",
       "234           0.0           0.0         0.0           0.0            0.0   \n",
       "235           0.0           0.0         0.0           0.0            0.0   \n",
       "\n",
       "     Cover_rye  Cover_sorghum  Cover_soy  Cover_wheat  Pre_Chisel  Pre_Cult  \\\n",
       "0          0.0            0.0        1.0          0.0         0.0       1.0   \n",
       "1          0.0            0.0        0.0          0.0         0.0       1.0   \n",
       "2          0.0            0.0        1.0          0.0         0.0       1.0   \n",
       "3          0.0            0.0        1.0          0.0         0.0       1.0   \n",
       "4          0.0            0.0        1.0          0.0         0.0       1.0   \n",
       "..         ...            ...        ...          ...         ...       ...   \n",
       "231        0.0            0.0        0.0          0.0         1.0       1.0   \n",
       "232        0.0            0.0        0.0          0.0         1.0       1.0   \n",
       "233        0.0            0.0        1.0          0.0         0.0       1.0   \n",
       "234        0.0            0.0        1.0          0.0         0.0       1.0   \n",
       "235        0.0            0.0        1.0          0.0         0.0       0.0   \n",
       "\n",
       "     Pre_Disc  Pre_MinTill  \n",
       "0         0.0          0.0  \n",
       "1         0.0          0.0  \n",
       "2         0.0          0.0  \n",
       "3         0.0          0.0  \n",
       "4         0.0          0.0  \n",
       "..        ...          ...  \n",
       "231       1.0          0.0  \n",
       "232       1.0          0.0  \n",
       "233       0.0          0.0  \n",
       "234       0.0          0.0  \n",
       "235       1.0          0.0  \n",
       "\n",
       "[236 rows x 18 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_small = phno.loc[:, ['Env', 'Env_Idx']].drop_duplicates().merge(meta.loc[:, [\n",
    "    'Env',\n",
    " 'Pounds_Needed_Soil_Moisture',\n",
    " 'Cover_beet',\n",
    " 'Cover_corn',\n",
    " 'Cover_cotton',\n",
    " 'Cover_fallow',\n",
    " 'Cover_lima',\n",
    " 'Cover_peanut',\n",
    " 'Cover_pumpkin',\n",
    " 'Cover_rye',\n",
    " 'Cover_sorghum',\n",
    " 'Cover_soy',\n",
    " 'Cover_wheat',\n",
    " 'Pre_Chisel',\n",
    " 'Pre_Cult',\n",
    " 'Pre_Disc',\n",
    " 'Pre_MinTill']].drop_duplicates(), how = 'left').sort_values('Env_Idx')\n",
    "\n",
    "meta_small"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4264c0e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "mgmtMat = np.array(meta_small.loc[:, ].drop(columns = ['Env', 'Env_Idx']))\n",
    "np.save(cache_path+'mgmtMatNames.npy', list(meta_small.loc[:, ].drop(columns = ['Env', 'Env_Idx'])))\n",
    "np.save(cache_path+'mgmtMat.npy', mgmtMat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf9e9990",
   "metadata": {},
   "source": [
    "### Observation to Genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef8bb24f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phno_Idx</th>\n",
       "      <th>Geno_Idx</th>\n",
       "      <th>Is_Phno_Idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133052</th>\n",
       "      <td>133052</td>\n",
       "      <td>4871</td>\n",
       "      <td>123234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133053</th>\n",
       "      <td>133053</td>\n",
       "      <td>4872</td>\n",
       "      <td>123235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133054</th>\n",
       "      <td>133054</td>\n",
       "      <td>4873</td>\n",
       "      <td>123236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133055</th>\n",
       "      <td>133055</td>\n",
       "      <td>4875</td>\n",
       "      <td>123238</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133056</th>\n",
       "      <td>133056</td>\n",
       "      <td>4897</td>\n",
       "      <td>123684</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133057 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Phno_Idx  Geno_Idx  Is_Phno_Idx\n",
       "0              0         0            0\n",
       "1              1         1            1\n",
       "2              2         2            2\n",
       "3              3         3            3\n",
       "4              4         4            4\n",
       "...          ...       ...          ...\n",
       "133052    133052      4871       123234\n",
       "133053    133053      4872       123235\n",
       "133054    133054      4873       123236\n",
       "133055    133055      4875       123238\n",
       "133056    133056      4897       123684\n",
       "\n",
       "[133057 rows x 3 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniq_geno = phno_geno.loc[:, ['Geno_Idx']].drop_duplicates().reset_index().rename(columns = {'index':'Is_Phno_Idx'})\n",
    "obs_geno_lookup = phno_geno.loc[:, ['Phno_Idx', 'Geno_Idx']].merge(uniq_geno).sort_values('Phno_Idx').reset_index(drop = True).copy()\n",
    "obs_geno_lookup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df34d521",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save out obs_geno_lookup\n",
    "np.save(cache_path+'obs_geno_lookup.npy', np.asarray(obs_geno_lookup))\n",
    "obs_geno_lookup.to_csv(cache_path+'obs_geno_lookup.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad8d81b7",
   "metadata": {},
   "source": [
    "### Observation to Environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17a1fce7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phno_Idx</th>\n",
       "      <th>Env_Idx</th>\n",
       "      <th>Is_Phno_Idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133052</th>\n",
       "      <td>133052</td>\n",
       "      <td>235</td>\n",
       "      <td>132615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133053</th>\n",
       "      <td>133053</td>\n",
       "      <td>235</td>\n",
       "      <td>132615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133054</th>\n",
       "      <td>133054</td>\n",
       "      <td>235</td>\n",
       "      <td>132615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133055</th>\n",
       "      <td>133055</td>\n",
       "      <td>235</td>\n",
       "      <td>132615</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133056</th>\n",
       "      <td>133056</td>\n",
       "      <td>235</td>\n",
       "      <td>132615</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133057 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Phno_Idx  Env_Idx  Is_Phno_Idx\n",
       "0              0        0            0\n",
       "1              1        0            0\n",
       "2              2        0            0\n",
       "3              3        0            0\n",
       "4              4        0            0\n",
       "...          ...      ...          ...\n",
       "133052    133052      235       132615\n",
       "133053    133053      235       132615\n",
       "133054    133054      235       132615\n",
       "133055    133055      235       132615\n",
       "133056    133056      235       132615\n",
       "\n",
       "[133057 rows x 3 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniq_env = phno_geno.loc[:, ['Env_Idx']].drop_duplicates().reset_index().rename(columns = {'index':'Is_Phno_Idx'})\n",
    "obs_env_lookup = phno_geno.loc[:, ['Phno_Idx', 'Env_Idx']].merge(uniq_env).sort_values('Phno_Idx').reset_index(drop = True).copy()\n",
    "obs_env_lookup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d14eb13e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save out obs_geno_lookup\n",
    "np.save(cache_path+'obs_env_lookup.npy', np.asarray(obs_env_lookup))\n",
    "obs_env_lookup.to_csv(cache_path+'obs_env_lookup.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05d090ef",
   "metadata": {},
   "source": [
    "## Y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56c358cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "YMat = np.array(phno_geno.Yield_Mg_ha)\n",
    "np.save(cache_path+'YMat.npy', YMat)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19b703a1",
   "metadata": {},
   "source": [
    "## Genomic Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aabb9cab",
   "metadata": {},
   "source": [
    "### One Hot Encode G (Deduplicated)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9fd8ee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# If phno and GMat observations are 1:1, then GMat is 135793, 2250. After deduplicating this will become 4926, 2250. That's only 3.6%!\n",
    "# Doing this removes several hybrids so it's actually a little smaller 4926, 2203"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "978cc79f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Phno_Idx</th>\n",
       "      <th>Geno_Idx</th>\n",
       "      <th>Is_Phno_Idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Phno_Idx  Geno_Idx  Is_Phno_Idx\n",
       "0         0         0            0\n",
       "1         1         1            1\n",
       "2         2         2            2\n",
       "3         3         3            3\n",
       "4         4         4            4"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs_geno_lookup.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bc5250c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Env</th>\n",
       "      <th>Year</th>\n",
       "      <th>Hybrid</th>\n",
       "      <th>Yield_Mg_ha</th>\n",
       "      <th>Replicate</th>\n",
       "      <th>Block</th>\n",
       "      <th>Plot</th>\n",
       "      <th>Phno_Idx</th>\n",
       "      <th>Env_Idx</th>\n",
       "      <th>Geno_Idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0088/LH185</td>\n",
       "      <td>5.721725</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0143/LH185</td>\n",
       "      <td>11.338246</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0003/LH185</td>\n",
       "      <td>6.540810</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0035/LH185</td>\n",
       "      <td>10.366857</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DEH1_2014</td>\n",
       "      <td>2014</td>\n",
       "      <td>M0052/LH185</td>\n",
       "      <td>10.908814</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4921</th>\n",
       "      <td>IAH2_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>GEMN-0096_LH212HT_0014/LH244</td>\n",
       "      <td>10.929844</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>124682</td>\n",
       "      <td>217</td>\n",
       "      <td>4921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4922</th>\n",
       "      <td>IAH2_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>NC370/LH244</td>\n",
       "      <td>5.306059</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>125036</td>\n",
       "      <td>217</td>\n",
       "      <td>4922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4923</th>\n",
       "      <td>IAH3_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>GEMN-0097_LH212HT_0046/LH244</td>\n",
       "      <td>13.179435</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>125273</td>\n",
       "      <td>218</td>\n",
       "      <td>4923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4924</th>\n",
       "      <td>IAH4_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>GEMN-0225_PHK76_0025/LH244</td>\n",
       "      <td>11.453161</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>126009</td>\n",
       "      <td>219</td>\n",
       "      <td>4924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4925</th>\n",
       "      <td>MIH1_2022</td>\n",
       "      <td>2022</td>\n",
       "      <td>GEMN-0097_PHN46_0012/LH244</td>\n",
       "      <td>6.012986</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>-999.0</td>\n",
       "      <td>127279</td>\n",
       "      <td>222</td>\n",
       "      <td>4925</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4926 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Env  Year                        Hybrid  Yield_Mg_ha  Replicate  \\\n",
       "0     DEH1_2014  2014                   M0088/LH185     5.721725        1.0   \n",
       "1     DEH1_2014  2014                   M0143/LH185    11.338246        1.0   \n",
       "2     DEH1_2014  2014                   M0003/LH185     6.540810        1.0   \n",
       "3     DEH1_2014  2014                   M0035/LH185    10.366857        1.0   \n",
       "4     DEH1_2014  2014                   M0052/LH185    10.908814        1.0   \n",
       "...         ...   ...                           ...          ...        ...   \n",
       "4921  IAH2_2022  2022  GEMN-0096_LH212HT_0014/LH244    10.929844     -999.0   \n",
       "4922  IAH2_2022  2022                   NC370/LH244     5.306059     -999.0   \n",
       "4923  IAH3_2022  2022  GEMN-0097_LH212HT_0046/LH244    13.179435     -999.0   \n",
       "4924  IAH4_2022  2022    GEMN-0225_PHK76_0025/LH244    11.453161     -999.0   \n",
       "4925  MIH1_2022  2022    GEMN-0097_PHN46_0012/LH244     6.012986     -999.0   \n",
       "\n",
       "      Block   Plot  Phno_Idx  Env_Idx  Geno_Idx  \n",
       "0       1.0    1.0         0        0         0  \n",
       "1       1.0    2.0         1        0         1  \n",
       "2       1.0    3.0         2        0         2  \n",
       "3       1.0    4.0         3        0         3  \n",
       "4       1.0    5.0         4        0         4  \n",
       "...     ...    ...       ...      ...       ...  \n",
       "4921 -999.0 -999.0    124682      217      4921  \n",
       "4922 -999.0 -999.0    125036      217      4922  \n",
       "4923 -999.0 -999.0    125273      218      4923  \n",
       "4924 -999.0 -999.0    126009      219      4924  \n",
       "4925 -999.0 -999.0    127279      222      4925  \n",
       "\n",
       "[4926 rows x 10 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = obs_geno_lookup.drop(columns = 'Phno_Idx').drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "mask = phno_geno.index.isin(list(obs_geno_lookup['Is_Phno_Idx'].drop_duplicates()))\n",
    "temp = phno_geno.loc[mask, ]#'Hybrid']\n",
    "temp = temp.reset_index(drop = True) # drop so that the index in the DataFrame and ndarray match\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef52cff1",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = pd.concat([temp, temp.Hybrid.str.split('/', expand=True)], axis=1\n",
    "        ).rename(columns = {0:'P0', 1:'P1'})\n",
    "\n",
    "uniq_parents = list(set(pd.concat([temp['P0'], temp['P1']])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "874b0f1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "if os.path.exists(cache_path+'GMat.npy'):\n",
    "    GMat = np.load(cache_path+'GMat.npy')\n",
    "else:\n",
    "    GMat = np.zeros([temp.shape[0], len(uniq_parents)])\n",
    "\n",
    "    for j in tqdm.tqdm(range(len(uniq_parents))):\n",
    "        for parent in ['P0', 'P1']:\n",
    "            mask = (temp[parent] == uniq_parents[j]) \n",
    "            GMat[temp.loc[mask, ].index, j] += 1\n",
    "            \n",
    "    np.save(cache_path+'GMat.npy', GMat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b28dac93",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save(cache_path+'GMatNames.npy', uniq_parents)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab8f5120",
   "metadata": {},
   "outputs": [],
   "source": [
    "# confirm there are two parents encoded for each observation\n",
    "assert 2 == np.min(np.sum(GMat, axis = 1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be0fbd9f",
   "metadata": {},
   "source": [
    "### Nucleotides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a23cb81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'unk': 0,\n",
       " 'A': 1,\n",
       " 'C': 2,\n",
       " 'G': 3,\n",
       " 'T': 4,\n",
       " 'K': 5,\n",
       " 'M': 6,\n",
       " 'N': 7,\n",
       " 'R': 8,\n",
       " 'S': 9,\n",
       " 'W': 10,\n",
       " 'Y': 11}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert IUPAC codes into one hot\n",
    "# https://www.bioinformatics.org/sms/iupac.html\n",
    "IUPAC = ['unk', 'A', 'C', 'G', 'T', 'K', 'M', 'N', 'R', 'S', 'W', 'Y']\n",
    "IUPAC = dict(zip(IUPAC, [i for i in range(len(IUPAC))])) # begin with unk.\n",
    "IUPAC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b692b85",
   "metadata": {},
   "outputs": [],
   "source": [
    "put_cached_result(cache_path+'ACGT_OneHot_dict.pkl', IUPAC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ae820a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "if os.path.exists(cache_path+'ACGT_OneHot.npy'):\n",
    "#     ACGT = np.load(cache_path+'ACGT_OneHot.npy')\n",
    "    pass\n",
    "else:\n",
    "    temp = obs_geno_lookup.drop(columns = 'Phno_Idx').drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "    # use the first Hybrid to figure out what the length of the sequence is.\n",
    "    i = 0\n",
    "    phno_idx = temp.loc[i, 'Is_Phno_Idx']\n",
    "    ith_hybrid = phno.loc[phno_idx, 'Hybrid']\n",
    "    geno_seq_len = len(get_geno(taxa_to_filename(taxa = ith_hybrid))[1:])    \n",
    "\n",
    "    # setup ndarray to hold data\n",
    "    ACGT = np.ndarray(shape = (temp.shape[0],        # obs\n",
    "                               geno_seq_len,         # values\n",
    "                               len(IUPAC.keys())+1)) # channels\n",
    "\n",
    "    # This is inefficient but needs only to be run once.\n",
    "    # go over observations, channels, fill in values\n",
    "\n",
    "    for i in tqdm.tqdm(temp.index): \n",
    "        phno_idx = temp.loc[i, 'Is_Phno_Idx']\n",
    "        ith_hybrid = phno.loc[phno_idx, 'Hybrid']\n",
    "        res = get_geno(taxa_to_filename(taxa = ith_hybrid)) \n",
    "        res = res[1:] # drop taxa\n",
    "\n",
    "        res = [e.strip('\\n') for e in res] # remove end of line\n",
    "\n",
    "        for key in IUPAC.keys():\n",
    "            IUPAC_mask = [True if e == key else False for e in res]\n",
    "            ACGT[i, \n",
    "                 IUPAC_mask, \n",
    "                 IUPAC[key]] = 1  \n",
    "\n",
    "    # Swap axes to match pytorch convention\n",
    "    ACGT = np.swapaxes(ACGT, 1, 2)\n",
    "\n",
    "    np.save(cache_path+'ACGT_OneHot.npy', ACGT)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a83c761",
   "metadata": {},
   "source": [
    "### Nucleotide percents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55f9b600",
   "metadata": {},
   "outputs": [],
   "source": [
    "if os.path.exists(cache_path+'ACGT.npy'):\n",
    "#     ACGT = np.load(cache_path+'ACGT.npy')\n",
    "    pass\n",
    "else:\n",
    "    temp = obs_geno_lookup.drop(columns = 'Phno_Idx').drop_duplicates().reset_index(drop=True)\n",
    "\n",
    "    \n",
    "    # use the first Hybrid to figure out what the length of the sequence is.\n",
    "    i = 0\n",
    "    phno_idx = temp.loc[i, 'Is_Phno_Idx']\n",
    "    ith_hybrid = phno.loc[phno_idx, 'Hybrid']\n",
    "    geno_seq_len = len(get_geno(taxa_to_filename(taxa = ith_hybrid))[1:])    \n",
    "    \n",
    "    # setup ndarray to hold data\n",
    "    ACGT = np.ndarray(shape = (temp.shape[0], geno_seq_len, 4))\n",
    "\n",
    "    for i in tqdm.tqdm(temp.index):\n",
    "        phno_idx = temp.loc[i, 'Is_Phno_Idx']\n",
    "        ith_hybrid = phno.loc[phno_idx, 'Hybrid']\n",
    "        res = get_geno(taxa_to_filename(taxa = ith_hybrid)) \n",
    "        res = res[1:] # drop taxa\n",
    "        res = list_to_ACGT(in_seq = res)\n",
    "        ACGT[i, :, :] = res[None, :, :]\n",
    "        \n",
    "    # Swap axes to match pytorch convention\n",
    "    # (4926, 125891, 4)\n",
    "    ACGT = np.swapaxes(ACGT, 1, 2)\n",
    "\n",
    "    # set missings to 0 \n",
    "    ACGT[np.isnan(ACGT)] = 0\n",
    "\n",
    "    np.save(cache_path+'ACGT.npy', ACGT)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e924a0d7",
   "metadata": {},
   "source": [
    "### Hilbert Nucleotide percents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4508fb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "if os.path.exists(cache_path+'ACGT_hilb.npy'):\n",
    "#     ACGT_hilb = np.load(cache_path+'ACGT_hilb.npy')\n",
    "    pass\n",
    "else:\n",
    "    ACGT_hilb = np_3d_to_hilbert( np.swapaxes(ACGT, 1, 2) ) # swap channels back to dim 2 before running\n",
    "\n",
    "    # ACGT_hilb.shape\n",
    "    # (4926, 256, 512, 4)\n",
    "    # Pytorch standard has channels second\n",
    "    ACGT_hilb = np.swapaxes(ACGT_hilb, 1, 3)\n",
    "    ACGT_hilb = np.swapaxes(ACGT_hilb, 2, 3)\n",
    "\n",
    "    # set missings to 0\n",
    "    ACGT_hilb[np.isnan(ACGT_hilb)] = 0\n",
    "\n",
    "    np.save(cache_path+'ACGT_hilb.npy', ACGT_hilb)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95b87fee",
   "metadata": {},
   "source": [
    "## Environmental Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb34a593",
   "metadata": {},
   "source": [
    "### Make Environment to Observation Lookup "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f0c5816",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Env_Idx</th>\n",
       "      <th>Is_Phno_Idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>2222</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Env_Idx  Is_Phno_Idx\n",
       "0        0            0\n",
       "1        1          461\n",
       "2        2          936\n",
       "3        3         1785\n",
       "4        4         2222"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "obs_env_lookup_small = obs_env_lookup.drop(columns = 'Phno_Idx'\n",
    "                                    ).drop_duplicates(\n",
    "                                    ).sort_values('Env_Idx'\n",
    "                                    ).reset_index(drop = True)\n",
    "\n",
    "obs_env_lookup_small.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04fd878c",
   "metadata": {},
   "source": [
    "### Make S Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9313151b",
   "metadata": {},
   "outputs": [],
   "source": [
    "SMat = phno.loc[:, ['Env']].merge(soil.drop(columns = ['Unnamed: 0', 'Year'])).drop(columns = ['Env'])\n",
    "SMatNames = list(SMat)\n",
    "SMat = np.array(SMat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99170c8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(133057, 23)"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SMat.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10d08d56",
   "metadata": {},
   "outputs": [],
   "source": [
    "SMatSmall = np.zeros([\n",
    "    obs_env_lookup_small.shape[0], # unique environments\n",
    "    len(SMatNames)                 # Cin\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1ac47b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in obs_env_lookup_small.index:\n",
    "    ii = obs_env_lookup_small.loc[(obs_env_lookup_small.Env_Idx == i), 'Is_Phno_Idx']\n",
    "    ii = int(ii)\n",
    "    SMatSmall[i, ] = SMat[ii, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7af6f4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# overwrite with much smaller version (rows from 135793 -> 236)\n",
    "SMat = SMatSmall.copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dac4543f",
   "metadata": {},
   "source": [
    "### Make W Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1dae4c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input: (N,Cin,Lin)(N,Cin,Lin) or (Cin,Lin)(Cin,Lin)\n",
    "wthr.DOY = wthr.DOY.astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c2f0502",
   "metadata": {},
   "outputs": [],
   "source": [
    "WMatNames = list(wthr.drop(columns = ['Unnamed: 0', 'Env', 'Year', 'Date', 'DOY']))\n",
    "WMat = np.zeros([                  # Pytorch uses\n",
    "    obs_env_lookup_small.shape[0], # N\n",
    "    len(WMatNames),                # Cin\n",
    "    np.max(wthr.DOY)               # Lin\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a16010f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 236/236 [00:00<00:00, 258.22it/s]\n"
     ]
    }
   ],
   "source": [
    "for i in tqdm.tqdm(obs_env_lookup_small.index):\n",
    "    ii = obs_env_lookup_small.loc[(obs_env_lookup_small.Env_Idx == i), 'Is_Phno_Idx']\n",
    "    ii = int(ii)\n",
    "\n",
    "    ith_Env = phno_geno.loc[(phno_geno.index == ii), 'Env']\n",
    "    ith_Env = list(ith_Env)[0]\n",
    "\n",
    "    # selected data is transposed to match correct shape\n",
    "    wthr_mask = (wthr.Env == ith_Env)\n",
    "    WMat[i, :, :] = wthr.loc[wthr_mask, \n",
    "                       ].sort_values('DOY'\n",
    "                       ).drop(columns = ['Unnamed: 0', 'Env', \n",
    "                                         'Year', 'Date', 'DOY']).T"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a11f0094",
   "metadata": {},
   "source": [
    "## Prep CGMV?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3555aea",
   "metadata": {},
   "outputs": [],
   "source": [
    "cgmv_Env = phno_geno.loc[:, ['Env', 'Env_Idx']\n",
    "                   ].drop_duplicates(\n",
    "                   ).merge(cgmv\n",
    "                   ).sort_values('Env_Idx'\n",
    "                   ).reset_index(\n",
    "                   ).drop(columns = ['index', 'Env', 'Env_Idx', 'Unnamed: 0', 'Year'])\n",
    "\n",
    "MMatNames = list(cgmv_Env)\n",
    "MMat = np.asarray(cgmv_Env)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b3405c8",
   "metadata": {},
   "source": [
    "# Save data\n",
    "This will streamline model generation. I'll just need to load these files in and can directly begin modeling."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7defc79",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save(cache_path+'SMatNames.npy', SMatNames)\n",
    "np.save(cache_path+'WMatNames.npy', WMatNames)\n",
    "np.save(cache_path+'MMatNames.npy', MMatNames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51ce6d71",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save(cache_path+'SMat.npy', SMat)\n",
    "np.save(cache_path+'WMat.npy', WMat)\n",
    "np.save(cache_path+'MMat.npy', MMat)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
