{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6356970",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Hacky way to schedule. Here I'm setting these to sleep until the gpus should be free.\n",
    "# At the end of the notebooks  os._exit(00) will kill the kernel freeing the gpu. \n",
    "#                          Hours to wait\n",
    "# import time; time.sleep( 6 * (60*60))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aadccc0",
   "metadata": {},
   "source": [
    "# G only KEGG based network architecture\n",
    "\n",
    "> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bc2f33b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "import plotly.express as px\n",
    "import plotly.io as pio\n",
    "pio.templates.default = \"plotly_white\"\n",
    "\n",
    "# import hilbertcurve\n",
    "# from hilbertcurve.hilbertcurve import HilbertCurve\n",
    "\n",
    "from EnvDL.core import * # includes remove_matching_files\n",
    "from EnvDL.dna import *\n",
    "from EnvDL.dlfn import *\n",
    "\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7f0d9c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from graphviz import Digraph\n",
    "import torchviz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb4600ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "cache_path = '../nbs_artifacts/01.25_g2fc_demo_G_ACGT_graph/'\n",
    "ensure_dir_path_exists(dir_path = cache_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b08eeda",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24a78dda",
   "metadata": {},
   "outputs": [],
   "source": [
    "load_from = '../nbs_artifacts/01.03_g2fc_prep_matrices/'\n",
    "# phno = pd.read_csv(load_from+'phno.csv')\n",
    "phno_geno = pd.read_csv(load_from+'phno_geno.csv')\n",
    "phno = phno_geno\n",
    "\n",
    "obs_geno_lookup = np.load(load_from+'obs_geno_lookup.npy') # Phno_Idx\tGeno_Idx\tIs_Phno_Idx\n",
    "YMat = np.load(load_from+'YMat.npy')\n",
    "# GMat = np.load(load_from+'GMat.npy')\n",
    "# ACGT_OneHot = np.load(load_from+'ACGT_OneHot.npy')\n",
    "# ACGT = np.load(load_from+'ACGT.npy')\n",
    "# ACGT_hilb = np.load(load_from+'ACGT_hilb.npy')\n",
    "# SMat = np.load(load_from+'SMat3.npy')\n",
    "# WMat = np.load(load_from+'WMat3.npy')\n",
    "# MMat = np.load(load_from+'MMat3.npy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce01a01c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using cuda device\n"
     ]
    }
   ],
   "source": [
    "# dataloader_batch_size = 8 #16 #64\n",
    "# run_epochs = 200\n",
    "\n",
    "use_gpu_num = 0\n",
    "\n",
    "# Imports --------------------------------------------------------------------\n",
    "import torch\n",
    "from torch.utils.data import Dataset\n",
    "from torch.utils.data import DataLoader\n",
    "from torch import nn\n",
    "\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "if use_gpu_num in [0, 1]: \n",
    "    torch.cuda.set_device(use_gpu_num)\n",
    "print(f\"Using {device} device\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc4a4118",
   "metadata": {},
   "source": [
    "## Simple case: Building a Neural Net from an arbitrary graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aae582d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:00<00:00, 62415.24it/s]\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (0)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"205pt\" height=\"332pt\"\n",
       " viewBox=\"0.00 0.00 205.39 332.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 328)\">\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-328 201.39,-328 201.39,4 -4,4\"/>\n",
       "<!-- yhat -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>yhat</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"86.7\" cy=\"-18\" rx=\"31.4\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"86.7\" y=\"-14.3\" font-family=\"Times,serif\" font-size=\"14.00\">yhat</text>\n",
       "</g>\n",
       "<!-- c1 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>c1</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"122.7\" cy=\"-90\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"122.7\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">c1</text>\n",
       "</g>\n",
       "<!-- c1&#45;&gt;yhat -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>c1&#45;&gt;yhat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M114.35,-72.76C110.05,-64.4 104.7,-54.02 99.86,-44.61\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"102.94,-42.93 95.25,-35.63 96.71,-46.13 102.94,-42.93\"/>\n",
       "</g>\n",
       "<!-- c2 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>c2</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"50.7\" cy=\"-90\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"50.7\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">c2</text>\n",
       "</g>\n",
       "<!-- c2&#45;&gt;yhat -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>c2&#45;&gt;yhat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M59.05,-72.76C63.34,-64.4 68.69,-54.02 73.53,-44.61\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"76.68,-46.13 78.14,-35.63 70.46,-42.93 76.68,-46.13\"/>\n",
       "</g>\n",
       "<!-- b1 -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>b1</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"122.7\" cy=\"-162\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"122.7\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">b1</text>\n",
       "</g>\n",
       "<!-- b1&#45;&gt;c1 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>b1&#45;&gt;c1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M122.7,-143.7C122.7,-135.98 122.7,-126.71 122.7,-118.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"126.2,-118.1 122.7,-108.1 119.2,-118.1 126.2,-118.1\"/>\n",
       "</g>\n",
       "<!-- b2 -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>b2</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"50.7\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"50.7\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">b2</text>\n",
       "</g>\n",
       "<!-- b2&#45;&gt;c2 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>b2&#45;&gt;c2</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M50.7,-215.87C50.7,-191.67 50.7,-147.21 50.7,-118.39\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"54.2,-118.19 50.7,-108.19 47.2,-118.19 54.2,-118.19\"/>\n",
       "</g>\n",
       "<!-- b2&#45;&gt;b1 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>b2&#45;&gt;b1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M65.27,-218.83C75.45,-208.94 89.22,-195.55 100.73,-184.36\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"103.17,-186.87 107.9,-177.38 98.29,-181.85 103.17,-186.87\"/>\n",
       "</g>\n",
       "<!-- a1_input -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>a1_input</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"146.7\" cy=\"-234\" rx=\"50.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"146.7\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">a1_input</text>\n",
       "</g>\n",
       "<!-- a1_input&#45;&gt;b1 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>a1_input&#45;&gt;b1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M140.89,-216.05C138.14,-208.06 134.81,-198.33 131.75,-189.4\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"135.01,-188.11 128.45,-179.79 128.38,-190.38 135.01,-188.11\"/>\n",
       "</g>\n",
       "<!-- a2_input -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>a2_input</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"50.7\" cy=\"-306\" rx=\"50.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"50.7\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">a2_input</text>\n",
       "</g>\n",
       "<!-- a2_input&#45;&gt;b2 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>a2_input&#45;&gt;b2</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M50.7,-287.7C50.7,-279.98 50.7,-270.71 50.7,-262.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"54.2,-262.1 50.7,-252.1 47.2,-262.1 54.2,-262.1\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph>"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the simplest possible way to do this:\n",
    "# input1 (a1) -> b1  -> c1 \\\n",
    "#             /             -> yhat\n",
    "# input1 (a2) -> b2  -> c2 /\n",
    "\n",
    "example_dict = {\n",
    "    'yhat':['c1', 'c2'], # <-  extra node that makes a prediction from the top level nodes in the graph\n",
    "      'c1':['b1'],\n",
    "      'c2':['b2'],\n",
    "      'b1':['a1_input', 'b2'],\n",
    "      'b2':['a2_input']}\n",
    "\n",
    "dot = Digraph()\n",
    "for key in tqdm(example_dict.keys()):\n",
    "    dot.node(key)\n",
    "    for value in example_dict[key]:\n",
    "        # edge takes a head/tail whereas edges takes name pairs concatednated (A, B -> AB)in a list\n",
    "        dot.edge(value, key)    \n",
    "dot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e53c5358",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['yhat'], ['a1_input', 'a2_input'])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# start by finding the top level -- all those keys which are theselves not values\n",
    "# helper function to get all keys and all value from a dict. Useful for when keys don't have unique values.\n",
    "def find_uniq_keys_values(input_dict):\n",
    "    all_keys = list(input_dict.keys())\n",
    "    all_values = []\n",
    "    for e in all_keys:\n",
    "        all_values.extend(input_dict[e])\n",
    "    all_values = list(set(all_values))\n",
    "\n",
    "    return({'all_keys': all_keys,\n",
    "           'all_values': all_values})\n",
    "\n",
    "res = find_uniq_keys_values(input_dict = example_dict)\n",
    "all_keys = res['all_keys']\n",
    "all_values = res['all_values']\n",
    "\n",
    "# use the keys to find the input/outputs of the graph\n",
    "output_nodes = [e for e in all_keys if e not in all_values]\n",
    "input_nodes = [e for e in all_values if e not in all_keys]\n",
    "\n",
    "(output_nodes, input_nodes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e1d05ca",
   "metadata": {},
   "source": [
    "### Find order that nodes in the graph should be called to have all dependencies run when they are called."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95ba96a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['yhat']\n",
      "['yhat', 'c1', 'c2']\n",
      "['yhat', 'c1', 'c2', 'b1']\n",
      "['yhat', 'c1', 'c2', 'b1', 'b2']\n"
     ]
    }
   ],
   "source": [
    "temp = example_dict.copy()\n",
    "\n",
    "# find the dependancies for run order from many dependancies to none\n",
    "# wrapper function to find the nodes that aren't any other nodes dependancies.\n",
    "def find_top_nodes(all_key_value_dict):\n",
    "    return([e for e in all_key_value_dict['all_keys'] if e not in all_key_value_dict['all_values']])\n",
    "# wrapper function to find the input nodes. They don't occur in the keys and thus won't be added to the list otherwise.\n",
    "# another way to do this would have been to \n",
    "def find_input_nodes(all_key_value_dict):\n",
    "    return([e for e in all_key_value_dict['all_values'] if e not in all_key_value_dict['all_keys']])\n",
    "\n",
    "no_dependants = find_input_nodes(all_key_value_dict = find_uniq_keys_values(input_dict = temp))\n",
    "\n",
    "# first pass. Same as the output nodes identified above\n",
    "dependancy_order = []\n",
    "\n",
    "# Then iterate\n",
    "for ith in range(100):\n",
    "    top_nodes = find_top_nodes(all_key_value_dict = find_uniq_keys_values(input_dict = temp))\n",
    "    if top_nodes == []:\n",
    "        break\n",
    "    else:\n",
    "        dependancy_order += top_nodes    \n",
    "        # remove nodes from the graph that are at the 'top' level and haven't already been removed\n",
    "        for key in [e for e in dependancy_order if e in temp.keys()]:\n",
    "             temp.pop(key)\n",
    "\n",
    "        \n",
    "    print(dependancy_order)\n",
    "\n",
    "# finally add in the inputs.        \n",
    "# dependancy_order += no_dependants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0c3f198",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['b2', 'b1', 'c2', 'c1', 'yhat']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reverse to get the order that the nodes should be called\n",
    "dependancy_order.reverse()\n",
    "dependancy_order"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b46aba7b",
   "metadata": {},
   "source": [
    "### Set up output tensor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc9180b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "y_true = np.random.uniform(-1, 1, 10).reshape(-1, 1) # (b, 3)\n",
    "y_true = torch.from_numpy(y_true)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66e35d29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[ 0.1338],\n",
       "        [-0.2289],\n",
       "        [ 0.9818],\n",
       "        [-0.7927],\n",
       "        [-0.6640],\n",
       "        [ 0.0265],\n",
       "        [ 0.8208],\n",
       "        [ 0.5117],\n",
       "        [-0.9432],\n",
       "        [ 0.6721]], dtype=torch.float64)"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_true"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69f7d429",
   "metadata": {},
   "source": [
    "### Set up dict of input tensors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c49f7adf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['a1_input', 'a2_input']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_dependants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a16af78f",
   "metadata": {},
   "outputs": [],
   "source": [
    "a1_input = torch.from_numpy(np.random.uniform(-1, 1, 20).reshape(-1, 2)) # (b, 2))\n",
    "a2_input = torch.from_numpy(np.random.uniform(-1, 1, 30).reshape(-1, 3)) # (b, 3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfdc5975",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'a1_input': tensor([[ 0.9501,  0.0389],\n",
       "         [ 0.0875,  0.5887],\n",
       "         [-0.1009, -0.7724],\n",
       "         [-0.2083,  0.1433],\n",
       "         [ 0.4781,  0.6267],\n",
       "         [ 0.6603, -0.0211],\n",
       "         [-0.1465, -0.1847],\n",
       "         [-0.8806, -0.2331],\n",
       "         [ 0.4776,  0.0646],\n",
       "         [-0.4393, -0.8689]]),\n",
       " 'a2_input': tensor([[-0.1532,  0.7324, -0.8856],\n",
       "         [ 0.2143,  0.7152, -0.9613],\n",
       "         [-0.6621,  0.8213, -0.0444],\n",
       "         [-0.7321,  0.6988, -0.3910],\n",
       "         [ 0.9717,  0.8391,  0.8297],\n",
       "         [ 0.7658, -0.0153, -0.9037],\n",
       "         [ 0.1787, -0.7985, -0.7837],\n",
       "         [-0.2730, -0.0121,  0.0575],\n",
       "         [-0.8051,  0.2993,  0.0398],\n",
       "         [-0.7323, -0.0048,  0.5394]])}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_tensor_dict = {}\n",
    "for e in no_dependants:\n",
    "    input_tensor_dict[e] = eval(e).to(torch.float) # I don't know if this is a safe thing to do.\n",
    "    \n",
    "input_tensor_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57511313",
   "metadata": {},
   "source": [
    "### Figure out expected input/output shapes\n",
    "\n",
    "_==NOTE! This assumes only dense connections!==_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28571664",
   "metadata": {},
   "outputs": [],
   "source": [
    "# now let's figure out the size that each node should expect as an input.\n",
    "# For now I assume that each node will have the same number of outputs as its inputs. \n",
    "# ==NOTE== if instead each node had a fixed output (e.g. 10 channels) then all non-inputs could be replaced with this value\n",
    "example_dict_input_size = example_dict.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "797f1e97",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add in channel number\n",
    "for e in no_dependants:\n",
    "    example_dict_input_size[e] = list(input_tensor_dict[e].shape)[1] # <- NOTE! THIS ASSUMES ONLY DENSE CONNECTIONS (i.e. only the 1st dim is needed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63405f74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'yhat': 8, 'c1': 5, 'c2': 3, 'b1': 5, 'b2': 3, 'a1_input': 2, 'a2_input': 3}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# walk through dependancy order\n",
    "for e in dependancy_order:\n",
    "    # get the channel number for each dependency and sum to get the input channels expected\n",
    "    res = sum([example_dict_input_size[key] for key in example_dict_input_size[e]])\n",
    "\n",
    "    # overwrite with the value\n",
    "    example_dict_input_size[e] = res\n",
    "\n",
    "example_dict_input_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc27ddf6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'yhat': 1, 'c1': 5, 'c2': 3, 'b1': 5, 'b2': 3, 'a1_input': 2, 'a2_input': 3}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now make an output dict\n",
    "example_dict_output_size = example_dict_input_size.copy()\n",
    "example_dict_output_size['yhat'] = 1\n",
    "example_dict_output_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4cafe48",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75c374d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['yhat', 'c1', 'c2', 'b1', 'b2'])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_dict.keys()\n",
    "# example_dict_input_size"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b348c27",
   "metadata": {},
   "source": [
    "### Set up DataSet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6046e283",
   "metadata": {},
   "source": [
    "Now we have\n",
    "\n",
    "- A dictionary with the connections: `example_dict`\n",
    "- The expected input sizes for each node: `example_dict_input_size`\n",
    "- A dictionary with the input tensors: `input_tensor_dict`\n",
    "- A list of the input tensors' names: `no_dependants` \n",
    "- A list with the order that each module should be called: `dependancy_order`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e4a8ea2",
   "metadata": {},
   "outputs": [],
   "source": [
    "class ListDataset(Dataset): # for any G containing matix with many (phno) to one (geno)\n",
    "    def __init__(self, \n",
    "                 y, \n",
    "                 x_list,\n",
    "                 transform = None, target_transform = None,\n",
    "                 use_gpu_num = 0,\n",
    "                 device = 'cuda',\n",
    "                 **kwargs \n",
    "                ):\n",
    "        self.device = device\n",
    "        self.y = y \n",
    "        self.x_list = x_list,\n",
    "        self.transform = transform\n",
    "        self.target_transform = target_transform    \n",
    "        \n",
    "    def __len__(self):\n",
    "        return len(self.y)\n",
    "    \n",
    "    def __getitem__(self, idx):\n",
    "        y_idx =self.y[idx]\n",
    "        x_idx =[x[idx] for x in self.x_list[0]] # must be the 0th entry because it's in a tuple\n",
    "        \n",
    "        if self.target_transform:\n",
    "            y_idx = self.transform(y_idx)\n",
    "            x_idx = [self.transform(x) for x in x_idx]\n",
    "            \n",
    "        return y_idx, x_idx\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "803387eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[tensor([[-0.7927],\n",
       "         [ 0.8208],\n",
       "         [ 0.1338],\n",
       "         [ 0.5117],\n",
       "         [ 0.9818]], dtype=torch.float64),\n",
       " [tensor([[-0.2083,  0.1433],\n",
       "          [-0.1465, -0.1847],\n",
       "          [ 0.9501,  0.0389],\n",
       "          [-0.8806, -0.2331],\n",
       "          [-0.1009, -0.7724]]),\n",
       "  tensor([[-0.7321,  0.6988, -0.3910],\n",
       "          [ 0.1787, -0.7985, -0.7837],\n",
       "          [-0.1532,  0.7324, -0.8856],\n",
       "          [-0.2730, -0.0121,  0.0575],\n",
       "          [-0.6621,  0.8213, -0.0444]])]]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "training_dataloader = DataLoader(ListDataset(\n",
    "        y = y_true,\n",
    "        x_list = [input_tensor_dict[key] for key in input_tensor_dict.keys()],\n",
    "        use_gpu_num = 0,\n",
    "        device = 'cuda',\n",
    "    ),\n",
    "    batch_size = 5,\n",
    "    shuffle = True\n",
    ")\n",
    "next(iter(training_dataloader))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6568d72",
   "metadata": {},
   "source": [
    "### Set up NeuralNetwork"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e592b4e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input_tensor_names = list(input_tensor_dict.keys())\n",
    "# input_tensor_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ad25a1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first pass is by hand.\n",
    "class NeuralNetwork(nn.Module):\n",
    "    def __init__(self, \n",
    "                 example_dict, # contains the node (excluding input tensors)\n",
    "                 example_dict_input_size, # contains the input sizes (including the tensors)\n",
    "                 example_dict_output_size,\n",
    "                 input_tensor_names,\n",
    "                 dependancy_order\n",
    "                ):\n",
    "        super(NeuralNetwork, self).__init__()\n",
    "        def Linear_block(in_size, out_size, drop_pr):\n",
    "            block = nn.Sequential(\n",
    "                nn.Linear(in_size, out_size),\n",
    "                nn.ReLU(),\n",
    "                nn.Dropout(drop_pr))\n",
    "            return(block)   \n",
    "        \n",
    "        self.nn_dict = {}\n",
    "        # iterate through the input dictionary and add a linear block for each node (key) included \n",
    "        for key in example_dict.keys():\n",
    "            self.nn_dict[key] = Linear_block(in_size=example_dict_input_size[key], \n",
    "                                             out_size=example_dict_output_size[key], \n",
    "                                             drop_pr=0)\n",
    "        # things for get_input_node in forward to work.\n",
    "        self.example_dict = example_dict\n",
    "        self.input_tensor_names = input_tensor_names\n",
    "        self.dependancy_order = dependancy_order\n",
    "        \n",
    "        self.input_tensor_lookup = dict(zip(input_tensor_names, \n",
    "                                            [i for i in range(len(input_tensor_names))]))\n",
    "        self.result_list = []\n",
    "        self.result_list_lookup = {}\n",
    "\n",
    "    def forward(self, x):\n",
    "        # Note: x will be a list. input_tensor_lookup will contain the name: list index pairs.\n",
    "        # I use a dict instead of a list comprehension here because there could be an arbitrarily\n",
    "        # large number of inputs in the list. \n",
    "        def get_input_node(self, input_node, get_x):    \n",
    "            if input_node in self.input_tensor_names:\n",
    "                return(get_x[self.input_tensor_lookup[input_node]])\n",
    "            else:\n",
    "                return(self.result_list[self.result_list_lookup[input_node]])\n",
    "            \n",
    "        for key in self.dependancy_order:\n",
    "            input_nodes = self.example_dict[key]\n",
    "            \n",
    "            if key not in self.result_list_lookup.keys():\n",
    "                # I realllllly hope these are being copied by reference...\n",
    "                # must be first before growing result_list\n",
    "                self.result_list_lookup[key] = len(self.result_list_lookup)\n",
    "                self.result_list += [self.nn_dict[key](torch.concat(\n",
    "                    [get_input_node(self, input_node = e, get_x = x) for e in input_nodes], \n",
    "                    1))]\n",
    "            else:\n",
    "                # Note! here I'm saving the tensor without wrapping it in a list.\n",
    "                self.result_list[self.result_list_lookup[key]] = self.nn_dict[key](torch.concat(\n",
    "                    [get_input_node(self, input_node = e, get_x = x) for e in input_nodes], \n",
    "                    1))\n",
    "                \n",
    "        return self.result_list[self.result_list_lookup['yhat']]\n",
    "\n",
    "model = NeuralNetwork(example_dict = example_dict, \n",
    "                      example_dict_input_size = example_dict_input_size,\n",
    "                      example_dict_output_size = example_dict_output_size,\n",
    "                      input_tensor_names = list(input_tensor_dict.keys()),\n",
    "                      dependancy_order = dependancy_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45764795",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[0.],\n",
       "        [0.],\n",
       "        [0.],\n",
       "        [0.],\n",
       "        [0.]], grad_fn=<ReluBackward0>)"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "yhat = model(next(iter(training_dataloader))[1])\n",
    "yhat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c3ac529",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (0)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"604pt\" height=\"886pt\"\n",
       " viewBox=\"0.00 0.00 603.59 886.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(0.92 0.92) rotate(0) translate(4 956)\">\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-956 650,-956 650,4 -4,4\"/>\n",
       "<!-- 140169411065040 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>140169411065040</title>\n",
       "<polygon fill=\"#caff70\" stroke=\"black\" points=\"470.5,-31 416.5,-31 416.5,0 470.5,0 470.5,-31\"/>\n",
       "<text text-anchor=\"middle\" x=\"443.5\" y=\"-7\" font-family=\"monospace\" font-size=\"10.00\"> ()</text>\n",
       "</g>\n",
       "<!-- 140169393687136 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>140169393687136</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"491,-86 396,-86 396,-67 491,-67 491,-86\"/>\n",
       "<text text-anchor=\"middle\" x=\"443.5\" y=\"-74\" font-family=\"monospace\" font-size=\"10.00\">MeanBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393687136&#45;&gt;140169411065040 -->\n",
       "<g id=\"edge41\" class=\"edge\">\n",
       "<title>140169393687136&#45;&gt;140169411065040</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M443.5,-66.79C443.5,-60.07 443.5,-50.4 443.5,-41.34\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"447,-41.19 443.5,-31.19 440,-41.19 447,-41.19\"/>\n",
       "</g>\n",
       "<!-- 140169393692368 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>140169393692368</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"491,-141 396,-141 396,-122 491,-122 491,-141\"/>\n",
       "<text text-anchor=\"middle\" x=\"443.5\" y=\"-129\" font-family=\"monospace\" font-size=\"10.00\">ReluBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393692368&#45;&gt;140169393687136 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>140169393692368&#45;&gt;140169393687136</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M443.5,-121.75C443.5,-114.8 443.5,-104.85 443.5,-96.13\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"447,-96.09 443.5,-86.09 440,-96.09 447,-96.09\"/>\n",
       "</g>\n",
       "<!-- 140169393686944 -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>140169393686944</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"494,-196 393,-196 393,-177 494,-177 494,-196\"/>\n",
       "<text text-anchor=\"middle\" x=\"443.5\" y=\"-184\" font-family=\"monospace\" font-size=\"10.00\">AddmmBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393686944&#45;&gt;140169393692368 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>140169393686944&#45;&gt;140169393692368</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M443.5,-176.75C443.5,-169.8 443.5,-159.85 443.5,-151.13\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"447,-151.09 443.5,-141.09 440,-151.09 447,-151.09\"/>\n",
       "</g>\n",
       "<!-- 140169393690880 -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>140169393690880</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"377,-251 276,-251 276,-232 377,-232 377,-251\"/>\n",
       "<text text-anchor=\"middle\" x=\"326.5\" y=\"-239\" font-family=\"monospace\" font-size=\"10.00\">AccumulateGrad</text>\n",
       "</g>\n",
       "<!-- 140169393690880&#45;&gt;140169393686944 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>140169393690880&#45;&gt;140169393686944</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M345.3,-231.98C364.12,-223.46 393.32,-210.23 415,-200.41\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"416.7,-203.48 424.36,-196.17 413.81,-197.11 416.7,-203.48\"/>\n",
       "</g>\n",
       "<!-- 140169422730448 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>140169422730448</title>\n",
       "<polygon fill=\"lightblue\" stroke=\"black\" points=\"344.5,-318 290.5,-318 290.5,-287 344.5,-287 344.5,-318\"/>\n",
       "<text text-anchor=\"middle\" x=\"317.5\" y=\"-294\" font-family=\"monospace\" font-size=\"10.00\"> (1)</text>\n",
       "</g>\n",
       "<!-- 140169422730448&#45;&gt;140169393690880 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>140169422730448&#45;&gt;140169393690880</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M319.72,-286.92C320.91,-279.13 322.39,-269.47 323.66,-261.15\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"327.12,-261.66 325.17,-251.25 320.2,-260.61 327.12,-261.66\"/>\n",
       "</g>\n",
       "<!-- 140169393686656 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>140169393686656</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"488,-251 399,-251 399,-232 488,-232 488,-251\"/>\n",
       "<text text-anchor=\"middle\" x=\"443.5\" y=\"-239\" font-family=\"monospace\" font-size=\"10.00\">CatBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393686656&#45;&gt;140169393686944 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>140169393686656&#45;&gt;140169393686944</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M443.5,-231.75C443.5,-224.8 443.5,-214.85 443.5,-206.13\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"447,-206.09 443.5,-196.09 440,-206.09 447,-206.09\"/>\n",
       "</g>\n",
       "<!-- 140169393689584 -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>140169393689584</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"496,-312 401,-312 401,-293 496,-293 496,-312\"/>\n",
       "<text text-anchor=\"middle\" x=\"448.5\" y=\"-300\" font-family=\"monospace\" font-size=\"10.00\">ReluBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393689584&#45;&gt;140169393686656 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>140169393689584&#45;&gt;140169393686656</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M447.76,-292.79C447.06,-284.52 445.98,-271.79 445.09,-261.21\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"448.57,-260.91 444.24,-251.24 441.6,-261.5 448.57,-260.91\"/>\n",
       "</g>\n",
       "<!-- 140169393692224 -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>140169393692224</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"508,-379 407,-379 407,-360 508,-360 508,-379\"/>\n",
       "<text text-anchor=\"middle\" x=\"457.5\" y=\"-367\" font-family=\"monospace\" font-size=\"10.00\">AddmmBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393692224&#45;&gt;140169393689584 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>140169393692224&#45;&gt;140169393689584</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M456.29,-359.73C454.96,-350.18 452.81,-334.62 451.1,-322.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"454.53,-321.55 449.69,-312.13 447.6,-322.51 454.53,-321.55\"/>\n",
       "</g>\n",
       "<!-- 140169393692128 -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>140169393692128</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"646,-440 545,-440 545,-421 646,-421 646,-440\"/>\n",
       "<text text-anchor=\"middle\" x=\"595.5\" y=\"-428\" font-family=\"monospace\" font-size=\"10.00\">AccumulateGrad</text>\n",
       "</g>\n",
       "<!-- 140169393692128&#45;&gt;140169393692224 -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>140169393692128&#45;&gt;140169393692224</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M575.43,-420.92C552.13,-410.96 513.49,-394.44 486.82,-383.04\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"488.16,-379.8 477.59,-379.09 485.41,-386.24 488.16,-379.8\"/>\n",
       "</g>\n",
       "<!-- 140169422731648 -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>140169422731648</title>\n",
       "<polygon fill=\"lightblue\" stroke=\"black\" points=\"623.5,-507 569.5,-507 569.5,-476 623.5,-476 623.5,-507\"/>\n",
       "<text text-anchor=\"middle\" x=\"596.5\" y=\"-483\" font-family=\"monospace\" font-size=\"10.00\"> (5)</text>\n",
       "</g>\n",
       "<!-- 140169422731648&#45;&gt;140169393692128 -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>140169422731648&#45;&gt;140169393692128</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M596.25,-475.92C596.12,-468.22 595.96,-458.69 595.82,-450.43\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"599.32,-450.19 595.65,-440.25 592.32,-450.31 599.32,-450.19\"/>\n",
       "</g>\n",
       "<!-- 140169393692080 -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>140169393692080</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"429,-440 340,-440 340,-421 429,-421 429,-440\"/>\n",
       "<text text-anchor=\"middle\" x=\"384.5\" y=\"-428\" font-family=\"monospace\" font-size=\"10.00\">CatBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393692080&#45;&gt;140169393692224 -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>140169393692080&#45;&gt;140169393692224</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M395.27,-420.79C406.7,-411.56 424.98,-396.78 438.78,-385.63\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"441.11,-388.25 446.68,-379.24 436.71,-382.81 441.11,-388.25\"/>\n",
       "</g>\n",
       "<!-- 140169393692992 -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>140169393692992</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"432,-501 337,-501 337,-482 432,-482 432,-501\"/>\n",
       "<text text-anchor=\"middle\" x=\"384.5\" y=\"-489\" font-family=\"monospace\" font-size=\"10.00\">ReluBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393692992&#45;&gt;140169393692080 -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>140169393692992&#45;&gt;140169393692080</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M384.5,-481.79C384.5,-473.6 384.5,-461.06 384.5,-450.55\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"388,-450.24 384.5,-440.24 381,-450.24 388,-450.24\"/>\n",
       "</g>\n",
       "<!-- 140169393693088 -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>140169393693088</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"444,-568 343,-568 343,-549 444,-549 444,-568\"/>\n",
       "<text text-anchor=\"middle\" x=\"393.5\" y=\"-556\" font-family=\"monospace\" font-size=\"10.00\">AddmmBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393693088&#45;&gt;140169393692992 -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>140169393693088&#45;&gt;140169393692992</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M392.29,-548.73C390.96,-539.18 388.81,-523.62 387.1,-511.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"390.53,-510.55 385.69,-501.13 383.6,-511.51 390.53,-510.55\"/>\n",
       "</g>\n",
       "<!-- 140169393684784 -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>140169393684784</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"529,-629 428,-629 428,-610 529,-610 529,-629\"/>\n",
       "<text text-anchor=\"middle\" x=\"478.5\" y=\"-617\" font-family=\"monospace\" font-size=\"10.00\">AccumulateGrad</text>\n",
       "</g>\n",
       "<!-- 140169393684784&#45;&gt;140169393693088 -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>140169393684784&#45;&gt;140169393693088</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M465.96,-609.79C452.32,-600.33 430.31,-585.05 414.12,-573.81\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"416.03,-570.88 405.82,-568.05 412.04,-576.63 416.03,-570.88\"/>\n",
       "</g>\n",
       "<!-- 140169421990608 -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>140169421990608</title>\n",
       "<polygon fill=\"lightblue\" stroke=\"black\" points=\"505.5,-696 451.5,-696 451.5,-665 505.5,-665 505.5,-696\"/>\n",
       "<text text-anchor=\"middle\" x=\"478.5\" y=\"-672\" font-family=\"monospace\" font-size=\"10.00\"> (5)</text>\n",
       "</g>\n",
       "<!-- 140169421990608&#45;&gt;140169393684784 -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>140169421990608&#45;&gt;140169393684784</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M478.5,-664.92C478.5,-657.22 478.5,-647.69 478.5,-639.43\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"482,-639.25 478.5,-629.25 475,-639.25 482,-639.25\"/>\n",
       "</g>\n",
       "<!-- 140169393693136 -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>140169393693136</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"410,-629 321,-629 321,-610 410,-610 410,-629\"/>\n",
       "<text text-anchor=\"middle\" x=\"365.5\" y=\"-617\" font-family=\"monospace\" font-size=\"10.00\">CatBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393693136&#45;&gt;140169393693088 -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>140169393693136&#45;&gt;140169393693088</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M369.63,-609.79C373.64,-601.34 379.85,-588.25 384.94,-577.55\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"388.23,-578.78 389.35,-568.24 381.9,-575.78 388.23,-578.78\"/>\n",
       "</g>\n",
       "<!-- 140169393684880 -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>140169393684880</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"367,-690 272,-690 272,-671 367,-671 367,-690\"/>\n",
       "<text text-anchor=\"middle\" x=\"319.5\" y=\"-678\" font-family=\"monospace\" font-size=\"10.00\">ReluBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393684880&#45;&gt;140169393693136 -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>140169393684880&#45;&gt;140169393693136</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M326.29,-670.79C333.15,-661.99 343.93,-648.17 352.45,-637.23\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"355.3,-639.28 358.68,-629.24 349.78,-634.98 355.3,-639.28\"/>\n",
       "</g>\n",
       "<!-- 140169393694336 -->\n",
       "<g id=\"node35\" class=\"node\">\n",
       "<title>140169393694336</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"303,-629 214,-629 214,-610 303,-610 303,-629\"/>\n",
       "<text text-anchor=\"middle\" x=\"258.5\" y=\"-617\" font-family=\"monospace\" font-size=\"10.00\">CatBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393684880&#45;&gt;140169393694336 -->\n",
       "<g id=\"edge34\" class=\"edge\">\n",
       "<title>140169393684880&#45;&gt;140169393694336</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M310.5,-670.79C301.13,-661.73 286.25,-647.34 274.8,-636.27\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"277.16,-633.68 267.54,-629.24 272.29,-638.71 277.16,-633.68\"/>\n",
       "</g>\n",
       "<!-- 140169393693232 -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>140169393693232</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"370,-757 269,-757 269,-738 370,-738 370,-757\"/>\n",
       "<text text-anchor=\"middle\" x=\"319.5\" y=\"-745\" font-family=\"monospace\" font-size=\"10.00\">AddmmBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393693232&#45;&gt;140169393684880 -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>140169393693232&#45;&gt;140169393684880</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M319.5,-737.73C319.5,-728.18 319.5,-712.62 319.5,-700.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"323,-700.13 319.5,-690.13 316,-700.13 323,-700.13\"/>\n",
       "</g>\n",
       "<!-- 140169393693664 -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>140169393693664</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"317,-818 216,-818 216,-799 317,-799 317,-818\"/>\n",
       "<text text-anchor=\"middle\" x=\"266.5\" y=\"-806\" font-family=\"monospace\" font-size=\"10.00\">AccumulateGrad</text>\n",
       "</g>\n",
       "<!-- 140169393693664&#45;&gt;140169393693232 -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>140169393693664&#45;&gt;140169393693232</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M274.32,-798.79C282.31,-789.91 294.89,-775.89 304.76,-764.91\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"307.57,-767.02 311.65,-757.24 302.36,-762.34 307.57,-767.02\"/>\n",
       "</g>\n",
       "<!-- 140169449682528 -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>140169449682528</title>\n",
       "<polygon fill=\"lightblue\" stroke=\"black\" points=\"293.5,-885 239.5,-885 239.5,-854 293.5,-854 293.5,-885\"/>\n",
       "<text text-anchor=\"middle\" x=\"266.5\" y=\"-861\" font-family=\"monospace\" font-size=\"10.00\"> (3)</text>\n",
       "</g>\n",
       "<!-- 140169449682528&#45;&gt;140169393693664 -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>140169449682528&#45;&gt;140169393693664</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M266.5,-853.92C266.5,-846.22 266.5,-836.69 266.5,-828.43\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"270,-828.25 266.5,-818.25 263,-828.25 270,-828.25\"/>\n",
       "</g>\n",
       "<!-- 140169393693424 -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>140169393693424</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"412,-818 335,-818 335,-799 412,-799 412,-818\"/>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-806\" font-family=\"monospace\" font-size=\"10.00\">TBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393693424&#45;&gt;140169393693232 -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>140169393693424&#45;&gt;140169393693232</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M365.53,-798.79C357.4,-789.91 344.57,-775.89 334.52,-764.91\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"336.83,-762.26 327.5,-757.24 331.67,-766.98 336.83,-762.26\"/>\n",
       "</g>\n",
       "<!-- 140169393694240 -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>140169393694240</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"424,-879 323,-879 323,-860 424,-860 424,-879\"/>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-867\" font-family=\"monospace\" font-size=\"10.00\">AccumulateGrad</text>\n",
       "</g>\n",
       "<!-- 140169393694240&#45;&gt;140169393693424 -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>140169393694240&#45;&gt;140169393693424</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M373.5,-859.79C373.5,-851.6 373.5,-839.06 373.5,-828.55\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"377,-828.24 373.5,-818.24 370,-828.24 377,-828.24\"/>\n",
       "</g>\n",
       "<!-- 140169449676768 -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>140169449676768</title>\n",
       "<polygon fill=\"lightblue\" stroke=\"black\" points=\"403,-952 344,-952 344,-921 403,-921 403,-952\"/>\n",
       "<text text-anchor=\"middle\" x=\"373.5\" y=\"-928\" font-family=\"monospace\" font-size=\"10.00\"> (3, 3)</text>\n",
       "</g>\n",
       "<!-- 140169449676768&#45;&gt;140169393694240 -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>140169449676768&#45;&gt;140169393694240</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M373.5,-920.75C373.5,-911.39 373.5,-899.19 373.5,-889.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"377,-889.02 373.5,-879.02 370,-889.02 377,-889.02\"/>\n",
       "</g>\n",
       "<!-- 140169393693040 -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>140169393693040</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"624,-629 547,-629 547,-610 624,-610 624,-629\"/>\n",
       "<text text-anchor=\"middle\" x=\"585.5\" y=\"-617\" font-family=\"monospace\" font-size=\"10.00\">TBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393693040&#45;&gt;140169393693088 -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>140169393693040&#45;&gt;140169393693088</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M557.58,-609.92C524.31,-599.7 468.55,-582.56 431.4,-571.15\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"432.04,-567.68 421.45,-568.09 429.98,-574.37 432.04,-567.68\"/>\n",
       "</g>\n",
       "<!-- 140169393685072 -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>140169393685072</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"636,-690 535,-690 535,-671 636,-671 636,-690\"/>\n",
       "<text text-anchor=\"middle\" x=\"585.5\" y=\"-678\" font-family=\"monospace\" font-size=\"10.00\">AccumulateGrad</text>\n",
       "</g>\n",
       "<!-- 140169393685072&#45;&gt;140169393693040 -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>140169393685072&#45;&gt;140169393693040</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M585.5,-670.79C585.5,-662.6 585.5,-650.06 585.5,-639.55\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"589,-639.24 585.5,-629.24 582,-639.24 589,-639.24\"/>\n",
       "</g>\n",
       "<!-- 140173173709616 -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>140173173709616</title>\n",
       "<polygon fill=\"lightblue\" stroke=\"black\" points=\"615,-763 556,-763 556,-732 615,-732 615,-763\"/>\n",
       "<text text-anchor=\"middle\" x=\"585.5\" y=\"-739\" font-family=\"monospace\" font-size=\"10.00\"> (5, 5)</text>\n",
       "</g>\n",
       "<!-- 140173173709616&#45;&gt;140169393685072 -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>140173173709616&#45;&gt;140169393685072</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M585.5,-731.75C585.5,-722.39 585.5,-710.19 585.5,-700.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"589,-700.02 585.5,-690.02 582,-700.02 589,-700.02\"/>\n",
       "</g>\n",
       "<!-- 140169393692464 -->\n",
       "<g id=\"node28\" class=\"node\">\n",
       "<title>140169393692464</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"526,-440 449,-440 449,-421 526,-421 526,-440\"/>\n",
       "<text text-anchor=\"middle\" x=\"487.5\" y=\"-428\" font-family=\"monospace\" font-size=\"10.00\">TBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393692464&#45;&gt;140169393692224 -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>140169393692464&#45;&gt;140169393692224</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M483.07,-420.79C478.73,-412.25 471.98,-398.98 466.51,-388.22\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"469.6,-386.57 461.94,-379.24 463.36,-389.74 469.6,-386.57\"/>\n",
       "</g>\n",
       "<!-- 140169393684832 -->\n",
       "<g id=\"node29\" class=\"node\">\n",
       "<title>140169393684832</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"551,-501 450,-501 450,-482 551,-482 551,-501\"/>\n",
       "<text text-anchor=\"middle\" x=\"500.5\" y=\"-489\" font-family=\"monospace\" font-size=\"10.00\">AccumulateGrad</text>\n",
       "</g>\n",
       "<!-- 140169393684832&#45;&gt;140169393692464 -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>140169393684832&#45;&gt;140169393692464</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M498.58,-481.79C496.76,-473.52 493.95,-460.79 491.62,-450.21\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"495,-449.25 489.43,-440.24 488.16,-450.76 495,-449.25\"/>\n",
       "</g>\n",
       "<!-- 140169433297808 -->\n",
       "<g id=\"node30\" class=\"node\">\n",
       "<title>140169433297808</title>\n",
       "<polygon fill=\"lightblue\" stroke=\"black\" points=\"530,-574 471,-574 471,-543 530,-543 530,-574\"/>\n",
       "<text text-anchor=\"middle\" x=\"500.5\" y=\"-550\" font-family=\"monospace\" font-size=\"10.00\"> (5, 5)</text>\n",
       "</g>\n",
       "<!-- 140169433297808&#45;&gt;140169393684832 -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>140169433297808&#45;&gt;140169393684832</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M500.5,-542.75C500.5,-533.39 500.5,-521.19 500.5,-511.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"504,-511.02 500.5,-501.02 497,-511.02 504,-511.02\"/>\n",
       "</g>\n",
       "<!-- 140169393692656 -->\n",
       "<g id=\"node31\" class=\"node\">\n",
       "<title>140169393692656</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"374,-379 279,-379 279,-360 374,-360 374,-379\"/>\n",
       "<text text-anchor=\"middle\" x=\"326.5\" y=\"-367\" font-family=\"monospace\" font-size=\"10.00\">ReluBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393692656&#45;&gt;140169393686656 -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>140169393692656&#45;&gt;140169393686656</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M332.47,-359.97C343.38,-344.65 367.61,-311.73 391.5,-287 401.69,-276.45 414.14,-265.73 424.31,-257.45\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"426.58,-260.11 432.21,-251.13 422.21,-254.65 426.58,-260.11\"/>\n",
       "</g>\n",
       "<!-- 140169393693184 -->\n",
       "<g id=\"node32\" class=\"node\">\n",
       "<title>140169393693184</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"281,-568 180,-568 180,-549 281,-549 281,-568\"/>\n",
       "<text text-anchor=\"middle\" x=\"230.5\" y=\"-556\" font-family=\"monospace\" font-size=\"10.00\">AddmmBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393693184&#45;&gt;140169393692656 -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>140169393693184&#45;&gt;140169393692656</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M234.87,-548.98C249.56,-520.38 297.27,-427.43 317.41,-388.21\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"320.65,-389.57 322.1,-379.07 314.42,-386.37 320.65,-389.57\"/>\n",
       "</g>\n",
       "<!-- 140169393685120 -->\n",
       "<g id=\"node33\" class=\"node\">\n",
       "<title>140169393685120</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"101,-629 0,-629 0,-610 101,-610 101,-629\"/>\n",
       "<text text-anchor=\"middle\" x=\"50.5\" y=\"-617\" font-family=\"monospace\" font-size=\"10.00\">AccumulateGrad</text>\n",
       "</g>\n",
       "<!-- 140169393685120&#45;&gt;140169393693184 -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>140169393685120&#45;&gt;140169393693184</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M76.67,-609.92C107.73,-599.74 159.69,-582.71 194.52,-571.29\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"195.88,-574.53 204.29,-568.09 193.7,-567.88 195.88,-574.53\"/>\n",
       "</g>\n",
       "<!-- 140173173709456 -->\n",
       "<g id=\"node34\" class=\"node\">\n",
       "<title>140173173709456</title>\n",
       "<polygon fill=\"lightblue\" stroke=\"black\" points=\"77.5,-696 23.5,-696 23.5,-665 77.5,-665 77.5,-696\"/>\n",
       "<text text-anchor=\"middle\" x=\"50.5\" y=\"-672\" font-family=\"monospace\" font-size=\"10.00\"> (3)</text>\n",
       "</g>\n",
       "<!-- 140173173709456&#45;&gt;140169393685120 -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>140173173709456&#45;&gt;140169393685120</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M50.5,-664.92C50.5,-657.22 50.5,-647.69 50.5,-639.43\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"54,-639.25 50.5,-629.25 47,-639.25 54,-639.25\"/>\n",
       "</g>\n",
       "<!-- 140169393694336&#45;&gt;140169393693184 -->\n",
       "<g id=\"edge33\" class=\"edge\">\n",
       "<title>140169393694336&#45;&gt;140169393693184</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M254.37,-609.79C250.36,-601.34 244.15,-588.25 239.06,-577.55\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"242.1,-575.78 234.65,-568.24 235.77,-578.78 242.1,-575.78\"/>\n",
       "</g>\n",
       "<!-- 140169393692320 -->\n",
       "<g id=\"node36\" class=\"node\">\n",
       "<title>140169393692320</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"196,-629 119,-629 119,-610 196,-610 196,-629\"/>\n",
       "<text text-anchor=\"middle\" x=\"157.5\" y=\"-617\" font-family=\"monospace\" font-size=\"10.00\">TBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393692320&#45;&gt;140169393693184 -->\n",
       "<g id=\"edge35\" class=\"edge\">\n",
       "<title>140169393692320&#45;&gt;140169393693184</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M168.27,-609.79C179.7,-600.56 197.98,-585.78 211.78,-574.63\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"214.11,-577.25 219.68,-568.24 209.71,-571.81 214.11,-577.25\"/>\n",
       "</g>\n",
       "<!-- 140169393694528 -->\n",
       "<g id=\"node37\" class=\"node\">\n",
       "<title>140169393694528</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"208,-690 107,-690 107,-671 208,-671 208,-690\"/>\n",
       "<text text-anchor=\"middle\" x=\"157.5\" y=\"-678\" font-family=\"monospace\" font-size=\"10.00\">AccumulateGrad</text>\n",
       "</g>\n",
       "<!-- 140169393694528&#45;&gt;140169393692320 -->\n",
       "<g id=\"edge36\" class=\"edge\">\n",
       "<title>140169393694528&#45;&gt;140169393692320</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M157.5,-670.79C157.5,-662.6 157.5,-650.06 157.5,-639.55\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"161,-639.24 157.5,-629.24 154,-639.24 161,-639.24\"/>\n",
       "</g>\n",
       "<!-- 140169429537072 -->\n",
       "<g id=\"node38\" class=\"node\">\n",
       "<title>140169429537072</title>\n",
       "<polygon fill=\"lightblue\" stroke=\"black\" points=\"187,-763 128,-763 128,-732 187,-732 187,-763\"/>\n",
       "<text text-anchor=\"middle\" x=\"157.5\" y=\"-739\" font-family=\"monospace\" font-size=\"10.00\"> (3, 3)</text>\n",
       "</g>\n",
       "<!-- 140169429537072&#45;&gt;140169393694528 -->\n",
       "<g id=\"edge37\" class=\"edge\">\n",
       "<title>140169429537072&#45;&gt;140169393694528</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M157.5,-731.75C157.5,-722.39 157.5,-710.19 157.5,-700.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"161,-700.02 157.5,-690.02 154,-700.02 161,-700.02\"/>\n",
       "</g>\n",
       "<!-- 140169393692512 -->\n",
       "<g id=\"node39\" class=\"node\">\n",
       "<title>140169393692512</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"593,-251 516,-251 516,-232 593,-232 593,-251\"/>\n",
       "<text text-anchor=\"middle\" x=\"554.5\" y=\"-239\" font-family=\"monospace\" font-size=\"10.00\">TBackward0</text>\n",
       "</g>\n",
       "<!-- 140169393692512&#45;&gt;140169393686944 -->\n",
       "<g id=\"edge38\" class=\"edge\">\n",
       "<title>140169393692512&#45;&gt;140169393686944</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M536.67,-231.98C518.89,-223.5 491.35,-210.35 470.82,-200.54\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"472.19,-197.32 461.66,-196.17 469.17,-203.64 472.19,-197.32\"/>\n",
       "</g>\n",
       "<!-- 140169393692944 -->\n",
       "<g id=\"node40\" class=\"node\">\n",
       "<title>140169393692944</title>\n",
       "<polygon fill=\"lightgrey\" stroke=\"black\" points=\"615,-312 514,-312 514,-293 615,-293 615,-312\"/>\n",
       "<text text-anchor=\"middle\" x=\"564.5\" y=\"-300\" font-family=\"monospace\" font-size=\"10.00\">AccumulateGrad</text>\n",
       "</g>\n",
       "<!-- 140169393692944&#45;&gt;140169393692512 -->\n",
       "<g id=\"edge39\" class=\"edge\">\n",
       "<title>140169393692944&#45;&gt;140169393692512</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M563.02,-292.79C561.62,-284.52 559.46,-271.79 557.67,-261.21\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"561.1,-260.52 555.98,-251.24 554.2,-261.69 561.1,-260.52\"/>\n",
       "</g>\n",
       "<!-- 140169429176464 -->\n",
       "<g id=\"node41\" class=\"node\">\n",
       "<title>140169429176464</title>\n",
       "<polygon fill=\"lightblue\" stroke=\"black\" points=\"594,-385 535,-385 535,-354 594,-354 594,-385\"/>\n",
       "<text text-anchor=\"middle\" x=\"564.5\" y=\"-361\" font-family=\"monospace\" font-size=\"10.00\"> (1, 8)</text>\n",
       "</g>\n",
       "<!-- 140169429176464&#45;&gt;140169393692944 -->\n",
       "<g id=\"edge40\" class=\"edge\">\n",
       "<title>140169429176464&#45;&gt;140169393692944</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M564.5,-353.75C564.5,-344.39 564.5,-332.19 564.5,-322.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"568,-322.02 564.5,-312.02 561,-322.02 568,-322.02\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph>"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "torchviz.make_dot(yhat.mean(), params=dict(model.named_parameters()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e7c9ac9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2179883b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d26a6fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "93d03773",
   "metadata": {},
   "source": [
    "## KEGG graph "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4299f2b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pickle as pkl\n",
    "# with open('../data/zma/kegg/kegg_gene_entries.pkl', 'rb') as handle:\n",
    "#     parsed_kegg_gene_entries = pkl.load(handle)\n",
    "\n",
    "# Testing using the cleaned up version in '../nbs_artifacts/01.05_g2fc_demo_model/'\n",
    "load_from = '../nbs_artifacts/01.05_g2fc_demo_model/'\n",
    "parsed_kegg_gene_entries = get_cached_result(load_from+'filtered_kegg_gene_entries.pkl')\n",
    "\n",
    "ACGT_gene_slice_list = get_cached_result(load_from+'ACGT_gene_slice_list.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dc4d1f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(ACGT_gene_slice_list) == len(parsed_kegg_gene_entries)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8edc4ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Restrict to only those with pathway\n",
    "kegg_gene_brite = [e for e in parsed_kegg_gene_entries if 'BRITE' in e.keys()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbb791b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# also require to have a non-empty path\n",
    "kegg_gene_brite = [e for e in kegg_gene_brite if not e['BRITE']['BRITE_PATHS'] == []]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab9f3eed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retaining 43.53%, 6067/13939 Entries\n"
     ]
    }
   ],
   "source": [
    "print('Retaining '+ str(round(len(kegg_gene_brite)/len(parsed_kegg_gene_entries), 4)*100)+'%, '+str(len(kegg_gene_brite)\n",
    "     )+'/'+str(len(parsed_kegg_gene_entries)\n",
    "     )+' Entries'\n",
    "     )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d365781b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['KEGG Orthology (KO) [BR:zma00001]',\n",
       "  '09100 Metabolism',\n",
       "  '09102 Energy metabolism',\n",
       "  '00190 Oxidative phosphorylation',\n",
       "  '100383860'],\n",
       " ['Enzymes [BR:zma01000]',\n",
       "  '7. Translocases',\n",
       "  '7.1  Catalysing the translocation of hydrons',\n",
       "  '7.1.2  Linked to the hydrolysis of a nucleoside triphosphate',\n",
       "  '7.1.2.1  P-type H+-exporting transporter',\n",
       "  '100383860']]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_gene_brite[1]['BRITE']['BRITE_PATHS']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0bea378",
   "metadata": {},
   "source": [
    "### Tiny 2 gene version of the network:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30b38dcb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6067"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(kegg_gene_brite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cc93bad",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00, 30283.78it/s]\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "The goal here is to have a dict with each node and a list of it's children. \n",
    "For example, the graph\n",
    "a--b--d\n",
    " |-c--e\n",
    "Would be parsed into     \n",
    "{'a':['b', 'c'],\n",
    " 'b':['d'],\n",
    " 'c':['e']}\n",
    "\"\"\"\n",
    "kegg_connections = {}\n",
    "\n",
    "# for all genes in list\n",
    "for i in tqdm(range(2)): # <------------------------- only using 2 inputs genes\n",
    "# for i in tqdm(range(len(kegg_gene_brite))):    \n",
    "    temp = kegg_gene_brite[i]['BRITE']['BRITE_PATHS']\n",
    "    # clean up to make sure that there are no \":\" characters. These can mess up graphviz\n",
    "    temp = [[temp[j][i].replace(':', '-') for i in range(len(temp[j])) ] for j in range(len(temp))]\n",
    "    # all paths through graph associated with a gene\n",
    "    for j in range(len(temp)):\n",
    "        # steps of the path through the graph\n",
    "        for k in range(len(temp[j])-1):\n",
    "            # if this is a new key, add it and add the k+1 entry as it's child\n",
    "            if temp[j][k] not in kegg_connections.keys():\n",
    "                kegg_connections[temp[j][k]] = [temp[j][k+1]]\n",
    "            else: \n",
    "                # Check to see if there's a new child to add   \n",
    "                if temp[j][k+1] not in kegg_connections[temp[j][k]]:\n",
    "                    # add it.\n",
    "                    kegg_connections[temp[j][k]].extend([temp[j][k+1]])\n",
    "\n",
    "# kegg_connections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9c8cc67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add yhat node to the graph\n",
    "temp_values = []\n",
    "for key in kegg_connections.keys():\n",
    "    temp_values += kegg_connections[key]\n",
    "\n",
    "kegg_connections['y_hat'] = [key for key in kegg_connections.keys() if key not in temp_values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "597288ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:00<00:00, 58319.02it/s]\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (0)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"2997pt\" height=\"476pt\"\n",
       " viewBox=\"0.00 0.00 2997.48 476.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 472)\">\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-472 2993.48,-472 2993.48,4 -4,4\"/>\n",
       "<!-- KEGG Orthology (KO) [BR&#45;zma00001] -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>KEGG Orthology (KO) [BR&#45;zma00001]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1304.09\" cy=\"-162\" rx=\"186.57\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1304.09\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">KEGG Orthology (KO) [BR&#45;zma00001]</text>\n",
       "</g>\n",
       "<!-- y_hat -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>y_hat</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1897.09\" cy=\"-18\" rx=\"36\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1897.09\" y=\"-14.3\" font-family=\"Times,serif\" font-size=\"14.00\">y_hat</text>\n",
       "</g>\n",
       "<!-- KEGG Orthology (KO) [BR&#45;zma00001]&#45;&gt;y_hat -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>KEGG Orthology (KO) [BR&#45;zma00001]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1344.47,-144.36C1394.17,-124.47 1481.54,-91.44 1559.09,-72 1662.45,-46.08 1786.93,-30.46 1852.17,-23.43\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1852.55,-26.91 1862.12,-22.38 1851.81,-19.95 1852.55,-26.91\"/>\n",
       "</g>\n",
       "<!-- 09180 Brite Hierarchies -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>09180 Brite Hierarchies</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1304.09\" cy=\"-234\" rx=\"123.48\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1304.09\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">09180 Brite Hierarchies</text>\n",
       "</g>\n",
       "<!-- 09180 Brite Hierarchies&#45;&gt;KEGG Orthology (KO) [BR&#45;zma00001] -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>09180 Brite Hierarchies&#45;&gt;KEGG Orthology (KO) [BR&#45;zma00001]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1304.09,-215.7C1304.09,-207.98 1304.09,-198.71 1304.09,-190.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1307.59,-190.1 1304.09,-180.1 1300.59,-190.1 1307.59,-190.1\"/>\n",
       "</g>\n",
       "<!-- 09100 Metabolism -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>09100 Metabolism</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"618.09\" cy=\"-234\" rx=\"97.48\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"618.09\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">09100 Metabolism</text>\n",
       "</g>\n",
       "<!-- 09100 Metabolism&#45;&gt;KEGG Orthology (KO) [BR&#45;zma00001] -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>09100 Metabolism&#45;&gt;KEGG Orthology (KO) [BR&#45;zma00001]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M701.17,-224.52C817.84,-212.62 1031.48,-190.82 1170.59,-176.62\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1171.3,-180.07 1180.9,-175.57 1170.59,-173.1 1171.3,-180.07\"/>\n",
       "</g>\n",
       "<!-- 09182 Protein families&#45; genetic information processing -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>09182 Protein families&#45; genetic information processing</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1304.09\" cy=\"-306\" rx=\"266.45\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1304.09\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">09182 Protein families&#45; genetic information processing</text>\n",
       "</g>\n",
       "<!-- 09182 Protein families&#45; genetic information processing&#45;&gt;09180 Brite Hierarchies -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>09182 Protein families&#45; genetic information processing&#45;&gt;09180 Brite Hierarchies</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1304.09,-287.7C1304.09,-279.98 1304.09,-270.71 1304.09,-262.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1307.59,-262.1 1304.09,-252.1 1300.59,-262.1 1307.59,-262.1\"/>\n",
       "</g>\n",
       "<!-- 03016 Transfer RNA biogenesis [BR&#45;zma03016] -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>03016 Transfer RNA biogenesis [BR&#45;zma03016]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1001.09\" cy=\"-378\" rx=\"233.16\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1001.09\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">03016 Transfer RNA biogenesis [BR&#45;zma03016]</text>\n",
       "</g>\n",
       "<!-- 03016 Transfer RNA biogenesis [BR&#45;zma03016]&#45;&gt;09182 Protein families&#45; genetic information processing -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>03016 Transfer RNA biogenesis [BR&#45;zma03016]&#45;&gt;09182 Protein families&#45; genetic information processing</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1071.36,-360.76C1116.73,-350.28 1175.91,-336.61 1223.23,-325.68\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1224.19,-329.05 1233.14,-323.39 1222.61,-322.23 1224.19,-329.05\"/>\n",
       "</g>\n",
       "<!-- 03036 Chromosome and associated proteins [BR&#45;zma03036] -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>03036 Chromosome and associated proteins [BR&#45;zma03036]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1544.09\" cy=\"-378\" rx=\"291.65\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1544.09\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">03036 Chromosome and associated proteins [BR&#45;zma03036]</text>\n",
       "</g>\n",
       "<!-- 03036 Chromosome and associated proteins [BR&#45;zma03036]&#45;&gt;09182 Protein families&#45; genetic information processing -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>03036 Chromosome and associated proteins [BR&#45;zma03036]&#45;&gt;09182 Protein families&#45; genetic information processing</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1486.9,-360.32C1452.18,-350.19 1407.68,-337.22 1371.28,-326.6\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1371.82,-323.11 1361.24,-323.67 1369.86,-329.83 1371.82,-323.11\"/>\n",
       "</g>\n",
       "<!-- 100278565 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>100278565</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"2112.09\" cy=\"-450\" rx=\"63.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"2112.09\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">100278565</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;03016 Transfer RNA biogenesis [BR&#45;zma03016] -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;03016 Transfer RNA biogenesis [BR&#45;zma03016]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2050.44,-445.44C1908.16,-437.17 1545.86,-415.87 1243.09,-396 1221.3,-394.57 1198.42,-393.02 1175.78,-391.46\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1175.96,-387.97 1165.74,-390.77 1175.48,-394.95 1175.96,-387.97\"/>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;03036 Chromosome and associated proteins [BR&#45;zma03036] -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;03036 Chromosome and associated proteins [BR&#45;zma03036]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2055.02,-441.97C1967.6,-431.19 1798.07,-410.3 1678.31,-395.54\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1678.69,-392.06 1668.33,-394.31 1677.83,-399.01 1678.69,-392.06\"/>\n",
       "</g>\n",
       "<!-- 2.1.1.34 &#160;tRNA (guanosine18&#45;2&#39;&#45;O)&#45;methyltransferase -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>2.1.1.34 &#160;tRNA (guanosine18&#45;2&#39;&#45;O)&#45;methyltransferase</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"2112.09\" cy=\"-378\" rx=\"258.05\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"2112.09\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">2.1.1.34 &#160;tRNA (guanosine18&#45;2&#39;&#45;O)&#45;methyltransferase</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;2.1.1.34 &#160;tRNA (guanosine18&#45;2&#39;&#45;O)&#45;methyltransferase -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;2.1.1.34 &#160;tRNA (guanosine18&#45;2&#39;&#45;O)&#45;methyltransferase</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2112.09,-431.7C2112.09,-423.98 2112.09,-414.71 2112.09,-406.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"2115.59,-406.1 2112.09,-396.1 2108.59,-406.1 2115.59,-406.1\"/>\n",
       "</g>\n",
       "<!-- Methyltransferases -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>Methyltransferases</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"2888.09\" cy=\"-378\" rx=\"101.28\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"2888.09\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">Methyltransferases</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;Methyltransferases -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;Methyltransferases</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2174.12,-445.56C2291.9,-438.7 2555.92,-421.72 2777.09,-396 2786.5,-394.91 2796.34,-393.61 2806.11,-392.24\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"2806.84,-395.67 2816.24,-390.78 2805.84,-388.74 2806.84,-395.67\"/>\n",
       "</g>\n",
       "<!-- RISC (RNA&#45;induced silencing complex) -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>RISC (RNA&#45;induced silencing complex)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"2578.09\" cy=\"-378\" rx=\"190.37\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"2578.09\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">RISC (RNA&#45;induced silencing complex)</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;RISC (RNA&#45;induced silencing complex) -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;RISC (RNA&#45;induced silencing complex)</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2166.91,-440.76C2241.5,-429.56 2376.52,-409.28 2471.44,-395.02\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"2471.99,-398.48 2481.36,-393.53 2470.95,-391.55 2471.99,-398.48\"/>\n",
       "</g>\n",
       "<!-- Enzymes [BR&#45;zma01000] -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>Enzymes [BR&#45;zma01000]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1695.09\" cy=\"-90\" rx=\"127.28\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1695.09\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">Enzymes [BR&#45;zma01000]</text>\n",
       "</g>\n",
       "<!-- Enzymes [BR&#45;zma01000]&#45;&gt;y_hat -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>Enzymes [BR&#45;zma01000]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1740.92,-73.12C1776.63,-60.74 1825.6,-43.77 1859.12,-32.15\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1860.27,-35.46 1868.57,-28.88 1857.98,-28.85 1860.27,-35.46\"/>\n",
       "</g>\n",
       "<!-- 2. Transferases -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>2. Transferases</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1993.09\" cy=\"-162\" rx=\"82.59\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1993.09\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">2. Transferases</text>\n",
       "</g>\n",
       "<!-- 2. Transferases&#45;&gt;Enzymes [BR&#45;zma01000] -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>2. Transferases&#45;&gt;Enzymes [BR&#45;zma01000]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1939.12,-148.32C1891.14,-137.05 1820.6,-120.48 1767.68,-108.05\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1768.21,-104.58 1757.68,-105.7 1766.61,-111.4 1768.21,-104.58\"/>\n",
       "</g>\n",
       "<!-- 7. Translocases -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>7. Translocases</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1685.09\" cy=\"-162\" rx=\"82.59\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1685.09\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">7. Translocases</text>\n",
       "</g>\n",
       "<!-- 7. Translocases&#45;&gt;Enzymes [BR&#45;zma01000] -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>7. Translocases&#45;&gt;Enzymes [BR&#45;zma01000]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1687.56,-143.7C1688.66,-135.98 1689.98,-126.71 1691.21,-118.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1694.69,-118.5 1692.64,-108.1 1687.76,-117.51 1694.69,-118.5\"/>\n",
       "</g>\n",
       "<!-- 2.1 &#160;Transferring one&#45;carbon groups -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>2.1 &#160;Transferring one&#45;carbon groups</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"2108.09\" cy=\"-234\" rx=\"178.97\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"2108.09\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">2.1 &#160;Transferring one&#45;carbon groups</text>\n",
       "</g>\n",
       "<!-- 2.1 &#160;Transferring one&#45;carbon groups&#45;&gt;2. Transferases -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>2.1 &#160;Transferring one&#45;carbon groups&#45;&gt;2. Transferases</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2080.25,-216.05C2064.65,-206.56 2045.06,-194.63 2028.39,-184.49\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"2029.84,-181.28 2019.48,-179.07 2026.2,-187.26 2029.84,-181.28\"/>\n",
       "</g>\n",
       "<!-- 2.1.1 &#160;Methyltransferases -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>2.1.1 &#160;Methyltransferases</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"2111.09\" cy=\"-306\" rx=\"130.78\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"2111.09\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">2.1.1 &#160;Methyltransferases</text>\n",
       "</g>\n",
       "<!-- 2.1.1 &#160;Methyltransferases&#45;&gt;2.1 &#160;Transferring one&#45;carbon groups -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>2.1.1 &#160;Methyltransferases&#45;&gt;2.1 &#160;Transferring one&#45;carbon groups</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2110.34,-287.7C2110.01,-279.98 2109.62,-270.71 2109.25,-262.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"2112.74,-261.95 2108.82,-252.1 2105.75,-262.25 2112.74,-261.95\"/>\n",
       "</g>\n",
       "<!-- 2.1.1.34 &#160;tRNA (guanosine18&#45;2&#39;&#45;O)&#45;methyltransferase&#45;&gt;2.1.1 &#160;Methyltransferases -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>2.1.1.34 &#160;tRNA (guanosine18&#45;2&#39;&#45;O)&#45;methyltransferase&#45;&gt;2.1.1 &#160;Methyltransferases</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2111.84,-359.7C2111.73,-351.98 2111.6,-342.71 2111.47,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"2114.97,-334.05 2111.33,-324.1 2107.97,-334.15 2114.97,-334.05\"/>\n",
       "</g>\n",
       "<!-- Transfer RNA biogenesis [BR&#45;zma03016] -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>Transfer RNA biogenesis [BR&#45;zma03016]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"2578.09\" cy=\"-90\" rx=\"200.66\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"2578.09\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">Transfer RNA biogenesis [BR&#45;zma03016]</text>\n",
       "</g>\n",
       "<!-- Transfer RNA biogenesis [BR&#45;zma03016]&#45;&gt;y_hat -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>Transfer RNA biogenesis [BR&#45;zma03016]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2451.59,-76C2297.08,-60.11 2043.69,-34.07 1941.69,-23.58\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1942.03,-20.1 1931.72,-22.56 1941.31,-27.06 1942.03,-20.1\"/>\n",
       "</g>\n",
       "<!-- Eukaryotic type -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>Eukaryotic type</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"2578.09\" cy=\"-234\" rx=\"84.49\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"2578.09\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">Eukaryotic type</text>\n",
       "</g>\n",
       "<!-- Eukaryotic type&#45;&gt;Transfer RNA biogenesis [BR&#45;zma03016] -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>Eukaryotic type&#45;&gt;Transfer RNA biogenesis [BR&#45;zma03016]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2578.09,-215.87C2578.09,-191.67 2578.09,-147.21 2578.09,-118.39\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"2581.59,-118.19 2578.09,-108.19 2574.59,-118.19 2581.59,-118.19\"/>\n",
       "</g>\n",
       "<!-- Chromosome and associated proteins [BR&#45;zma03036] -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>Chromosome and associated proteins [BR&#45;zma03036]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"2100.09\" cy=\"-90\" rx=\"259.15\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"2100.09\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">Chromosome and associated proteins [BR&#45;zma03036]</text>\n",
       "</g>\n",
       "<!-- Eukaryotic type&#45;&gt;Chromosome and associated proteins [BR&#45;zma03036] -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>Eukaryotic type&#45;&gt;Chromosome and associated proteins [BR&#45;zma03036]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2530.94,-219C2445.21,-193.53 2263.61,-139.58 2165.79,-110.52\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"2166.59,-107.11 2156.01,-107.61 2164.6,-113.82 2166.59,-107.11\"/>\n",
       "</g>\n",
       "<!-- tRNA modification factors -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>tRNA modification factors</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"2847.09\" cy=\"-306\" rx=\"130.78\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"2847.09\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">tRNA modification factors</text>\n",
       "</g>\n",
       "<!-- tRNA modification factors&#45;&gt;Eukaryotic type -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>tRNA modification factors&#45;&gt;Eukaryotic type</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2788.72,-289.81C2744.61,-278.33 2684.23,-262.62 2639.42,-250.96\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"2640.09,-247.52 2629.53,-248.39 2638.33,-254.29 2640.09,-247.52\"/>\n",
       "</g>\n",
       "<!-- Gene silencing -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>Gene silencing</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"2578.09\" cy=\"-306\" rx=\"79.09\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"2578.09\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">Gene silencing</text>\n",
       "</g>\n",
       "<!-- Gene silencing&#45;&gt;Eukaryotic type -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>Gene silencing&#45;&gt;Eukaryotic type</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2578.09,-287.7C2578.09,-279.98 2578.09,-270.71 2578.09,-262.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"2581.59,-262.1 2578.09,-252.1 2574.59,-262.1 2581.59,-262.1\"/>\n",
       "</g>\n",
       "<!-- Methyltransferases&#45;&gt;tRNA modification factors -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>Methyltransferases&#45;&gt;tRNA modification factors</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2878.16,-360.05C2873.4,-351.92 2867.58,-342 2862.28,-332.94\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"2865.28,-331.14 2857.2,-324.28 2859.24,-334.67 2865.28,-331.14\"/>\n",
       "</g>\n",
       "<!-- Chromosome and associated proteins [BR&#45;zma03036]&#45;&gt;y_hat -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>Chromosome and associated proteins [BR&#45;zma03036]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2051.46,-72.23C2015.96,-59.99 1968.48,-43.62 1935.59,-32.28\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1936.48,-28.88 1925.88,-28.93 1934.2,-35.5 1936.48,-28.88\"/>\n",
       "</g>\n",
       "<!-- RISC (RNA&#45;induced silencing complex)&#45;&gt;Gene silencing -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>RISC (RNA&#45;induced silencing complex)&#45;&gt;Gene silencing</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M2578.09,-359.7C2578.09,-351.98 2578.09,-342.71 2578.09,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"2581.59,-334.1 2578.09,-324.1 2574.59,-334.1 2581.59,-334.1\"/>\n",
       "</g>\n",
       "<!-- 09102 Energy metabolism -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>09102 Energy metabolism</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"171.09\" cy=\"-306\" rx=\"133.78\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"171.09\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">09102 Energy metabolism</text>\n",
       "</g>\n",
       "<!-- 09102 Energy metabolism&#45;&gt;09100 Metabolism -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>09102 Energy metabolism&#45;&gt;09100 Metabolism</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M254.64,-291.92C335.21,-279.3 456.39,-260.32 536.35,-247.8\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"537.04,-251.24 546.37,-246.23 535.95,-244.32 537.04,-251.24\"/>\n",
       "</g>\n",
       "<!-- 00190 Oxidative phosphorylation -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>00190 Oxidative phosphorylation</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"165.09\" cy=\"-378\" rx=\"165.17\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"165.09\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">00190 Oxidative phosphorylation</text>\n",
       "</g>\n",
       "<!-- 00190 Oxidative phosphorylation&#45;&gt;09102 Energy metabolism -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>00190 Oxidative phosphorylation&#45;&gt;09102 Energy metabolism</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M166.57,-359.7C167.23,-351.98 168.02,-342.71 168.76,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"172.25,-334.37 169.62,-324.1 165.28,-333.77 172.25,-334.37\"/>\n",
       "</g>\n",
       "<!-- 100383860 -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>100383860</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"357.09\" cy=\"-450\" rx=\"63.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"357.09\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">100383860</text>\n",
       "</g>\n",
       "<!-- 100383860&#45;&gt;00190 Oxidative phosphorylation -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>100383860&#45;&gt;00190 Oxidative phosphorylation</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M319.61,-435.34C291.13,-424.95 251.53,-410.52 219.55,-398.86\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"220.48,-395.47 209.89,-395.33 218.08,-402.05 220.48,-395.47\"/>\n",
       "</g>\n",
       "<!-- 7.1.2.1 &#160;P&#45;type H+&#45;exporting transporter -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>7.1.2.1 &#160;P&#45;type H+&#45;exporting transporter</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"549.09\" cy=\"-378\" rx=\"200.66\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"549.09\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">7.1.2.1 &#160;P&#45;type H+&#45;exporting transporter</text>\n",
       "</g>\n",
       "<!-- 100383860&#45;&gt;7.1.2.1 &#160;P&#45;type H+&#45;exporting transporter -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>100383860&#45;&gt;7.1.2.1 &#160;P&#45;type H+&#45;exporting transporter</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M394.56,-435.34C422.85,-425.02 462.09,-410.72 493.96,-399.1\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"495.39,-402.3 503.59,-395.59 492.99,-395.73 495.39,-402.3\"/>\n",
       "</g>\n",
       "<!-- 7.1 &#160;Catalysing the translocation of hydrons -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>7.1 &#160;Catalysing the translocation of hydrons</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1680.09\" cy=\"-234\" rx=\"214.46\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1680.09\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">7.1 &#160;Catalysing the translocation of hydrons</text>\n",
       "</g>\n",
       "<!-- 7.1 &#160;Catalysing the translocation of hydrons&#45;&gt;7. Translocases -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>7.1 &#160;Catalysing the translocation of hydrons&#45;&gt;7. Translocases</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1681.32,-215.7C1681.87,-207.98 1682.53,-198.71 1683.15,-190.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1686.64,-190.33 1683.86,-180.1 1679.66,-189.83 1686.64,-190.33\"/>\n",
       "</g>\n",
       "<!-- 7.1.2 &#160;Linked to the hydrolysis of a nucleoside triphosphate -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>7.1.2 &#160;Linked to the hydrolysis of a nucleoside triphosphate</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"692.09\" cy=\"-306\" rx=\"286.75\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"692.09\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">7.1.2 &#160;Linked to the hydrolysis of a nucleoside triphosphate</text>\n",
       "</g>\n",
       "<!-- 7.1.2 &#160;Linked to the hydrolysis of a nucleoside triphosphate&#45;&gt;7.1 &#160;Catalysing the translocation of hydrons -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>7.1.2 &#160;Linked to the hydrolysis of a nucleoside triphosphate&#45;&gt;7.1 &#160;Catalysing the translocation of hydrons</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M874.45,-292.08C1056.72,-279.17 1334.67,-259.47 1511.81,-246.92\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1512.28,-250.4 1522.01,-246.2 1511.78,-243.42 1512.28,-250.4\"/>\n",
       "</g>\n",
       "<!-- 7.1.2.1 &#160;P&#45;type H+&#45;exporting transporter&#45;&gt;7.1.2 &#160;Linked to the hydrolysis of a nucleoside triphosphate -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>7.1.2.1 &#160;P&#45;type H+&#45;exporting transporter&#45;&gt;7.1.2 &#160;Linked to the hydrolysis of a nucleoside triphosphate</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M583.34,-360.23C602.91,-350.65 627.66,-338.54 648.6,-328.29\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"650.17,-331.41 657.62,-323.87 647.1,-325.13 650.17,-331.41\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph>"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This is too big to render in full\n",
    "\n",
    "dot = Digraph()\n",
    "for key in tqdm(kegg_connections.keys()):\n",
    "    dot.node(key)\n",
    "    for value in kegg_connections[key]:\n",
    "        # edge takes a head/tail whereas edges takes name pairs concatednated (A, B -> AB)in a list\n",
    "        dot.edge(value, key)    \n",
    "    \n",
    "dot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f1e6077",
   "metadata": {},
   "source": [
    "Version with the node names masked for size "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8ddd86c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:00<00:00, 146041.23it/s]\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (0)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"495pt\" height=\"476pt\"\n",
       " viewBox=\"0.00 0.00 494.69 476.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 472)\">\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-472 490.69,-472 490.69,4 -4,4\"/>\n",
       "<!-- 0 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>0</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"156.69\" cy=\"-162\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"156.69\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">0</text>\n",
       "</g>\n",
       "<!-- 24 -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>24</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"285.69\" cy=\"-18\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"285.69\" y=\"-14.3\" font-family=\"Times,serif\" font-size=\"14.00\">24</text>\n",
       "</g>\n",
       "<!-- 0&#45;&gt;24 -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>0&#45;&gt;24</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M164.8,-144.65C174.66,-125.79 192.62,-94.49 213.69,-72 226.61,-58.21 243.53,-45.55 257.72,-36.07\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"259.66,-38.98 266.14,-30.59 255.85,-33.11 259.66,-38.98\"/>\n",
       "</g>\n",
       "<!-- 1 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>1</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"156.69\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"156.69\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">1</text>\n",
       "</g>\n",
       "<!-- 1&#45;&gt;0 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>1&#45;&gt;0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M156.69,-215.7C156.69,-207.98 156.69,-198.71 156.69,-190.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"160.19,-190.1 156.69,-180.1 153.19,-190.1 160.19,-190.1\"/>\n",
       "</g>\n",
       "<!-- 17 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>17</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"59.69\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"59.69\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">17</text>\n",
       "</g>\n",
       "<!-- 17&#45;&gt;0 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>17&#45;&gt;0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M77.49,-220.16C92.39,-209.41 113.87,-193.91 130.7,-181.76\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"132.95,-184.45 139.02,-175.76 128.86,-178.77 132.95,-184.45\"/>\n",
       "</g>\n",
       "<!-- 2 -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>2</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"171.69\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"171.69\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">2</text>\n",
       "</g>\n",
       "<!-- 2&#45;&gt;1 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>2&#45;&gt;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M168.06,-288.05C166.39,-280.26 164.37,-270.82 162.5,-262.08\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"165.91,-261.32 160.4,-252.28 159.07,-262.79 165.91,-261.32\"/>\n",
       "</g>\n",
       "<!-- 3 -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>3</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"171.69\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"171.69\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">3</text>\n",
       "</g>\n",
       "<!-- 3&#45;&gt;2 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>3&#45;&gt;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M171.69,-359.7C171.69,-351.98 171.69,-342.71 171.69,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"175.19,-334.1 171.69,-324.1 168.19,-334.1 175.19,-334.1\"/>\n",
       "</g>\n",
       "<!-- 4 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>4</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"243.69\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"243.69\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">4</text>\n",
       "</g>\n",
       "<!-- 4&#45;&gt;2 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>4&#45;&gt;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M229.12,-362.83C218.94,-352.94 205.17,-339.55 193.66,-328.36\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"196.1,-325.85 186.49,-321.38 191.22,-330.87 196.1,-325.85\"/>\n",
       "</g>\n",
       "<!-- 100278565 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>100278565</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"315.69\" cy=\"-450\" rx=\"63.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"315.69\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">100278565</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;3 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M285.16,-434.15C260.67,-422.25 226.47,-405.63 202.03,-393.75\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"203.39,-390.52 192.86,-389.29 200.33,-396.81 203.39,-390.52\"/>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;4 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;4</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M298.63,-432.41C288.86,-422.91 276.49,-410.88 265.96,-400.65\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"268.29,-398.03 258.68,-393.57 263.41,-403.05 268.29,-398.03\"/>\n",
       "</g>\n",
       "<!-- 9 -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>9</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"315.69\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"315.69\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">9</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;9 -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;9</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M315.69,-431.7C315.69,-423.98 315.69,-414.71 315.69,-406.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"319.19,-406.1 315.69,-396.1 312.19,-406.1 319.19,-406.1\"/>\n",
       "</g>\n",
       "<!-- 13 -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>13</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"459.69\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"459.69\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">13</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;13 -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;13</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M346.23,-434.15C370.72,-422.25 404.92,-405.63 429.35,-393.75\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"431.06,-396.81 438.52,-389.29 428,-390.52 431.06,-396.81\"/>\n",
       "</g>\n",
       "<!-- 16 -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>16</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"387.69\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"387.69\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">16</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;16 -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;16</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M332.76,-432.41C342.53,-422.91 354.9,-410.88 365.43,-400.65\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"367.98,-403.05 372.71,-393.57 363.1,-398.03 367.98,-403.05\"/>\n",
       "</g>\n",
       "<!-- 5 -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>5</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"249.69\" cy=\"-90\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"249.69\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">5</text>\n",
       "</g>\n",
       "<!-- 5&#45;&gt;24 -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>5&#45;&gt;24</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M258.04,-72.76C262.41,-64.28 267.84,-53.71 272.74,-44.2\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"275.93,-45.64 277.39,-35.15 269.7,-42.44 275.93,-45.64\"/>\n",
       "</g>\n",
       "<!-- 6 -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>6</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"300.69\" cy=\"-162\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"300.69\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">6</text>\n",
       "</g>\n",
       "<!-- 6&#45;&gt;5 -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>6&#45;&gt;5</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M289.38,-145.46C282.77,-136.4 274.31,-124.79 266.9,-114.61\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"269.7,-112.51 260.98,-106.49 264.04,-116.64 269.7,-112.51\"/>\n",
       "</g>\n",
       "<!-- 20 -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>20</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"228.69\" cy=\"-162\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"228.69\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">20</text>\n",
       "</g>\n",
       "<!-- 20&#45;&gt;5 -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>20&#45;&gt;5</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M233.78,-144.05C236.18,-136.06 239.1,-126.33 241.77,-117.4\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"245.14,-118.37 244.66,-107.79 238.43,-116.36 245.14,-118.37\"/>\n",
       "</g>\n",
       "<!-- 7 -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>7</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"303.69\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"303.69\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">7</text>\n",
       "</g>\n",
       "<!-- 7&#45;&gt;6 -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>7&#45;&gt;6</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M302.95,-215.7C302.62,-207.98 302.22,-198.71 301.86,-190.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"305.35,-189.95 301.43,-180.1 298.36,-190.25 305.35,-189.95\"/>\n",
       "</g>\n",
       "<!-- 8 -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>8</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"313.69\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"313.69\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">8</text>\n",
       "</g>\n",
       "<!-- 8&#45;&gt;7 -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>8&#45;&gt;7</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M311.27,-288.05C310.17,-280.35 308.84,-271.03 307.6,-262.36\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"311.04,-261.68 306.16,-252.28 304.11,-262.67 311.04,-261.68\"/>\n",
       "</g>\n",
       "<!-- 9&#45;&gt;8 -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>9&#45;&gt;8</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M315.2,-359.7C314.98,-351.98 314.71,-342.71 314.47,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"317.97,-334 314.18,-324.1 310.97,-334.2 317.97,-334\"/>\n",
       "</g>\n",
       "<!-- 10 -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>10</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"393.69\" cy=\"-90\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"393.69\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">10</text>\n",
       "</g>\n",
       "<!-- 10&#45;&gt;24 -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>10&#45;&gt;24</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M374.88,-76.81C357.69,-65.67 332.08,-49.06 312.7,-36.5\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"314.59,-33.56 304.29,-31.05 310.78,-39.43 314.59,-33.56\"/>\n",
       "</g>\n",
       "<!-- 11 -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>11</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"392.69\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"392.69\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">11</text>\n",
       "</g>\n",
       "<!-- 11&#45;&gt;10 -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>11&#45;&gt;10</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M392.81,-215.87C392.99,-191.67 393.3,-147.21 393.5,-118.39\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"397,-118.21 393.57,-108.19 390,-118.16 397,-118.21\"/>\n",
       "</g>\n",
       "<!-- 14 -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>14</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"321.69\" cy=\"-90\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"321.69\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">14</text>\n",
       "</g>\n",
       "<!-- 11&#45;&gt;14 -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>11&#45;&gt;14</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M384.55,-216.71C372.31,-192.23 349.06,-145.74 334.53,-116.67\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"337.57,-114.92 329.96,-107.54 331.3,-118.05 337.57,-114.92\"/>\n",
       "</g>\n",
       "<!-- 12 -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>12</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"459.69\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"459.69\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">12</text>\n",
       "</g>\n",
       "<!-- 12&#45;&gt;11 -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>12&#45;&gt;11</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M445.81,-290.5C436.56,-280.83 424.22,-267.94 413.77,-257.02\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"416,-254.29 406.56,-249.49 410.95,-259.13 416,-254.29\"/>\n",
       "</g>\n",
       "<!-- 15 -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>15</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"387.69\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"387.69\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">15</text>\n",
       "</g>\n",
       "<!-- 15&#45;&gt;11 -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>15&#45;&gt;11</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M388.93,-287.7C389.48,-279.98 390.14,-270.71 390.76,-262.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"394.25,-262.33 391.47,-252.1 387.27,-261.83 394.25,-262.33\"/>\n",
       "</g>\n",
       "<!-- 13&#45;&gt;12 -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>13&#45;&gt;12</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M459.69,-359.7C459.69,-351.98 459.69,-342.71 459.69,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"463.19,-334.1 459.69,-324.1 456.19,-334.1 463.19,-334.1\"/>\n",
       "</g>\n",
       "<!-- 14&#45;&gt;24 -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>14&#45;&gt;24</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M313.34,-72.76C308.98,-64.28 303.55,-53.71 298.65,-44.2\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"301.68,-42.44 294,-35.15 295.46,-45.64 301.68,-42.44\"/>\n",
       "</g>\n",
       "<!-- 16&#45;&gt;15 -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>16&#45;&gt;15</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M387.69,-359.7C387.69,-351.98 387.69,-342.71 387.69,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"391.19,-334.1 387.69,-324.1 384.19,-334.1 391.19,-334.1\"/>\n",
       "</g>\n",
       "<!-- 18 -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>18</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"27.69\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"27.69\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">18</text>\n",
       "</g>\n",
       "<!-- 18&#45;&gt;17 -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>18&#45;&gt;17</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M35.28,-288.41C39.06,-280.13 43.73,-269.92 47.96,-260.66\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"51.19,-262.02 52.17,-251.47 44.82,-259.11 51.19,-262.02\"/>\n",
       "</g>\n",
       "<!-- 19 -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>19</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"27.69\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"27.69\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">19</text>\n",
       "</g>\n",
       "<!-- 19&#45;&gt;18 -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>19&#45;&gt;18</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M27.69,-359.7C27.69,-351.98 27.69,-342.71 27.69,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"31.19,-334.1 27.69,-324.1 24.19,-334.1 31.19,-334.1\"/>\n",
       "</g>\n",
       "<!-- 100383860 -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>100383860</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"63.69\" cy=\"-450\" rx=\"63.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"63.69\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">100383860</text>\n",
       "</g>\n",
       "<!-- 100383860&#45;&gt;19 -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>100383860&#45;&gt;19</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M54.98,-432.05C50.72,-423.77 45.5,-413.62 40.77,-404.42\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"43.77,-402.6 36.08,-395.31 37.54,-405.8 43.77,-402.6\"/>\n",
       "</g>\n",
       "<!-- 23 -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>23</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"99.69\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"99.69\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">23</text>\n",
       "</g>\n",
       "<!-- 100383860&#45;&gt;23 -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>100383860&#45;&gt;23</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M72.41,-432.05C76.67,-423.77 81.89,-413.62 86.62,-404.42\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"89.85,-405.8 91.31,-395.31 83.62,-402.6 89.85,-405.8\"/>\n",
       "</g>\n",
       "<!-- 21 -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>21</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"228.69\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"228.69\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">21</text>\n",
       "</g>\n",
       "<!-- 21&#45;&gt;20 -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>21&#45;&gt;20</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M228.69,-215.7C228.69,-207.98 228.69,-198.71 228.69,-190.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"232.19,-190.1 228.69,-180.1 225.19,-190.1 232.19,-190.1\"/>\n",
       "</g>\n",
       "<!-- 22 -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>22</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"99.69\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"99.69\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">22</text>\n",
       "</g>\n",
       "<!-- 22&#45;&gt;21 -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>22&#45;&gt;21</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M120.13,-293.91C141.46,-282.34 175.09,-264.09 199.27,-250.97\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"201.23,-253.89 208.35,-246.04 197.89,-247.73 201.23,-253.89\"/>\n",
       "</g>\n",
       "<!-- 23&#45;&gt;22 -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>23&#45;&gt;22</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M99.69,-359.7C99.69,-351.98 99.69,-342.71 99.69,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"103.19,-334.1 99.69,-324.1 96.19,-334.1 103.19,-334.1\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph>"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "name_to_num_dict = dict(zip(list(kegg_connections.keys()),\n",
    "                            [str(i) for i in range(len(list(kegg_connections.keys())))]))\n",
    "\n",
    "temp = {}\n",
    "for key in kegg_connections.keys():\n",
    "    temp[name_to_num_dict[key]] = [name_to_num_dict[e] if e in name_to_num_dict.keys() else e for e in kegg_connections[key]]\n",
    "\n",
    "dot = Digraph()\n",
    "for key in tqdm(temp.keys()):\n",
    "    dot.node(key)\n",
    "    for value in temp[key]:\n",
    "        # edge takes a head/tail whereas edges takes name pairs concatednated (A, B -> AB)in a list\n",
    "        dot.edge(value, key)    \n",
    "\n",
    "# dot.render(directory=cache_path, view=True) \n",
    "dot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1cb7324",
   "metadata": {},
   "source": [
    "### Setup the DataLoader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d96f8dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['y_hat'], ['100278565', '100383860'])"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# start by finding the top level -- all those keys which are theselves not values\n",
    "res = find_uniq_keys_values(input_dict = kegg_connections)\n",
    "all_keys = res['all_keys']\n",
    "all_values = res['all_values']\n",
    "\n",
    "# use the keys to find the input/outputs of the graph\n",
    "output_nodes = [e for e in all_keys if e not in all_values]\n",
    "input_nodes = [e for e in all_values if e not in all_keys]\n",
    "\n",
    "(output_nodes, input_nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adc6c050",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find the dependancies for run order from many dependancies to none\n",
    "temp = kegg_connections.copy()\n",
    "\n",
    "no_dependants = find_input_nodes(all_key_value_dict = find_uniq_keys_values(input_dict = temp))\n",
    "# first pass. Same as the output nodes identified above\n",
    "dependancy_order = []\n",
    "# Then iterate\n",
    "for ith in range(100): #TODO <- this should be set as a input parameter\n",
    "    top_nodes = find_top_nodes(all_key_value_dict = find_uniq_keys_values(input_dict = temp))\n",
    "    if top_nodes == []:\n",
    "        break\n",
    "    else:\n",
    "        dependancy_order += top_nodes    \n",
    "        # remove nodes from the graph that are at the 'top' level and haven't already been removed\n",
    "        for key in [e for e in dependancy_order if e in temp.keys()]:\n",
    "             temp.pop(key)\n",
    "\n",
    "# dependancy_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8251172",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['7.1.2.1  P-type H+-exporting transporter',\n",
       " \"2.1.1.34  tRNA (guanosine18-2'-O)-methyltransferase\",\n",
       " '7.1.2  Linked to the hydrolysis of a nucleoside triphosphate',\n",
       " '00190 Oxidative phosphorylation',\n",
       " 'RISC (RNA-induced silencing complex)',\n",
       " 'Methyltransferases',\n",
       " '2.1.1  Methyltransferases',\n",
       " '03036 Chromosome and associated proteins [BR-zma03036]',\n",
       " '03016 Transfer RNA biogenesis [BR-zma03016]',\n",
       " '7.1  Catalysing the translocation of hydrons',\n",
       " '09102 Energy metabolism',\n",
       " 'Gene silencing',\n",
       " 'tRNA modification factors',\n",
       " '2.1  Transferring one-carbon groups',\n",
       " '09182 Protein families- genetic information processing',\n",
       " '7. Translocases',\n",
       " '09100 Metabolism',\n",
       " 'Eukaryotic type',\n",
       " '2. Transferases',\n",
       " '09180 Brite Hierarchies',\n",
       " 'Chromosome and associated proteins [BR-zma03036]',\n",
       " 'Transfer RNA biogenesis [BR-zma03016]',\n",
       " 'Enzymes [BR-zma01000]',\n",
       " 'KEGG Orthology (KO) [BR-zma00001]',\n",
       " 'y_hat']"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reverse to get the order that the nodes should be called\n",
    "dependancy_order.reverse()\n",
    "dependancy_order"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8001e949",
   "metadata": {},
   "source": [
    "-### Set up dict of input tensors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53901bec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6067/6067 [00:00<00:00, 1465409.87it/s]\n"
     ]
    }
   ],
   "source": [
    "# build a dict to go from the node names in `no_dependants` to the list index in `ACGT_gene_slice_list`\n",
    "brite_node_to_list_idx_dict = {}\n",
    "for i in tqdm(range(len(kegg_gene_brite))):\n",
    "    brite_node_to_list_idx_dict[str(kegg_gene_brite[i]['BRITE']['BRITE_PATHS'][0][-1])] = i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e209df06",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_tensor_dict = {}\n",
    "for e in no_dependants:\n",
    "    input_tensor_dict[e] = ACGT_gene_slice_list[brite_node_to_list_idx_dict[e]]\n",
    "    \n",
    "# input_tensor_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f7b3a3a",
   "metadata": {},
   "source": [
    "-### Figure out expected input/output shapes\n",
    "\n",
    "_==NOTE! This assumes only dense connections!==_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ca643b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'7.1.2.1  P-type H+-exporting transporter': 5,\n",
       " \"2.1.1.34  tRNA (guanosine18-2'-O)-methyltransferase\": 5,\n",
       " '7.1.2  Linked to the hydrolysis of a nucleoside triphosphate': 5,\n",
       " '00190 Oxidative phosphorylation': 5,\n",
       " 'RISC (RNA-induced silencing complex)': 5,\n",
       " 'Methyltransferases': 5,\n",
       " '2.1.1  Methyltransferases': 5,\n",
       " '03036 Chromosome and associated proteins [BR-zma03036]': 5,\n",
       " '03016 Transfer RNA biogenesis [BR-zma03016]': 5,\n",
       " '7.1  Catalysing the translocation of hydrons': 5,\n",
       " '09102 Energy metabolism': 5,\n",
       " 'Gene silencing': 5,\n",
       " 'tRNA modification factors': 5,\n",
       " '2.1  Transferring one-carbon groups': 5,\n",
       " '09182 Protein families- genetic information processing': 5,\n",
       " '7. Translocases': 5,\n",
       " '09100 Metabolism': 5,\n",
       " 'Eukaryotic type': 5,\n",
       " '2. Transferases': 5,\n",
       " '09180 Brite Hierarchies': 5,\n",
       " 'Chromosome and associated proteins [BR-zma03036]': 5,\n",
       " 'Transfer RNA biogenesis [BR-zma03016]': 5,\n",
       " 'Enzymes [BR-zma01000]': 5,\n",
       " 'KEGG Orthology (KO) [BR-zma00001]': 5,\n",
       " 'y_hat': 1}"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This could be replaced by a sort of \"distance from output\" measure\n",
    "default_output_size = 5\n",
    "output_size_dict = dict(zip(dependancy_order, \n",
    "                        [default_output_size for i in range(len(dependancy_order))]))\n",
    "output_size_dict['y_hat'] = 1 \n",
    "output_size_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4703e891",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:00<00:00, 543303.63it/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:00<00:00, 168310.75it/s]\n"
     ]
    }
   ],
   "source": [
    "input_size_dict = kegg_connections.copy()\n",
    "\n",
    "# use the expected output sizes from `output_size_dict` to fill in the non-data sizes\n",
    "for e in tqdm(input_size_dict.keys()):\n",
    "    # overwrite named connections with the output size of those connections\n",
    "    # if the entry is in no_dependants it's data so it's size needs to be grabbed from the input_tensor_dict\n",
    "    input_size_dict[e] = [\n",
    "        list(input_tensor_dict[ee].shape)[-1] # <- NOTE! THIS ASSUMES ONLY DENSE CONNECTIONS (i.e. only the 1st dim is needed)\n",
    "        if ee in no_dependants\n",
    "        else output_size_dict[ee] for ee in input_size_dict[e]]\n",
    "\n",
    "# Now walk over entries and overwrite with the su of the inputs\n",
    "for e in tqdm(input_size_dict.keys()):\n",
    "    input_size_dict[e] = np.sum(input_size_dict[e])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07d9fb35",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 25/25 [00:00<00:00, 48011.72it/s]\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (0)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"690pt\" height=\"582pt\"\n",
       " viewBox=\"0.00 0.00 690.02 582.44\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 578.44)\">\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-578.44 686.02,-578.44 686.02,4 -4,4\"/>\n",
       "<!-- KEGG Orthology (KO) [BR&#45;zma00001] -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>KEGG Orthology (KO) [BR&#45;zma00001]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"236.01\" cy=\"-206.35\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"236.01\" y=\"-210.15\" font-family=\"Times,serif\" font-size=\"14.00\">in: 10</text>\n",
       "<text text-anchor=\"middle\" x=\"236.01\" y=\"-195.15\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- y_hat -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>y_hat</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"393.01\" cy=\"-26.87\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"393.01\" y=\"-30.67\" font-family=\"Times,serif\" font-size=\"14.00\">in: 20</text>\n",
       "<text text-anchor=\"middle\" x=\"393.01\" y=\"-15.67\" font-family=\"Times,serif\" font-size=\"14.00\">out: 1</text>\n",
       "</g>\n",
       "<!-- KEGG Orthology (KO) [BR&#45;zma00001]&#45;&gt;y_hat -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>KEGG Orthology (KO) [BR&#45;zma00001]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M243.89,-179.95C252.69,-154.78 268.94,-116.24 293.01,-89.74 309.01,-72.13 331.17,-57.73 350.56,-47.27\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"352.28,-50.32 359.53,-42.6 349.04,-44.11 352.28,-50.32\"/>\n",
       "</g>\n",
       "<!-- 09180 Brite Hierarchies -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>09180 Brite Hierarchies</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"236.01\" cy=\"-296.09\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"236.01\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 5</text>\n",
       "<text text-anchor=\"middle\" x=\"236.01\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 09180 Brite Hierarchies&#45;&gt;KEGG Orthology (KO) [BR&#45;zma00001] -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>09180 Brite Hierarchies&#45;&gt;KEGG Orthology (KO) [BR&#45;zma00001]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M236.01,-268.77C236.01,-260.8 236.01,-251.91 236.01,-243.42\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"239.51,-243.35 236.01,-233.35 232.51,-243.35 239.51,-243.35\"/>\n",
       "</g>\n",
       "<!-- 09100 Metabolism -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>09100 Metabolism</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"90.01\" cy=\"-296.09\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"90.01\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 5</text>\n",
       "<text text-anchor=\"middle\" x=\"90.01\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 09100 Metabolism&#45;&gt;KEGG Orthology (KO) [BR&#45;zma00001] -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>09100 Metabolism&#45;&gt;KEGG Orthology (KO) [BR&#45;zma00001]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M119.56,-277.33C142.12,-263.78 173.38,-244.99 197.84,-230.29\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"199.74,-233.23 206.51,-225.08 196.14,-227.23 199.74,-233.23\"/>\n",
       "</g>\n",
       "<!-- 09182 Protein families&#45; genetic information processing -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>09182 Protein families&#45; genetic information processing</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"241.01\" cy=\"-385.83\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"241.01\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 10</text>\n",
       "<text text-anchor=\"middle\" x=\"241.01\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 09182 Protein families&#45; genetic information processing&#45;&gt;09180 Brite Hierarchies -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>09182 Protein families&#45; genetic information processing&#45;&gt;09180 Brite Hierarchies</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M239.51,-358.51C239.06,-350.54 238.55,-341.65 238.07,-333.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"241.56,-332.88 237.49,-323.09 234.57,-333.28 241.56,-332.88\"/>\n",
       "</g>\n",
       "<!-- 03016 Transfer RNA biogenesis [BR&#45;zma03016] -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>03016 Transfer RNA biogenesis [BR&#45;zma03016]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"241.01\" cy=\"-475.57\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"241.01\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 11</text>\n",
       "<text text-anchor=\"middle\" x=\"241.01\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 03016 Transfer RNA biogenesis [BR&#45;zma03016]&#45;&gt;09182 Protein families&#45; genetic information processing -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>03016 Transfer RNA biogenesis [BR&#45;zma03016]&#45;&gt;09182 Protein families&#45; genetic information processing</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M241.01,-448.25C241.01,-440.28 241.01,-431.39 241.01,-422.9\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"244.51,-422.83 241.01,-412.83 237.51,-422.83 244.51,-422.83\"/>\n",
       "</g>\n",
       "<!-- 03036 Chromosome and associated proteins [BR&#45;zma03036] -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>03036 Chromosome and associated proteins [BR&#45;zma03036]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"341.01\" cy=\"-475.57\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"341.01\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 11</text>\n",
       "<text text-anchor=\"middle\" x=\"341.01\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 03036 Chromosome and associated proteins [BR&#45;zma03036]&#45;&gt;09182 Protein families&#45; genetic information processing -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>03036 Chromosome and associated proteins [BR&#45;zma03036]&#45;&gt;09182 Protein families&#45; genetic information processing</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M317.06,-453.55C303.8,-441.92 287.15,-427.31 272.81,-414.73\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"274.94,-411.94 265.12,-407.98 270.32,-417.21 274.94,-411.94\"/>\n",
       "</g>\n",
       "<!-- 100278565 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>100278565</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"441.01\" cy=\"-556.44\" rx=\"63.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"441.01\" y=\"-552.74\" font-family=\"Times,serif\" font-size=\"14.00\">100278565</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;03016 Transfer RNA biogenesis [BR&#45;zma03016] -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;03016 Transfer RNA biogenesis [BR&#45;zma03016]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M400,-542.65C369.52,-532.73 327.13,-518.08 291.01,-502.44 287.77,-501.04 284.46,-499.52 281.14,-497.94\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"282.42,-494.67 271.9,-493.39 279.33,-500.95 282.42,-494.67\"/>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;03036 Chromosome and associated proteins [BR&#45;zma03036] -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;03036 Chromosome and associated proteins [BR&#45;zma03036]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M420.29,-539.1C407.16,-528.74 389.87,-515.11 374.79,-503.21\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"376.7,-500.26 366.68,-496.82 372.37,-505.76 376.7,-500.26\"/>\n",
       "</g>\n",
       "<!-- 2.1.1.34 &#160;tRNA (guanosine18&#45;2&#39;&#45;O)&#45;methyltransferase -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>2.1.1.34 &#160;tRNA (guanosine18&#45;2&#39;&#45;O)&#45;methyltransferase</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"441.01\" cy=\"-475.57\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"441.01\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 11</text>\n",
       "<text text-anchor=\"middle\" x=\"441.01\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;2.1.1.34 &#160;tRNA (guanosine18&#45;2&#39;&#45;O)&#45;methyltransferase -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;2.1.1.34 &#160;tRNA (guanosine18&#45;2&#39;&#45;O)&#45;methyltransferase</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M441.01,-538.33C441.01,-530.84 441.01,-521.75 441.01,-512.87\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"444.51,-512.65 441.01,-502.65 437.51,-512.65 444.51,-512.65\"/>\n",
       "</g>\n",
       "<!-- Methyltransferases -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>Methyltransferases</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"641.01\" cy=\"-475.57\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"641.01\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 11</text>\n",
       "<text text-anchor=\"middle\" x=\"641.01\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;Methyltransferases -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;Methyltransferases</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M482.02,-542.65C512.51,-532.73 554.89,-518.08 591.01,-502.44 594.25,-501.04 597.57,-499.52 600.88,-497.94\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"602.7,-500.95 610.13,-493.39 599.61,-494.67 602.7,-500.95\"/>\n",
       "</g>\n",
       "<!-- RISC (RNA&#45;induced silencing complex) -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>RISC (RNA&#45;induced silencing complex)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"541.01\" cy=\"-475.57\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"541.01\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 11</text>\n",
       "<text text-anchor=\"middle\" x=\"541.01\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;RISC (RNA&#45;induced silencing complex) -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;RISC (RNA&#45;induced silencing complex)</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M461.73,-539.1C474.86,-528.74 492.15,-515.11 507.23,-503.21\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"509.66,-505.76 515.34,-496.82 505.32,-500.26 509.66,-505.76\"/>\n",
       "</g>\n",
       "<!-- Enzymes [BR&#45;zma01000] -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>Enzymes [BR&#45;zma01000]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"343.01\" cy=\"-116.61\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"343.01\" y=\"-120.41\" font-family=\"Times,serif\" font-size=\"14.00\">in: 10</text>\n",
       "<text text-anchor=\"middle\" x=\"343.01\" y=\"-105.41\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- Enzymes [BR&#45;zma01000]&#45;&gt;y_hat -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>Enzymes [BR&#45;zma01000]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M356.94,-91.17C362.25,-81.85 368.39,-71.07 374.09,-61.07\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"377.25,-62.6 379.16,-52.18 371.17,-59.13 377.25,-62.6\"/>\n",
       "</g>\n",
       "<!-- 2. Transferases -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>2. Transferases</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"436.01\" cy=\"-206.35\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"436.01\" y=\"-210.15\" font-family=\"Times,serif\" font-size=\"14.00\">in: 5</text>\n",
       "<text text-anchor=\"middle\" x=\"436.01\" y=\"-195.15\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 2. Transferases&#45;&gt;Enzymes [BR&#45;zma01000] -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>2. Transferases&#45;&gt;Enzymes [BR&#45;zma01000]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M413.02,-183.66C400.97,-172.29 386.06,-158.23 373.1,-146\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"375.41,-143.37 365.74,-139.05 370.61,-148.46 375.41,-143.37\"/>\n",
       "</g>\n",
       "<!-- 7. Translocases -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>7. Translocases</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"336.01\" cy=\"-206.35\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"336.01\" y=\"-210.15\" font-family=\"Times,serif\" font-size=\"14.00\">in: 5</text>\n",
       "<text text-anchor=\"middle\" x=\"336.01\" y=\"-195.15\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 7. Translocases&#45;&gt;Enzymes [BR&#45;zma01000] -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>7. Translocases&#45;&gt;Enzymes [BR&#45;zma01000]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M338.07,-179.5C338.72,-171.45 339.44,-162.42 340.13,-153.8\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"343.63,-153.83 340.94,-143.58 336.66,-153.27 343.63,-153.83\"/>\n",
       "</g>\n",
       "<!-- 2.1 &#160;Transferring one&#45;carbon groups -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>2.1 &#160;Transferring one&#45;carbon groups</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"437.01\" cy=\"-296.09\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"437.01\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 5</text>\n",
       "<text text-anchor=\"middle\" x=\"437.01\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 2.1 &#160;Transferring one&#45;carbon groups&#45;&gt;2. Transferases -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>2.1 &#160;Transferring one&#45;carbon groups&#45;&gt;2. Transferases</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M436.71,-268.77C436.62,-260.8 436.52,-251.91 436.42,-243.42\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"439.92,-243.31 436.31,-233.35 432.92,-243.39 439.92,-243.31\"/>\n",
       "</g>\n",
       "<!-- 2.1.1 &#160;Methyltransferases -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>2.1.1 &#160;Methyltransferases</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"440.01\" cy=\"-385.83\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"440.01\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 5</text>\n",
       "<text text-anchor=\"middle\" x=\"440.01\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 2.1.1 &#160;Methyltransferases&#45;&gt;2.1 &#160;Transferring one&#45;carbon groups -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>2.1.1 &#160;Methyltransferases&#45;&gt;2.1 &#160;Transferring one&#45;carbon groups</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M439.11,-358.51C438.84,-350.54 438.54,-341.65 438.25,-333.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"441.74,-332.97 437.9,-323.09 434.75,-333.21 441.74,-332.97\"/>\n",
       "</g>\n",
       "<!-- 2.1.1.34 &#160;tRNA (guanosine18&#45;2&#39;&#45;O)&#45;methyltransferase&#45;&gt;2.1.1 &#160;Methyltransferases -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>2.1.1.34 &#160;tRNA (guanosine18&#45;2&#39;&#45;O)&#45;methyltransferase&#45;&gt;2.1.1 &#160;Methyltransferases</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M440.71,-448.25C440.62,-440.28 440.52,-431.39 440.42,-422.9\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"443.92,-422.79 440.31,-412.83 436.92,-422.87 443.92,-422.79\"/>\n",
       "</g>\n",
       "<!-- Transfer RNA biogenesis [BR&#45;zma03016] -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>Transfer RNA biogenesis [BR&#45;zma03016]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"543.01\" cy=\"-116.61\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"543.01\" y=\"-120.41\" font-family=\"Times,serif\" font-size=\"14.00\">in: 5</text>\n",
       "<text text-anchor=\"middle\" x=\"543.01\" y=\"-105.41\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- Transfer RNA biogenesis [BR&#45;zma03016]&#45;&gt;y_hat -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>Transfer RNA biogenesis [BR&#45;zma03016]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M513.02,-98.07C489.62,-84.38 456.94,-65.26 431.59,-50.43\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"433.35,-47.41 422.95,-45.38 429.82,-53.45 433.35,-47.41\"/>\n",
       "</g>\n",
       "<!-- Eukaryotic type -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>Eukaryotic type</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"543.01\" cy=\"-296.09\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"543.01\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 10</text>\n",
       "<text text-anchor=\"middle\" x=\"543.01\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- Eukaryotic type&#45;&gt;Transfer RNA biogenesis [BR&#45;zma03016] -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>Eukaryotic type&#45;&gt;Transfer RNA biogenesis [BR&#45;zma03016]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M543.01,-268.9C543.01,-238.41 543.01,-187.85 543.01,-153.66\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"546.51,-153.62 543.01,-143.62 539.51,-153.62 546.51,-153.62\"/>\n",
       "</g>\n",
       "<!-- Chromosome and associated proteins [BR&#45;zma03036] -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>Chromosome and associated proteins [BR&#45;zma03036]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"443.01\" cy=\"-116.61\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"443.01\" y=\"-120.41\" font-family=\"Times,serif\" font-size=\"14.00\">in: 5</text>\n",
       "<text text-anchor=\"middle\" x=\"443.01\" y=\"-105.41\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- Eukaryotic type&#45;&gt;Chromosome and associated proteins [BR&#45;zma03036] -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>Eukaryotic type&#45;&gt;Chromosome and associated proteins [BR&#45;zma03036]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M532.04,-270.17C521.25,-246.37 503.86,-209.75 486.01,-179.48 480.11,-169.47 473.13,-158.96 466.53,-149.5\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"469.25,-147.29 460.61,-141.16 463.54,-151.34 469.25,-147.29\"/>\n",
       "</g>\n",
       "<!-- tRNA modification factors -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>tRNA modification factors</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"641.01\" cy=\"-385.83\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"641.01\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 5</text>\n",
       "<text text-anchor=\"middle\" x=\"641.01\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- tRNA modification factors&#45;&gt;Eukaryotic type -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>tRNA modification factors&#45;&gt;Eukaryotic type</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M617.29,-363.59C604.29,-351.95 588.02,-337.39 574.02,-324.85\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"576.3,-322.2 566.52,-318.13 571.63,-327.41 576.3,-322.2\"/>\n",
       "</g>\n",
       "<!-- Gene silencing -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>Gene silencing</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"541.01\" cy=\"-385.83\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"541.01\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 5</text>\n",
       "<text text-anchor=\"middle\" x=\"541.01\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- Gene silencing&#45;&gt;Eukaryotic type -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>Gene silencing&#45;&gt;Eukaryotic type</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M541.61,-358.51C541.79,-350.54 542,-341.65 542.19,-333.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"545.69,-333.17 542.42,-323.09 538.69,-333.01 545.69,-333.17\"/>\n",
       "</g>\n",
       "<!-- Methyltransferases&#45;&gt;tRNA modification factors -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>Methyltransferases&#45;&gt;tRNA modification factors</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M641.01,-448.25C641.01,-440.28 641.01,-431.39 641.01,-422.9\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"644.51,-422.83 641.01,-412.83 637.51,-422.83 644.51,-422.83\"/>\n",
       "</g>\n",
       "<!-- Chromosome and associated proteins [BR&#45;zma03036]&#45;&gt;y_hat -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>Chromosome and associated proteins [BR&#45;zma03036]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M429.08,-91.17C423.77,-81.85 417.63,-71.07 411.93,-61.07\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"414.86,-59.13 406.87,-52.18 408.78,-62.6 414.86,-59.13\"/>\n",
       "</g>\n",
       "<!-- RISC (RNA&#45;induced silencing complex)&#45;&gt;Gene silencing -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>RISC (RNA&#45;induced silencing complex)&#45;&gt;Gene silencing</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M541.01,-448.25C541.01,-440.28 541.01,-431.39 541.01,-422.9\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"544.51,-422.83 541.01,-412.83 537.51,-422.83 544.51,-422.83\"/>\n",
       "</g>\n",
       "<!-- 09102 Energy metabolism -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>09102 Energy metabolism</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"41.01\" cy=\"-385.83\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"41.01\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 5</text>\n",
       "<text text-anchor=\"middle\" x=\"41.01\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 09102 Energy metabolism&#45;&gt;09100 Metabolism -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>09102 Energy metabolism&#45;&gt;09100 Metabolism</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M54.66,-360.39C59.82,-351.15 65.79,-340.47 71.33,-330.54\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"74.45,-332.14 76.27,-321.7 68.33,-328.73 74.45,-332.14\"/>\n",
       "</g>\n",
       "<!-- 00190 Oxidative phosphorylation -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>00190 Oxidative phosphorylation</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"41.01\" cy=\"-475.57\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"41.01\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 2</text>\n",
       "<text text-anchor=\"middle\" x=\"41.01\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 00190 Oxidative phosphorylation&#45;&gt;09102 Energy metabolism -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>00190 Oxidative phosphorylation&#45;&gt;09102 Energy metabolism</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M41.01,-448.25C41.01,-440.28 41.01,-431.39 41.01,-422.9\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"44.51,-422.83 41.01,-412.83 37.51,-422.83 44.51,-422.83\"/>\n",
       "</g>\n",
       "<!-- 100383860 -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>100383860</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"91.01\" cy=\"-556.44\" rx=\"63.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"91.01\" y=\"-552.74\" font-family=\"Times,serif\" font-size=\"14.00\">100383860</text>\n",
       "</g>\n",
       "<!-- 100383860&#45;&gt;00190 Oxidative phosphorylation -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>100383860&#45;&gt;00190 Oxidative phosphorylation</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M80.16,-538.33C74.74,-529.78 68,-519.14 61.64,-509.12\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"64.51,-507.1 56.2,-500.53 58.6,-510.85 64.51,-507.1\"/>\n",
       "</g>\n",
       "<!-- 7.1.2.1 &#160;P&#45;type H+&#45;exporting transporter -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>7.1.2.1 &#160;P&#45;type H+&#45;exporting transporter</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"141.01\" cy=\"-475.57\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"141.01\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 2</text>\n",
       "<text text-anchor=\"middle\" x=\"141.01\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 100383860&#45;&gt;7.1.2.1 &#160;P&#45;type H+&#45;exporting transporter -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>100383860&#45;&gt;7.1.2.1 &#160;P&#45;type H+&#45;exporting transporter</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M101.86,-538.33C107.28,-529.78 114.03,-519.14 120.38,-509.12\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"123.42,-510.85 125.82,-500.53 117.51,-507.1 123.42,-510.85\"/>\n",
       "</g>\n",
       "<!-- 7.1 &#160;Catalysing the translocation of hydrons -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>7.1 &#160;Catalysing the translocation of hydrons</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"336.01\" cy=\"-296.09\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"336.01\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 5</text>\n",
       "<text text-anchor=\"middle\" x=\"336.01\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 7.1 &#160;Catalysing the translocation of hydrons&#45;&gt;7. Translocases -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>7.1 &#160;Catalysing the translocation of hydrons&#45;&gt;7. Translocases</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M336.01,-268.77C336.01,-260.8 336.01,-251.91 336.01,-243.42\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"339.51,-243.35 336.01,-233.35 332.51,-243.35 339.51,-243.35\"/>\n",
       "</g>\n",
       "<!-- 7.1.2 &#160;Linked to the hydrolysis of a nucleoside triphosphate -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>7.1.2 &#160;Linked to the hydrolysis of a nucleoside triphosphate</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"141.01\" cy=\"-385.83\" rx=\"41.02\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"141.01\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 5</text>\n",
       "<text text-anchor=\"middle\" x=\"141.01\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 5</text>\n",
       "</g>\n",
       "<!-- 7.1.2 &#160;Linked to the hydrolysis of a nucleoside triphosphate&#45;&gt;7.1 &#160;Catalysing the translocation of hydrons -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>7.1.2 &#160;Linked to the hydrolysis of a nucleoside triphosphate&#45;&gt;7.1 &#160;Catalysing the translocation of hydrons</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M171.95,-368.12C178.2,-364.95 184.76,-361.75 191.01,-358.96 232.24,-340.56 244.78,-341.37 286.01,-322.96 289.23,-321.52 292.54,-319.98 295.84,-318.38\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"297.67,-321.38 305.08,-313.8 294.56,-315.11 297.67,-321.38\"/>\n",
       "</g>\n",
       "<!-- 7.1.2.1 &#160;P&#45;type H+&#45;exporting transporter&#45;&gt;7.1.2 &#160;Linked to the hydrolysis of a nucleoside triphosphate -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>7.1.2.1 &#160;P&#45;type H+&#45;exporting transporter&#45;&gt;7.1.2 &#160;Linked to the hydrolysis of a nucleoside triphosphate</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M141.01,-448.25C141.01,-440.28 141.01,-431.39 141.01,-422.9\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"144.51,-422.83 141.01,-412.83 137.51,-422.83 144.51,-422.83\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph>"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dot = Digraph()\n",
    "for key in tqdm(kegg_connections.keys()):\n",
    "    key_label = 'in: '+str(input_size_dict[key])+'\\nout: '+str(output_size_dict[key])\n",
    "    dot.node(key, key_label)\n",
    "    for value in kegg_connections[key]:\n",
    "        # edge takes a head/tail whereas edges takes name pairs concatednated (A, B -> AB)in a list\n",
    "        dot.edge(value, key)    \n",
    "    \n",
    "dot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c9ebaea",
   "metadata": {},
   "source": [
    "### Set up DataSet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a550a369",
   "metadata": {},
   "source": [
    "Now we have\n",
    "\n",
    "- A dictionary with the connections: `example_dict`\n",
    "- The expected input sizes for each node: `example_dict_input_size`\n",
    "- A dictionary with the input tensors: `input_tensor_dict`\n",
    "- A list of the input tensors' names: `no_dependants` \n",
    "- A list with the order that each module should be called: `dependancy_order`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75500055",
   "metadata": {},
   "outputs": [],
   "source": [
    "class ListDataset(Dataset): # for any G containing matix with many (phno) to one (geno)\n",
    "    def __init__(self, \n",
    "                 y, \n",
    "                 x_list,\n",
    "                 transform = None, target_transform = None,\n",
    "                 use_gpu_num = 0,\n",
    "                 device = 'cuda',\n",
    "                 **kwargs \n",
    "                ):\n",
    "        self.device = device\n",
    "        self.y = y \n",
    "        self.x_list = x_list,\n",
    "        self.transform = transform\n",
    "        self.target_transform = target_transform    \n",
    "        \n",
    "    def __len__(self):\n",
    "        return len(self.y)\n",
    "    \n",
    "    def __getitem__(self, idx):\n",
    "        y_idx =self.y[idx]\n",
    "        x_idx =[x[idx] for x in self.x_list[0]] # must be the 0th entry because it's in a tuple\n",
    "        \n",
    "        if self.target_transform:\n",
    "            y_idx = self.transform(y_idx)\n",
    "            x_idx = [self.transform(x) for x in x_idx]\n",
    "            \n",
    "        return y_idx, x_idx\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26f5c6e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "x_list_temp = [torch.from_numpy(input_tensor_dict[key]).to(torch.float) for key in input_tensor_dict.keys()]\n",
    "[e.shape for e in x_list_temp]\n",
    "y_temp = torch.from_numpy(YMat[0:4926,]).to(torch.float)#[:, None]\n",
    "y_temp.shape\n",
    "\n",
    "training_dataloader = DataLoader(ListDataset(\n",
    "        y = y_temp,\n",
    "        x_list = x_list_temp,\n",
    "        use_gpu_num = 0,\n",
    "        device = 'cuda',\n",
    "    ),\n",
    "    batch_size = 5,\n",
    "    shuffle = True\n",
    ")\n",
    "\n",
    "# next(iter(training_dataloader))[1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c21850c",
   "metadata": {},
   "source": [
    "### Set up NeuralNetwork"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a661128",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a86d1db0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[[0.0600],\n",
       "         [0.0696],\n",
       "         [0.0642],\n",
       "         [0.0710]],\n",
       "\n",
       "        [[0.0620],\n",
       "         [0.0702],\n",
       "         [0.0646],\n",
       "         [0.0687]],\n",
       "\n",
       "        [[0.0605],\n",
       "         [0.0705],\n",
       "         [0.0652],\n",
       "         [0.0687]],\n",
       "\n",
       "        [[0.0616],\n",
       "         [0.0659],\n",
       "         [0.0678],\n",
       "         [0.0710]],\n",
       "\n",
       "        [[0.0605],\n",
       "         [0.0682],\n",
       "         [0.0698],\n",
       "         [0.0693]]], grad_fn=<ReluBackward0>)"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# first pass is by hand.\n",
    "class NeuralNetwork(nn.Module):\n",
    "    def __init__(self, \n",
    "                 example_dict, # contains the node (excluding input tensors)\n",
    "                 example_dict_input_size, # contains the input sizes (including the tensors)\n",
    "                 example_dict_output_size,\n",
    "                 input_tensor_names,\n",
    "                 dependancy_order\n",
    "                ):\n",
    "        super(NeuralNetwork, self).__init__()\n",
    "        def Linear_block(in_size, out_size, drop_pr):\n",
    "            block = nn.Sequential(\n",
    "                nn.Linear(in_size, out_size),\n",
    "                nn.ReLU(),\n",
    "                nn.Dropout(drop_pr))\n",
    "            return(block)   \n",
    "        \n",
    "        self.nn_dict = {}\n",
    "        # iterate through the input dictionary and add a linear block for each node (key) included \n",
    "        for key in example_dict.keys():\n",
    "            self.nn_dict[key] = Linear_block(in_size=example_dict_input_size[key], \n",
    "                                             out_size=example_dict_output_size[key], \n",
    "                                             drop_pr=0)\n",
    "        # things for get_input_node in forward to work.\n",
    "        self.example_dict = example_dict\n",
    "        self.input_tensor_names = input_tensor_names\n",
    "        self.dependancy_order = dependancy_order\n",
    "        \n",
    "        self.input_tensor_lookup = dict(zip(input_tensor_names, \n",
    "                                            [i for i in range(len(input_tensor_names))]))\n",
    "        self.result_list = []\n",
    "        self.result_list_lookup = {}\n",
    "\n",
    "    def forward(self, x):\n",
    "        # Note: x will be a list. input_tensor_lookup will contain the name: list index pairs.\n",
    "        # I use a dict instead of a list comprehension here because there could be an arbitrarily\n",
    "        # large number of inputs in the list. \n",
    "        def get_input_node(self, input_node, get_x):    \n",
    "            if input_node in self.input_tensor_names:\n",
    "                return(get_x[self.input_tensor_lookup[input_node]])\n",
    "            else:\n",
    "                return(self.result_list[self.result_list_lookup[input_node]])\n",
    "            \n",
    "        for key in self.dependancy_order:\n",
    "            input_nodes = self.example_dict[key]\n",
    "            \n",
    "            if key not in self.result_list_lookup.keys():\n",
    "                # I realllllly hope these are being copied by reference...\n",
    "                # must be first before growing result_list\n",
    "                self.result_list_lookup[key] = len(self.result_list_lookup)\n",
    "                    \n",
    "                self.result_list += [self.nn_dict[key](torch.concat(\n",
    "                    [get_input_node(self, input_node = e, get_x = x) for e in input_nodes], \n",
    "                    -1))] # Concat based on the last dim\n",
    "\n",
    "\n",
    "            else:\n",
    "                print('b')\n",
    "                # Note! here I'm saving the tensor without wrapping it in a list.\n",
    "                self.result_list[self.result_list_lookup[key]] = self.nn_dict[key](torch.concat(\n",
    "                    [get_input_node(self, input_node = e, get_x = x) for e in input_nodes], \n",
    "                    -1))\n",
    "                \n",
    "        return self.result_list[self.result_list_lookup['y_hat']]\n",
    "    \n",
    "model = NeuralNetwork(example_dict = kegg_connections, \n",
    "                      example_dict_input_size = input_size_dict,\n",
    "                      example_dict_output_size = output_size_dict,\n",
    "                      input_tensor_names = list(input_tensor_dict.keys()),\n",
    "                      dependancy_order = dependancy_order)    \n",
    "\n",
    "\n",
    "yhat = model(next(iter(training_dataloader))[1])\n",
    "yhat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30ddacf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# torchviz.make_dot(\n",
    "#     yhat.mean(), \n",
    "#                   params=dict(model.named_parameters()), \n",
    "# #                   show_attrs=True, \n",
    "# #                   show_saved=True\n",
    "#                  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa9d422c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# obs_geno_lookup # Phno_Idx\tGeno_Idx\tIs_Phno_Idx\n",
    "# YMat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56aaad6e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "171c02dc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4642a08",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db5841e8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "316c7cd2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a495dd5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d235c702",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9ac4085",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4129310d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2191863",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80c37ef3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "02205e69",
   "metadata": {},
   "source": [
    "## One Hot Encoded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3f93d53",
   "metadata": {},
   "outputs": [],
   "source": [
    "rng = np.random.default_rng(9230473) # note, must use rng.shuffle(arr) below for this to take effect."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50a91e51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hybrid</th>\n",
       "      <th>n</th>\n",
       "      <th>pct</th>\n",
       "      <th>pct_cumsum</th>\n",
       "      <th>random_order</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2369/DK3IIH6</td>\n",
       "      <td>113</td>\n",
       "      <td>0.000849</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2369/PHN82</td>\n",
       "      <td>62</td>\n",
       "      <td>0.000466</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2369/PHZ51</td>\n",
       "      <td>108</td>\n",
       "      <td>0.000812</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2FACC/DK3IIH6</td>\n",
       "      <td>40</td>\n",
       "      <td>0.000301</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4N506/DK3IIH6</td>\n",
       "      <td>37</td>\n",
       "      <td>0.000278</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4921</th>\n",
       "      <td>Z037E0054/LH162</td>\n",
       "      <td>10</td>\n",
       "      <td>0.000075</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4922</th>\n",
       "      <td>Z037E0054/PHZ51</td>\n",
       "      <td>23</td>\n",
       "      <td>0.000173</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4923</th>\n",
       "      <td>Z038E0057/DK3IIH6</td>\n",
       "      <td>37</td>\n",
       "      <td>0.000278</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4924</th>\n",
       "      <td>Z038E0057/LH162</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4925</th>\n",
       "      <td>Z038E0057/PHZ51</td>\n",
       "      <td>36</td>\n",
       "      <td>0.000271</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4926 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Hybrid    n       pct  pct_cumsum  random_order\n",
       "0          2369/DK3IIH6  113  0.000849         NaN             0\n",
       "1            2369/PHN82   62  0.000466         NaN             0\n",
       "2            2369/PHZ51  108  0.000812         NaN             0\n",
       "3         2FACC/DK3IIH6   40  0.000301         NaN             0\n",
       "4         4N506/DK3IIH6   37  0.000278         NaN             0\n",
       "...                 ...  ...       ...         ...           ...\n",
       "4921    Z037E0054/LH162   10  0.000075         NaN             0\n",
       "4922    Z037E0054/PHZ51   23  0.000173         NaN             0\n",
       "4923  Z038E0057/DK3IIH6   37  0.000278         NaN             0\n",
       "4924    Z038E0057/LH162   11  0.000083         NaN             0\n",
       "4925    Z038E0057/PHZ51   36  0.000271         NaN             0\n",
       "\n",
       "[4926 rows x 5 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pct_cumsum_thresh = .9\n",
    "\n",
    "# make a df to aid in creating train/test splits\n",
    "# the plan is to shuffle the rows of the df, calculate the cumulative sum of the percents obs, then \n",
    "# the entries above and below a given percent will be the train/test.\n",
    "obs_per_hybrid = phno.assign(n = 1).groupby('Hybrid').count().reset_index()\n",
    "obs_per_hybrid = obs_per_hybrid.loc[:, ['Hybrid', 'n']]\n",
    "obs_per_hybrid['pct'] = obs_per_hybrid.n / obs_per_hybrid.n.sum()\n",
    "obs_per_hybrid['pct_cumsum'] = np.nan\n",
    "obs_per_hybrid['random_order'] = 0\n",
    "\n",
    "obs_per_hybrid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9ca1bfb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hybrid</th>\n",
       "      <th>n</th>\n",
       "      <th>pct</th>\n",
       "      <th>pct_cumsum</th>\n",
       "      <th>random_order</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1738</th>\n",
       "      <td>MOG_MO45-074-1-1-1-1-B/LH185</td>\n",
       "      <td>13</td>\n",
       "      <td>0.000098</td>\n",
       "      <td>0.000098</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1196</th>\n",
       "      <td>M0276/LH198</td>\n",
       "      <td>21</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>0.000256</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>CG102/DK3IIH6</td>\n",
       "      <td>64</td>\n",
       "      <td>0.000481</td>\n",
       "      <td>0.000737</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4735</th>\n",
       "      <td>Z022E0023/PB80</td>\n",
       "      <td>13</td>\n",
       "      <td>0.000098</td>\n",
       "      <td>0.000834</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1647</th>\n",
       "      <td>MO44_PHW65_0416/LH195</td>\n",
       "      <td>47</td>\n",
       "      <td>0.000353</td>\n",
       "      <td>0.001187</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>BGEM-0211-N/LH195</td>\n",
       "      <td>45</td>\n",
       "      <td>0.000338</td>\n",
       "      <td>0.999557</td>\n",
       "      <td>4921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>B73_PHG39-15/LH82</td>\n",
       "      <td>8</td>\n",
       "      <td>0.000060</td>\n",
       "      <td>0.999617</td>\n",
       "      <td>4922</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2682</th>\n",
       "      <td>PHW52/PHK76</td>\n",
       "      <td>28</td>\n",
       "      <td>0.000210</td>\n",
       "      <td>0.999827</td>\n",
       "      <td>4923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3249</th>\n",
       "      <td>W10001_0013/PHB47</td>\n",
       "      <td>12</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.999917</td>\n",
       "      <td>4924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1330</th>\n",
       "      <td>MO44_LH145_0081/PB80</td>\n",
       "      <td>11</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>4925</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4926 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            Hybrid   n       pct  pct_cumsum  random_order\n",
       "1738  MOG_MO45-074-1-1-1-1-B/LH185  13  0.000098    0.000098             0\n",
       "1196                   M0276/LH198  21  0.000158    0.000256             1\n",
       "256                  CG102/DK3IIH6  64  0.000481    0.000737             2\n",
       "4735                Z022E0023/PB80  13  0.000098    0.000834             3\n",
       "1647         MO44_PHW65_0416/LH195  47  0.000353    0.001187             4\n",
       "...                            ...  ..       ...         ...           ...\n",
       "194              BGEM-0211-N/LH195  45  0.000338    0.999557          4921\n",
       "106              B73_PHG39-15/LH82   8  0.000060    0.999617          4922\n",
       "2682                   PHW52/PHK76  28  0.000210    0.999827          4923\n",
       "3249             W10001_0013/PHB47  12  0.000090    0.999917          4924\n",
       "1330          MO44_LH145_0081/PB80  11  0.000083    1.000000          4925\n",
       "\n",
       "[4926 rows x 5 columns]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fill in the random values to sort on\n",
    "arr = np.arange(obs_per_hybrid.shape[0])\n",
    "rng.shuffle(arr)\n",
    "obs_per_hybrid.random_order = arr\n",
    "\n",
    "obs_per_hybrid = obs_per_hybrid.sort_values('random_order')\n",
    "obs_per_hybrid['pct_cumsum'] = obs_per_hybrid.pct.cumsum()\n",
    "obs_per_hybrid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16bce9d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert back into phno indices\n",
    "train_hybrids = list(obs_per_hybrid.loc[(obs_per_hybrid.pct_cumsum <= pct_cumsum_thresh), 'Hybrid'])\n",
    "test_hybrids  = list(obs_per_hybrid.loc[(obs_per_hybrid.pct_cumsum >  pct_cumsum_thresh), 'Hybrid'])\n",
    "\n",
    "train_idx = phno.loc[(phno.Hybrid.isin(train_hybrids)), ].index\n",
    "test_idx  = phno.loc[(phno.Hybrid.isin(test_hybrids)), ].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2d0163e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[119735, 13322]"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[len(train_idx), len(test_idx)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbaedf07",
   "metadata": {},
   "outputs": [],
   "source": [
    "# confirm all observation idxs are have genomic information\n",
    "assert [] == [e for e in list(train_idx)+list(test_idx) if e not in obs_geno_lookup[:, 0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02004189",
   "metadata": {},
   "outputs": [],
   "source": [
    "YMat_cs = calc_cs(YMat[train_idx])\n",
    "y_cs = apply_cs(YMat, YMat_cs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f15ee06c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# training_dataloader = DataLoader(\n",
    "#     ACGTDataset(y = torch.from_numpy(y_cs[train_idx])[:, None].to(torch.float), \n",
    "#                 G = torch.from_numpy(ACGT_hilb).to(torch.float), \n",
    "#                 idx_original = torch.from_numpy(np.array(train_idx)),\n",
    "#                 idx_lookup   = torch.from_numpy(np.asarray(obs_geno_lookup)),\n",
    "#                 use_gpu_num = 0,\n",
    "#                 device = 'cuda'\n",
    "#                ),\n",
    "#     batch_size = 50,\n",
    "#     shuffle = True\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08eaaff0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# testing_dataloader = DataLoader(\n",
    "#     ACGTDataset(y = torch.from_numpy(y_cs[test_idx])[:, None].to(torch.float), \n",
    "#                 G = torch.from_numpy(ACGT_hilb).to(torch.float), \n",
    "#                 idx_original = torch.from_numpy(np.array(test_idx)),\n",
    "#                 idx_lookup   = torch.from_numpy(np.asarray(obs_geno_lookup)),\n",
    "#                 use_gpu_num = 0,\n",
    "#                 device = 'cuda'\n",
    "#                ),\n",
    "#     batch_size = 50,\n",
    "#     shuffle = True\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82b8b1fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# next(iter(training_dataloader))[0].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "788ba1e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cf76c08",
   "metadata": {},
   "outputs": [],
   "source": [
    "# class NeuralNetwork(nn.Module):\n",
    "#     def __init__(self):\n",
    "#         super(NeuralNetwork, self).__init__()    \n",
    "\n",
    "#         def Linear_block(in_size, out_size, drop_pr):\n",
    "#             block = nn.Sequential(\n",
    "#                 nn.Linear(in_size, out_size),\n",
    "#                 nn.ReLU(),\n",
    "#                 nn.Dropout(drop_pr)\n",
    "#             )\n",
    "#             return(block)         \n",
    "        \n",
    "        \n",
    "# #         def Conv1D_Max_block(in_channels, out_channels, kernel_size, stride):\n",
    "# #             block = nn.Sequential(\n",
    "# #                 nn.Conv1d(\n",
    "# #                     in_channels= in_channels, # second channel\n",
    "# #                     out_channels= out_channels,\n",
    "# #                     kernel_size= kernel_size,\n",
    "# #                     stride= stride\n",
    "# #                 ), \n",
    "# #                 nn.MaxPool1d((kernel_size,), stride=stride)\n",
    "# #             )\n",
    "# #             return(block)\n",
    "        \n",
    "#         self.x_network = nn.Sequential(\n",
    "#             nn.Conv2d(\n",
    "#                     in_channels= 4, \n",
    "#                     out_channels= 4,\n",
    "#                     kernel_size= (3, 3),\n",
    "#                     stride= 2,\n",
    "#                     padding = 1,\n",
    "#                     bias = True\n",
    "#                 ),\n",
    "#             nn.Conv2d(\n",
    "#                     in_channels= 4, \n",
    "#                     out_channels= 4,\n",
    "#                     kernel_size= (3, 3),\n",
    "#                     stride= 2,\n",
    "#                     padding = 1,\n",
    "#                     bias = True\n",
    "#                 ),\n",
    "#             nn.Conv2d(\n",
    "#                     in_channels= 4, \n",
    "#                     out_channels= 4,\n",
    "#                     kernel_size= (3, 3),\n",
    "#                     stride= 2,\n",
    "#                     padding = 1,\n",
    "#                     bias = True\n",
    "#                 ),\n",
    "#             nn.Conv2d(\n",
    "#                     in_channels= 4, \n",
    "#                     out_channels= 4,\n",
    "#                     kernel_size= (3, 3),\n",
    "#                     stride= 2,\n",
    "#                     padding = 1,\n",
    "#                     bias = True\n",
    "#                 ),\n",
    "#             nn.Conv2d(\n",
    "#                     in_channels= 4, \n",
    "#                     out_channels= 4,\n",
    "#                     kernel_size= (3, 3),\n",
    "#                     stride= 2,\n",
    "#                     padding = 1,\n",
    "#                     bias = True\n",
    "#                 ), \n",
    "#             nn.Flatten(),            \n",
    "# #             Linear_block(in_size = 116, out_size = 32, drop_pr = 0.3),\n",
    "#             nn.Linear(512, 1)\n",
    "#         )\n",
    "        \n",
    "#     def forward(self, x):\n",
    "#         x_out = self.x_network(x)\n",
    "#         return x_out\n",
    "\n",
    "# model = NeuralNetwork().to(device)\n",
    "\n",
    "# model(next(iter(training_dataloader))[0]).shape\n",
    "\n",
    "# # torch.Size([50, 4, 256, 512])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "654b3bf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # small scale test\n",
    "# model, loss_df = train_nn(\n",
    "#     cache_path,\n",
    "#     training_dataloader,\n",
    "#     testing_dataloader,\n",
    "#     model,\n",
    "#     learning_rate = 1e-3,\n",
    "#     batch_size = 50, #dataloader_batch_size,\n",
    "#     epochs = 1 #(run_epochs - epochs_run)\n",
    "# )\n",
    "\n",
    "# loss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e9cbc7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# estimate_iterations(sec_per_it = 95)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27248cc1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove_matching_files(\n",
    "#     cache_path,\n",
    "#     match_regex_list = ['model\\.pt'],\n",
    "#     dry_run = False\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3174cd52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dataloader_batch_size = 50\n",
    "# run_epochs = 100\n",
    "\n",
    "# # don't run if either of these exist because there may be cases where we want the results but not the model\n",
    "# import re\n",
    "\n",
    "# if not os.path.exists(cache_path+'/model.pt'): \n",
    "#     # Shared setup (train from scratch and load latest)\n",
    "#     model = NeuralNetwork()\n",
    "\n",
    "#     # find the biggest model to save\n",
    "#     saved_models = os.listdir(cache_path)\n",
    "#     saved_models = [e for e in saved_models if re.match('model*', e)]\n",
    "\n",
    "#     if saved_models == []:\n",
    "#         epochs_run = 0\n",
    "#     else:\n",
    "#         saved_models = [e for e in saved_models if e != 'model.pt']\n",
    "#         # if there are saved models reload and resume training\n",
    "#         saved_models_numbers = [int(e.replace('model_', ''\n",
    "#                                     ).replace('.pt', ''\n",
    "#                                     ).split('_')[0]) for e in saved_models]\n",
    "#         # saved_models\n",
    "#         epochs_run = max(saved_models_numbers)+1 # add 1 to account for 0 index\n",
    "#         latest_model = [e for e in saved_models if re.match(\n",
    "#             '^model_'+str(epochs_run-1)+'_.*\\.pt$', e)][0] # subtract 1 to convert back\n",
    "#         model.load_state_dict(torch.load(cache_path+latest_model))\n",
    "#         print('Resuming Training: '+str(epochs_run)+'/'+str(run_epochs)+' epochs run.')\n",
    "    \n",
    "#     model.to(device)    \n",
    "\n",
    "#     model, loss_df = train_nn(\n",
    "#         cache_path,\n",
    "#         training_dataloader,\n",
    "#         testing_dataloader,\n",
    "#         model,\n",
    "#         learning_rate = 1e-3,\n",
    "#         batch_size = dataloader_batch_size,\n",
    "#         epochs = (run_epochs - epochs_run)\n",
    "#     )\n",
    "    \n",
    "#     # experimental outputs:\n",
    "#     # 1. Model\n",
    "#     torch.save(model.state_dict(), cache_path+'/model.pt') # convention is to use .pt or .pth\n",
    "\n",
    "#     # 2. loss_df    \n",
    "#     # If this is resuming training, load and extend the existing loss dataframe\n",
    "#     if os.path.exists(cache_path+'/loss_df.csv'):\n",
    "#         loss_df_on_disk = pd.read_csv(cache_path+'/loss_df.csv')\n",
    "#         epoch_offset = 1 + loss_df_on_disk['Epoch'].max()\n",
    "#         loss_df['Epoch'] = loss_df['Epoch'] + epoch_offset\n",
    "#         loss_df = pd.concat([loss_df_on_disk, loss_df])\n",
    "#     loss_df.to_csv(cache_path+'/loss_df.csv', index=False)  \n",
    "    \n",
    "#     # 3. predictions \n",
    "#     yhats = pd.concat([\n",
    "#         yhat_loop(testing_dataloader, model).assign(Split = 'Test'),\n",
    "#         yhat_loop(training_dataloader, model).assign(Split = 'Train')], axis = 0)\n",
    "\n",
    "#     yhats.to_csv(cache_path+'/yhats.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4764b227",
   "metadata": {},
   "source": [
    "### Standard Visualizations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "127430f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2f27738",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fd9373f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87f93f99",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34602160",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad680db4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# scale_dict = {'y1':YMat_cs}\n",
    "# import plotly.graph_objects as go"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59662f6a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ec22f32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# naieve_yhat = training_dataloader.dataset.y.mean()\n",
    "\n",
    "# naieve_MSE_Train = reverse_cs( \n",
    "#     np.array(((naieve_yhat - training_dataloader.dataset.y)**2)).mean(),\n",
    "#     scale_dict['y1']\n",
    "# )\n",
    "\n",
    "# naieve_MSE_Test = reverse_cs( \n",
    "#     np.array(((naieve_yhat - testing_dataloader.dataset.y)**2)).mean(),\n",
    "#     scale_dict['y1']\n",
    "# )\n",
    "\n",
    "# naieve_MSE_Train, naieve_MSE_Test\n",
    "\n",
    "\n",
    "\n",
    "# loss_df = pd.read_csv(cache_path+'/loss_df.csv')\n",
    "\n",
    "# loss_df.TrainMSE = reverse_cs(loss_df.TrainMSE, scale_dict['y1'])\n",
    "# loss_df.TestMSE  = reverse_cs(loss_df.TestMSE , scale_dict['y1'])\n",
    "\n",
    "\n",
    "# fig = go.Figure()\n",
    "# fig.add_trace(go.Scatter(x=loss_df.Epoch, y=loss_df.TestMSE,\n",
    "#                     mode='lines', name='Test'))\n",
    "# fig.add_trace(go.Scatter(x=loss_df.Epoch, y=loss_df.TrainMSE,\n",
    "#                     mode='lines', name='Train'))\n",
    "\n",
    "# fig.add_trace(go.Scatter(x=loss_df.Epoch, y=[naieve_MSE_Test  for e in range(len(loss_df.Epoch))], \n",
    "#                          mode='lines', name='Naieve Test'))\n",
    "# fig.add_trace(go.Scatter(x=loss_df.Epoch, y=[naieve_MSE_Train for e in range(len(loss_df.Epoch))], \n",
    "#                          mode='lines', name='Naieve Train'))\n",
    "# fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6b66997",
   "metadata": {},
   "outputs": [],
   "source": [
    "# yhats = pd.read_csv(cache_path+'/yhats.csv')\n",
    "\n",
    "# # px.scatter(yhats, x = 'y_true', y = 'y_pred', color = 'Split')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "923b6dab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# yhats.y_true = reverse_cs(yhats.y_true, scale_dict['y1'])\n",
    "# yhats.y_pred = reverse_cs(yhats.y_pred, scale_dict['y1'])\n",
    "\n",
    "# # px.scatter(yhats, x = 'y_true', y = 'y_pred', color = 'Split', trendline=\"ols\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8460bc23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# yhats['Error'] = yhats.y_pred - yhats.y_true\n",
    "\n",
    "# px.histogram(yhats, x = 'Error', color = 'Split',\n",
    "#              marginal=\"box\", # can be `rug`, `violin`\n",
    "#              nbins= 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21076dbd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
