{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7aadccc0",
   "metadata": {},
   "source": [
    "# G only KEGG based network architecture\n",
    "\n",
    "> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "392ee9d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from EnvDL.core import ensure_dir_path_exists \n",
    "from EnvDL.dlfn import g2fc_datawrapper, BigDataset, plDNN_general\n",
    "from EnvDL.dlfn import ResNet2d, BasicBlock2d\n",
    "from EnvDL.dlfn import LSUV_\n",
    "\n",
    "import torch\n",
    "import torch.nn.functional as F # F.mse_loss\n",
    "from torch.utils.data import Dataset\n",
    "from torch.utils.data import DataLoader\n",
    "from torch import nn\n",
    "\n",
    "import lightning.pytorch as pl\n",
    "from lightning.pytorch.loggers import TensorBoardLogger"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "9f4f7771",
   "metadata": {},
   "outputs": [],
   "source": [
    "from tqdm import tqdm\n",
    "from graphviz import Digraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "302bf05a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run settings: \n",
    "max_epoch = 2\n",
    "batch_size = 24"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ce01a01c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using cuda device\n"
     ]
    }
   ],
   "source": [
    "use_gpu_num = 0\n",
    "\n",
    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
    "if use_gpu_num in [0, 1]: \n",
    "    torch.cuda.set_device(use_gpu_num)\n",
    "print(f\"Using {device} device\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "b6fba3df",
   "metadata": {},
   "outputs": [],
   "source": [
    "cache_path = '../nbs_artifacts/02.31_g2fc_G_ACGT_VNN__scratchpad/'\n",
    "ensure_dir_path_exists(dir_path = cache_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4f56cc68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Settings:\n",
    "## For graph linear blocks\n",
    "default_output_size = 50\n",
    "default_dropout_pr = 0.1\n",
    "default_block_reps = 1\n",
    "\n",
    "# ## Training settings\n",
    "# run_epochs = 15\n",
    "# epochs_run = 0\n",
    "# dataloader_batch_size = 1000"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff84bdd0",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "547c1e1f",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "9c52e21d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading and storing default `phno`.\n"
     ]
    }
   ],
   "source": [
    "X = g2fc_datawrapper()\n",
    "X.set_split()\n",
    "X.load_all(name_list = ['obs_geno_lookup', 'YMat', 'KEGG_slices',], store=True) \n",
    "\n",
    "X.calc_cs('YMat', version = 'np', filter = 'val:train')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "4c7f9e7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ACGT_gene_slice_list = X.get('KEGG_slices',     ops_string='')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87bfb7bb",
   "metadata": {},
   "source": [
    "## Generate Graph for DNN"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac6afc90",
   "metadata": {},
   "source": [
    "### Functions for Graph Construction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "5b924e03",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Building a Neural Net from an arbitrary graph\n",
    "# start by finding the top level -- all those keys which are theselves not values\n",
    "# helper function to get all keys and all value from a dict. Useful for when keys don't have unique values.\n",
    "def find_uniq_keys_values(input_dict):\n",
    "    all_keys = list(input_dict.keys())\n",
    "    all_values = []\n",
    "    for e in all_keys:\n",
    "        all_values.extend(input_dict[e])\n",
    "    all_values = list(set(all_values))\n",
    "\n",
    "    return({'all_keys': all_keys,\n",
    "           'all_values': all_values})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6dd645ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Find order that nodes in the graph should be called to have all dependencies run when they are called.\n",
    "# find the dependancies for run order from many dependancies to none\n",
    "# wrapper function to find the nodes that aren't any other nodes dependancies.\n",
    "def find_top_nodes(all_key_value_dict):\n",
    "    return([e for e in all_key_value_dict['all_keys'] if e not in all_key_value_dict['all_values']])\n",
    "# wrapper function to find the input nodes. They don't occur in the keys and thus won't be added to the list otherwise.\n",
    "# another way to do this would have been to \n",
    "def find_input_nodes(all_key_value_dict):\n",
    "    return([e for e in all_key_value_dict['all_values'] if e not in all_key_value_dict['all_keys']])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bd67297",
   "metadata": {},
   "source": [
    "### Process paths into graphs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "be294e5d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Retaining 43.53%, 6067/13939 Entries\n"
     ]
    }
   ],
   "source": [
    "parsed_kegg_gene_entries = X.get('KEGG_entries')\n",
    "\n",
    "# Restrict to only those with pathway\n",
    "kegg_gene_brite = [e for e in parsed_kegg_gene_entries if 'BRITE' in e.keys()]\n",
    "\n",
    "# also require to have a non-empty path\n",
    "kegg_gene_brite = [e for e in kegg_gene_brite if not e['BRITE']['BRITE_PATHS'] == []]\n",
    "\n",
    "print('Retaining '+ str(round(len(kegg_gene_brite)/len(parsed_kegg_gene_entries), 4)*100)+'%, '+str(len(kegg_gene_brite)\n",
    "     )+'/'+str(len(parsed_kegg_gene_entries)\n",
    "     )+' Entries'\n",
    "     )\n",
    "# kegg_gene_brite[1]['BRITE']['BRITE_PATHS']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0bea378",
   "metadata": {},
   "source": [
    "### Tiny `n`  gene version of the network:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "30b38dcb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using 6067/6067 genes.\n"
     ]
    }
   ],
   "source": [
    "n_genes = 6067 \n",
    "\n",
    "print('Using '+str(n_genes)+'/'+str(len(kegg_gene_brite))+' genes.')\n",
    "\n",
    "# if n_genes is too big, don't visualize.\n",
    "vis_dot_bool = True\n",
    "if n_genes > 10:\n",
    "    vis_dot_bool = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4bae9662",
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOD is list of dicts.\n",
    "# def kegg_brite_LOD_to_connections(\n",
    "def kegg_connections_build(\n",
    "        n_genes, \n",
    "        kegg_gene_brite):\n",
    "    \"\"\"\n",
    "    The goal here is to have a dict with each node and a list of it's children. \n",
    "    For example, the graph\n",
    "    a--b--d\n",
    "    |-c--e\n",
    "    Would be parsed into     \n",
    "    {'a':['b', 'c'],\n",
    "    'b':['d'],\n",
    "    'c':['e']}\n",
    "    \"\"\"\n",
    "    kegg_connections = {}\n",
    "    # for all genes in list\n",
    "    for i in tqdm(range(n_genes)): \n",
    "        temp = kegg_gene_brite[i]['BRITE']['BRITE_PATHS']\n",
    "        # clean up to make sure that there are no \":\" characters. These can mess up graphviz\n",
    "        temp = [[temp[j][i].replace(':', '-') for i in range(len(temp[j])) ] for j in range(len(temp))]\n",
    "        # all paths through graph associated with a gene\n",
    "        for j in range(len(temp)):\n",
    "            # steps of the path through the graph\n",
    "            for k in range(len(temp[j])-1):\n",
    "                \n",
    "                # name standardization \n",
    "                temp_jk  = temp[j][k]\n",
    "                temp_jk1 = temp[j][k+1]\n",
    "                temp_jk  = temp_jk.lower().title().replace(' ', '')\n",
    "                temp_jk1 = temp_jk1.lower().title().replace(' ', '')\n",
    "                \n",
    "                # if this is a new key, add it and add the k+1 entry as it's child\n",
    "                if temp_jk  not in kegg_connections.keys():\n",
    "                    kegg_connections[temp_jk] = [temp_jk1]\n",
    "                else: \n",
    "                    # Check to see if there's a new child to add   \n",
    "                    if temp_jk1 not in kegg_connections[temp_jk]:\n",
    "                        # make sure that no key contains itself. This was a problem for 'Others' which is now disallowed.\n",
    "                        if (temp_jk != temp_jk1):\n",
    "                            # add it.\n",
    "                            kegg_connections[temp_jk].extend([temp_jk1])\n",
    "    return(kegg_connections)          \n",
    "\n",
    "# kegg_brite_LOD_to_connections(\n",
    "#     n_genes = 6, \n",
    "#     kegg_gene_brite = kegg_gene_brite)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "0bc8ce75",
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"\"\"\n",
    "# The goal here is to have a dict with each node and a list of it's children. \n",
    "# For example, the graph\n",
    "# a--b--d\n",
    "#  |-c--e\n",
    "# Would be parsed into     \n",
    "# {'a':['b', 'c'],\n",
    "#  'b':['d'],\n",
    "#  'c':['e']}\n",
    "# \"\"\"\n",
    "# kegg_connections = {}\n",
    "\n",
    "# # for all genes in list\n",
    "# for i in tqdm(range(n_genes)): \n",
    "# # for i in tqdm(range(len(kegg_gene_brite))):    \n",
    "#     temp = kegg_gene_brite[i]['BRITE']['BRITE_PATHS']\n",
    "#     # clean up to make sure that there are no \":\" characters. These can mess up graphviz\n",
    "#     temp = [[temp[j][i].replace(':', '-') for i in range(len(temp[j])) ] for j in range(len(temp))]\n",
    "#     # all paths through graph associated with a gene\n",
    "#     for j in range(len(temp)):\n",
    "#         # steps of the path through the graph\n",
    "#         for k in range(len(temp[j])-1):\n",
    "            \n",
    "#             # name standardization \n",
    "#             temp_jk  = temp[j][k]\n",
    "#             temp_jk1 = temp[j][k+1]\n",
    "#             temp_jk  = temp_jk.lower().title().replace(' ', '')\n",
    "#             temp_jk1 = temp_jk1.lower().title().replace(' ', '')\n",
    "            \n",
    "#             # if this is a new key, add it and add the k+1 entry as it's child\n",
    "#             if temp_jk  not in kegg_connections.keys():\n",
    "#                 kegg_connections[temp_jk] = [temp_jk1]\n",
    "#             else: \n",
    "#                 # Check to see if there's a new child to add   \n",
    "#                 if temp_jk1 not in kegg_connections[temp_jk]:\n",
    "#                     # make sure that no key contains itself. This was a problem for 'Others' which is now disallowed.\n",
    "#                     if (temp_jk != temp_jk1):\n",
    "# #                         if ((temp_jk  != temp_jk1) & (temp_jk1 != 'Others')):\n",
    "#                         # add it.\n",
    "#                         kegg_connections[temp_jk].extend([temp_jk1])\n",
    "\n",
    "# # kegg_connections                        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "7ba4d769",
   "metadata": {},
   "outputs": [],
   "source": [
    "def kegg_connections_clean(kegg_connections):\n",
    "    if 'Others' in kegg_connections.keys():\n",
    "        del kegg_connections['Others']\n",
    "        print('Removed node \"Others\"')\n",
    "\n",
    "    # remove 'Others' as a possible value\n",
    "    for key in kegg_connections.keys():\n",
    "        kegg_connections[key] = [e for e in kegg_connections[key] if e != 'Others']\n",
    "\n",
    "    # Make sure that no list contains it's own key\n",
    "    for key in kegg_connections.keys():\n",
    "        kegg_connections[key] = [e for e in kegg_connections[key] if e != key]\n",
    "\n",
    "    # there might be associations with no dependants and with no dependants except those that have no dependants.\n",
    "    # Build up a list with those keys that don't connect back to snps then I'll pass over the connection dict once to remove references to them.\n",
    "    rm_list = []\n",
    "    rm_list_i = len(rm_list)\n",
    "    rm_list_j = -1\n",
    "    for i in range(100):\n",
    "        if rm_list_i == rm_list_j:\n",
    "            break\n",
    "        else:\n",
    "            rm_list = [key for key in kegg_connections.keys() if [e for e in kegg_connections[key] if e not in rm_list]\n",
    "        ==[]]\n",
    "            rm_list_j = rm_list_i \n",
    "            rm_list_i = len(rm_list)\n",
    "    # rm_list\n",
    "\n",
    "    for key in rm_list:\n",
    "        del kegg_connections[key]\n",
    "        \n",
    "    for key in kegg_connections.keys():\n",
    "        kegg_connections[key] = [e for e in kegg_connections[key] if e not in rm_list]\n",
    "    return kegg_connections\n",
    "\n",
    "# kegg_connections_clean(kegg_connections = kegg_connections_build(n_genes = 6, kegg_gene_brite = kegg_gene_brite) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "9c110e8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if 'Others' in kegg_connections.keys():\n",
    "#     del kegg_connections['Others']\n",
    "#     print('Removed node \"Others\"')\n",
    "\n",
    "# # remove 'Others' as a possible value\n",
    "# for key in kegg_connections.keys():\n",
    "#     kegg_connections[key] = [e for e in kegg_connections[key] if e != 'Others']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "9f1c1f1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Make sure that no list contains it's own key\n",
    "# for key in kegg_connections.keys():\n",
    "#     kegg_connections[key] = [e for e in kegg_connections[key] if e != key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "b66ec6db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # there might be associations with no dependants and with no dependants except those that have no dependants.\n",
    "# # Build up a list with those keys that don't connect back to snps then I'll pass over the connection dict once to remove references to them.\n",
    "# rm_list = []\n",
    "# rm_list_i = len(rm_list)\n",
    "# rm_list_j = -1\n",
    "# for i in range(100):\n",
    "#     if rm_list_i == rm_list_j:\n",
    "#         break\n",
    "#     else:\n",
    "#         rm_list = [key for key in kegg_connections.keys() if [e for e in kegg_connections[key] if e not in rm_list]\n",
    "#      ==[]]\n",
    "#         rm_list_j = rm_list_i \n",
    "#         rm_list_i = len(rm_list)\n",
    "# rm_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "057b2e76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for key in rm_list:\n",
    "#     del kegg_connections[key]\n",
    "    \n",
    "# for key in kegg_connections.keys():\n",
    "#     kegg_connections[key] = [e for e in kegg_connections[key] if e not in rm_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "65a59244",
   "metadata": {},
   "outputs": [],
   "source": [
    "def kegg_connections_append_y_hat(kegg_connections):\n",
    "    # add yhat node to the graph\n",
    "    temp_values = []\n",
    "    for key in kegg_connections.keys():\n",
    "        temp_values += kegg_connections[key]\n",
    "\n",
    "    kegg_connections['y_hat'] = [key for key in kegg_connections.keys() if key not in temp_values]\n",
    "    return kegg_connections\n",
    "\n",
    "# kegg_connections_append_y_hat(\n",
    "#     kegg_connections = kegg_connections_clean(\n",
    "#         kegg_connections = kegg_connections_build(\n",
    "#             n_genes = 6, kegg_gene_brite = kegg_gene_brite) ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "f9c8cc67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # add yhat node to the graph\n",
    "# temp_values = []\n",
    "# for key in kegg_connections.keys():\n",
    "#     temp_values += kegg_connections[key]\n",
    "\n",
    "# kegg_connections['y_hat'] = [key for key in kegg_connections.keys() if key not in temp_values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "fb52af7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6/6 [00:00<00:00, 39693.73it/s]\n"
     ]
    }
   ],
   "source": [
    "kegg_connections = kegg_connections_append_y_hat(\n",
    "    kegg_connections = kegg_connections_clean(\n",
    "        kegg_connections = kegg_connections_build(\n",
    "            n_genes = 6, kegg_gene_brite = kegg_gene_brite) ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "20f39cb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 46/46 [00:00<00:00, 180822.85it/s]\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (0)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"1253pt\" height=\"476pt\"\n",
       " viewBox=\"0.00 0.00 1252.69 476.00\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 472)\">\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-472 1248.69,-472 1248.69,4 -4,4\"/>\n",
       "<!-- 0 -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>0</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"675\" cy=\"-162\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"675\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">0</text>\n",
       "</g>\n",
       "<!-- 45 -->\n",
       "<g id=\"node52\" class=\"node\">\n",
       "<title>45</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"675\" cy=\"-18\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"675\" y=\"-14.3\" font-family=\"Times,serif\" font-size=\"14.00\">45</text>\n",
       "</g>\n",
       "<!-- 0&#45;&gt;45 -->\n",
       "<g id=\"edge59\" class=\"edge\">\n",
       "<title>0&#45;&gt;45</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M675,-143.87C675,-119.67 675,-75.21 675,-46.39\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"678.5,-46.19 675,-36.19 671.5,-46.19 678.5,-46.19\"/>\n",
       "</g>\n",
       "<!-- 1 -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>1</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"603\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"603\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">1</text>\n",
       "</g>\n",
       "<!-- 1&#45;&gt;0 -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>1&#45;&gt;0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M617.57,-218.83C627.75,-208.94 641.52,-195.55 653.03,-184.36\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"655.47,-186.87 660.2,-177.38 650.59,-181.85 655.47,-186.87\"/>\n",
       "</g>\n",
       "<!-- 17 -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>17</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"937\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"937\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">17</text>\n",
       "</g>\n",
       "<!-- 17&#45;&gt;0 -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>17&#45;&gt;0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M912.35,-226.41C865.44,-213.88 763.22,-186.57 709.52,-172.22\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"710.34,-168.82 699.77,-169.62 708.53,-175.58 710.34,-168.82\"/>\n",
       "</g>\n",
       "<!-- 24 -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>24</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"675\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"675\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">24</text>\n",
       "</g>\n",
       "<!-- 24&#45;&gt;0 -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>24&#45;&gt;0</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M675,-215.7C675,-207.98 675,-198.71 675,-190.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"678.5,-190.1 675,-180.1 671.5,-190.1 678.5,-190.1\"/>\n",
       "</g>\n",
       "<!-- 2 -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>2</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"374\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"374\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">2</text>\n",
       "</g>\n",
       "<!-- 2&#45;&gt;1 -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>2&#45;&gt;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M397.29,-296.44C405.38,-293.6 414.54,-290.52 423,-288 486.23,-269.17 504.87,-274.19 567,-252 568.86,-251.34 570.74,-250.61 572.63,-249.84\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"574.2,-252.98 581.97,-245.77 571.41,-246.56 574.2,-252.98\"/>\n",
       "</g>\n",
       "<!-- 36 -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>36</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"747\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"747\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">36</text>\n",
       "</g>\n",
       "<!-- 36&#45;&gt;1 -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>36&#45;&gt;1</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M725.75,-294.67C701.4,-282.83 661.28,-263.33 633.57,-249.86\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"634.86,-246.6 624.33,-245.37 631.8,-252.89 634.86,-246.6\"/>\n",
       "</g>\n",
       "<!-- 3 -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>3</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"243\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"243\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">3</text>\n",
       "</g>\n",
       "<!-- 3&#45;&gt;2 -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>3&#45;&gt;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M263.47,-366.06C285.29,-354.4 320.04,-335.83 344.75,-322.63\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"346.5,-325.66 353.67,-317.86 343.21,-319.49 346.5,-325.66\"/>\n",
       "</g>\n",
       "<!-- 4 -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>4</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"315\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"315\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">4</text>\n",
       "</g>\n",
       "<!-- 4&#45;&gt;2 -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>4&#45;&gt;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M327.8,-361.81C335.61,-352.55 345.75,-340.52 354.53,-330.09\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"357.45,-332.06 361.22,-322.16 352.1,-327.55 357.45,-332.06\"/>\n",
       "</g>\n",
       "<!-- 27 -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>27</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"387\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"387\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">27</text>\n",
       "</g>\n",
       "<!-- 27&#45;&gt;2 -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>27&#45;&gt;2</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M383.85,-360.05C382.42,-352.35 380.69,-343.03 379.08,-334.36\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"382.48,-333.47 377.21,-324.28 375.59,-334.75 382.48,-333.47\"/>\n",
       "</g>\n",
       "<!-- 100278565 -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>100278565</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"171\" cy=\"-450\" rx=\"63.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"171\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">100278565</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;3 -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;3</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M188.06,-432.41C197.83,-422.91 210.21,-410.88 220.74,-400.65\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"223.28,-403.05 228.01,-393.57 218.4,-398.03 223.28,-403.05\"/>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;4 -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;4</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M201.54,-434.15C226.03,-422.25 260.22,-405.63 284.66,-393.75\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"286.37,-396.81 293.83,-389.29 283.31,-390.52 286.37,-396.81\"/>\n",
       "</g>\n",
       "<!-- 9 -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>9</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"27\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"27\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">9</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;9 -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;9</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M140.46,-434.15C115.97,-422.25 81.78,-405.63 57.34,-393.75\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"58.69,-390.52 48.17,-389.29 55.63,-396.81 58.69,-390.52\"/>\n",
       "</g>\n",
       "<!-- 13 -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>13</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"99\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"99\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">13</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;13 -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;13</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M153.94,-432.41C144.17,-422.91 131.79,-410.88 121.26,-400.65\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"123.6,-398.03 113.99,-393.57 118.72,-403.05 123.6,-398.03\"/>\n",
       "</g>\n",
       "<!-- 16 -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>16</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"171\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"171\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">16</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;16 -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;16</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M171,-431.7C171,-423.98 171,-414.71 171,-406.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"174.5,-406.1 171,-396.1 167.5,-406.1 174.5,-406.1\"/>\n",
       "</g>\n",
       "<!-- 5 -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>5</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"730\" cy=\"-90\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"730\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">5</text>\n",
       "</g>\n",
       "<!-- 5&#45;&gt;45 -->\n",
       "<g id=\"edge60\" class=\"edge\">\n",
       "<title>5&#45;&gt;45</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M718.07,-73.81C710.79,-64.55 701.34,-52.52 693.15,-42.09\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"695.84,-39.86 686.91,-34.16 690.34,-44.18 695.84,-39.86\"/>\n",
       "</g>\n",
       "<!-- 6 -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>6</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"324\" cy=\"-162\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"324\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">6</text>\n",
       "</g>\n",
       "<!-- 6&#45;&gt;5 -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>6&#45;&gt;5</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M350.29,-157.04C417.97,-146.8 596.99,-119.48 656,-108 668.84,-105.5 682.84,-102.39 695.18,-99.51\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"696.16,-102.87 705.08,-97.17 694.55,-96.06 696.16,-102.87\"/>\n",
       "</g>\n",
       "<!-- 20 -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>20</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"967\" cy=\"-162\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"967\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">20</text>\n",
       "</g>\n",
       "<!-- 20&#45;&gt;5 -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>20&#45;&gt;5</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M942.53,-153.77C900.02,-141.22 812.45,-115.35 764.04,-101.05\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"764.92,-97.66 754.34,-98.19 762.94,-104.38 764.92,-97.66\"/>\n",
       "</g>\n",
       "<!-- 7 -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>7</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"240\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"240\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">7</text>\n",
       "</g>\n",
       "<!-- 7&#45;&gt;6 -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>7&#45;&gt;6</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M256.2,-219.5C268.63,-209.14 285.99,-194.67 300.03,-182.97\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"302.41,-185.55 307.85,-176.46 297.93,-180.17 302.41,-185.55\"/>\n",
       "</g>\n",
       "<!-- 8 -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>8</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"41\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"41\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">8</text>\n",
       "</g>\n",
       "<!-- 8&#45;&gt;7 -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>8&#45;&gt;7</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M64.61,-296.7C100.01,-284.24 166.97,-260.69 207.13,-246.56\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"208.46,-249.81 216.73,-243.19 206.13,-243.2 208.46,-249.81\"/>\n",
       "</g>\n",
       "<!-- 9&#45;&gt;8 -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>9&#45;&gt;8</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M30.39,-360.05C31.93,-352.35 33.79,-343.03 35.53,-334.36\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"39.02,-334.77 37.54,-324.28 32.15,-333.39 39.02,-334.77\"/>\n",
       "</g>\n",
       "<!-- 10 -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>10</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"548\" cy=\"-90\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"548\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">10</text>\n",
       "</g>\n",
       "<!-- 10&#45;&gt;45 -->\n",
       "<g id=\"edge61\" class=\"edge\">\n",
       "<title>10&#45;&gt;45</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M568.12,-77.91C588.99,-66.41 621.84,-48.3 645.61,-35.2\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"647.48,-38.16 654.55,-30.27 644.1,-32.03 647.48,-38.16\"/>\n",
       "</g>\n",
       "<!-- 11 -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>11</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"476\" cy=\"-162\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"476\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">11</text>\n",
       "</g>\n",
       "<!-- 11&#45;&gt;10 -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>11&#45;&gt;10</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M490.57,-146.83C500.75,-136.94 514.52,-123.55 526.03,-112.36\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"528.47,-114.87 533.2,-105.38 523.59,-109.85 528.47,-114.87\"/>\n",
       "</g>\n",
       "<!-- 14 -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>14</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"620\" cy=\"-90\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"620\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">14</text>\n",
       "</g>\n",
       "<!-- 11&#45;&gt;14 -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>11&#45;&gt;14</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M497.25,-150.67C521.6,-138.83 561.72,-119.33 589.43,-105.86\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"591.2,-108.89 598.67,-101.37 588.14,-102.6 591.2,-108.89\"/>\n",
       "</g>\n",
       "<!-- 28 -->\n",
       "<g id=\"node37\" class=\"node\">\n",
       "<title>28</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"476\" cy=\"-90\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"476\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">28</text>\n",
       "</g>\n",
       "<!-- 11&#45;&gt;28 -->\n",
       "<g id=\"edge41\" class=\"edge\">\n",
       "<title>11&#45;&gt;28</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M476,-143.7C476,-135.98 476,-126.71 476,-118.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"479.5,-118.1 476,-108.1 472.5,-118.1 479.5,-118.1\"/>\n",
       "</g>\n",
       "<!-- 12 -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>12</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"314\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"314\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">12</text>\n",
       "</g>\n",
       "<!-- 12&#45;&gt;11 -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>12&#45;&gt;11</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M335.47,-223C340.55,-220.68 345.96,-218.23 351,-216 382.41,-202.13 418.54,-186.88 443.86,-176.32\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"445.45,-179.45 453.33,-172.37 442.75,-172.99 445.45,-179.45\"/>\n",
       "</g>\n",
       "<!-- 15 -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>15</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"387\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"387\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">15</text>\n",
       "</g>\n",
       "<!-- 15&#45;&gt;11 -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>15&#45;&gt;11</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M403.74,-219.83C417.08,-209.34 435.96,-194.49 451.08,-182.6\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"453.34,-185.28 459.03,-176.35 449.01,-179.78 453.34,-185.28\"/>\n",
       "</g>\n",
       "<!-- 29 -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>29</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"459\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"459\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">29</text>\n",
       "</g>\n",
       "<!-- 29&#45;&gt;11 -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>29&#45;&gt;11</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M463.12,-216.05C465.01,-208.26 467.3,-198.82 469.42,-190.08\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"472.85,-190.82 471.8,-180.28 466.04,-189.17 472.85,-190.82\"/>\n",
       "</g>\n",
       "<!-- 32 -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>32</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"531\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"531\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">32</text>\n",
       "</g>\n",
       "<!-- 32&#45;&gt;11 -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>32&#45;&gt;11</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M519.07,-217.81C511.79,-208.55 502.34,-196.52 494.15,-186.09\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"496.84,-183.86 487.91,-178.16 491.34,-188.18 496.84,-183.86\"/>\n",
       "</g>\n",
       "<!-- 13&#45;&gt;12 -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>13&#45;&gt;12</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M117.7,-364.65C155.84,-339.46 242.71,-282.09 286.98,-252.85\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"289.08,-255.66 295.49,-247.22 285.22,-249.81 289.08,-255.66\"/>\n",
       "</g>\n",
       "<!-- 14&#45;&gt;45 -->\n",
       "<g id=\"edge62\" class=\"edge\">\n",
       "<title>14&#45;&gt;45</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M631.93,-73.81C639.21,-64.55 648.66,-52.52 656.85,-42.09\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"659.66,-44.18 663.09,-34.16 654.16,-39.86 659.66,-44.18\"/>\n",
       "</g>\n",
       "<!-- 16&#45;&gt;15 -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>16&#45;&gt;15</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M189.79,-364.65C228.1,-339.46 315.38,-282.09 359.85,-252.85\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"361.97,-255.64 368.41,-247.22 358.13,-249.79 361.97,-255.64\"/>\n",
       "</g>\n",
       "<!-- 18 -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>18</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1017\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1017\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">18</text>\n",
       "</g>\n",
       "<!-- 18&#45;&gt;17 -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>18&#45;&gt;17</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1001.19,-291.17C989.58,-281.01 973.61,-267.03 960.5,-255.56\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"962.6,-252.75 952.77,-248.8 957.99,-258.02 962.6,-252.75\"/>\n",
       "</g>\n",
       "<!-- 19 -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>19</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1036\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1036\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">19</text>\n",
       "</g>\n",
       "<!-- 19&#45;&gt;18 -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>19&#45;&gt;18</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1031.4,-360.05C1029.25,-352.14 1026.65,-342.54 1024.24,-333.69\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1027.55,-332.52 1021.56,-323.79 1020.8,-334.36 1027.55,-332.52\"/>\n",
       "</g>\n",
       "<!-- 100383860 -->\n",
       "<g id=\"node28\" class=\"node\">\n",
       "<title>100383860</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1036\" cy=\"-450\" rx=\"63.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1036\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">100383860</text>\n",
       "</g>\n",
       "<!-- 100383860&#45;&gt;19 -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>100383860&#45;&gt;19</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1036,-431.7C1036,-423.98 1036,-414.71 1036,-406.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1039.5,-406.1 1036,-396.1 1032.5,-406.1 1039.5,-406.1\"/>\n",
       "</g>\n",
       "<!-- 23 -->\n",
       "<g id=\"node32\" class=\"node\">\n",
       "<title>23</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1108\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1108\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">23</text>\n",
       "</g>\n",
       "<!-- 100383860&#45;&gt;23 -->\n",
       "<g id=\"edge34\" class=\"edge\">\n",
       "<title>100383860&#45;&gt;23</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1053.06,-432.41C1062.83,-422.91 1075.21,-410.88 1085.74,-400.65\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1088.28,-403.05 1093.01,-393.57 1083.4,-398.03 1088.28,-403.05\"/>\n",
       "</g>\n",
       "<!-- 103630585 -->\n",
       "<g id=\"node29\" class=\"node\">\n",
       "<title>103630585</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1181\" cy=\"-450\" rx=\"63.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1181\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">103630585</text>\n",
       "</g>\n",
       "<!-- 103630585&#45;&gt;19 -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>103630585&#45;&gt;19</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1150.25,-434.15C1125.59,-422.25 1091.16,-405.63 1066.55,-393.75\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1067.84,-390.49 1057.32,-389.29 1064.8,-396.79 1067.84,-390.49\"/>\n",
       "</g>\n",
       "<!-- 103630585&#45;&gt;23 -->\n",
       "<g id=\"edge35\" class=\"edge\">\n",
       "<title>103630585&#45;&gt;23</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1163.7,-432.41C1153.66,-422.78 1140.9,-410.55 1130.14,-400.23\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1132.35,-397.5 1122.71,-393.11 1127.51,-402.56 1132.35,-397.5\"/>\n",
       "</g>\n",
       "<!-- 21 -->\n",
       "<g id=\"node30\" class=\"node\">\n",
       "<title>21</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1022\" cy=\"-234\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1022\" y=\"-230.3\" font-family=\"Times,serif\" font-size=\"14.00\">21</text>\n",
       "</g>\n",
       "<!-- 21&#45;&gt;20 -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>21&#45;&gt;20</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1010.07,-217.81C1002.79,-208.55 993.34,-196.52 985.15,-186.09\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"987.84,-183.86 978.91,-178.16 982.34,-188.18 987.84,-183.86\"/>\n",
       "</g>\n",
       "<!-- 22 -->\n",
       "<g id=\"node31\" class=\"node\">\n",
       "<title>22</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1098\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"1098\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">22</text>\n",
       "</g>\n",
       "<!-- 22&#45;&gt;21 -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>22&#45;&gt;21</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1082.62,-290.83C1071.8,-280.87 1057.14,-267.37 1044.95,-256.14\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1047.08,-253.34 1037.36,-249.14 1042.34,-258.49 1047.08,-253.34\"/>\n",
       "</g>\n",
       "<!-- 23&#45;&gt;22 -->\n",
       "<g id=\"edge33\" class=\"edge\">\n",
       "<title>23&#45;&gt;22</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1105.58,-360.05C1104.48,-352.35 1103.15,-343.03 1101.91,-334.36\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1105.35,-333.68 1100.47,-324.28 1098.42,-334.67 1105.35,-333.68\"/>\n",
       "</g>\n",
       "<!-- 25 -->\n",
       "<g id=\"node33\" class=\"node\">\n",
       "<title>25</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"675\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"675\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">25</text>\n",
       "</g>\n",
       "<!-- 25&#45;&gt;24 -->\n",
       "<g id=\"edge36\" class=\"edge\">\n",
       "<title>25&#45;&gt;24</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M675,-287.7C675,-279.98 675,-270.71 675,-262.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"678.5,-262.1 675,-252.1 671.5,-262.1 678.5,-262.1\"/>\n",
       "</g>\n",
       "<!-- 26 -->\n",
       "<g id=\"node34\" class=\"node\">\n",
       "<title>26</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"603\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"603\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">26</text>\n",
       "</g>\n",
       "<!-- 26&#45;&gt;25 -->\n",
       "<g id=\"edge37\" class=\"edge\">\n",
       "<title>26&#45;&gt;25</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M617.57,-362.83C627.75,-352.94 641.52,-339.55 653.03,-328.36\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"655.47,-330.87 660.2,-321.38 650.59,-325.85 655.47,-330.87\"/>\n",
       "</g>\n",
       "<!-- 35 -->\n",
       "<g id=\"node35\" class=\"node\">\n",
       "<title>35</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"675\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"675\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">35</text>\n",
       "</g>\n",
       "<!-- 35&#45;&gt;25 -->\n",
       "<g id=\"edge38\" class=\"edge\">\n",
       "<title>35&#45;&gt;25</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M675,-359.7C675,-351.98 675,-342.71 675,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"678.5,-334.1 675,-324.1 671.5,-334.1 678.5,-334.1\"/>\n",
       "</g>\n",
       "<!-- 100383837 -->\n",
       "<g id=\"node36\" class=\"node\">\n",
       "<title>100383837</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"495\" cy=\"-450\" rx=\"63.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"495\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">100383837</text>\n",
       "</g>\n",
       "<!-- 100383837&#45;&gt;27 -->\n",
       "<g id=\"edge40\" class=\"edge\">\n",
       "<title>100383837&#45;&gt;27</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M470.76,-433.29C454.03,-422.44 431.75,-408 414.37,-396.74\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"416.08,-393.68 405.79,-391.18 412.27,-399.55 416.08,-393.68\"/>\n",
       "</g>\n",
       "<!-- 100383837&#45;&gt;26 -->\n",
       "<g id=\"edge39\" class=\"edge\">\n",
       "<title>100383837&#45;&gt;26</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M519.24,-433.29C535.97,-422.44 558.25,-408 575.63,-396.74\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"577.73,-399.55 584.21,-391.18 573.92,-393.68 577.73,-399.55\"/>\n",
       "</g>\n",
       "<!-- 31 -->\n",
       "<g id=\"node39\" class=\"node\">\n",
       "<title>31</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"459\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"459\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">31</text>\n",
       "</g>\n",
       "<!-- 100383837&#45;&gt;31 -->\n",
       "<g id=\"edge44\" class=\"edge\">\n",
       "<title>100383837&#45;&gt;31</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M486.29,-432.05C482.02,-423.77 476.8,-413.62 472.08,-404.42\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"475.07,-402.6 467.39,-395.31 468.85,-405.8 475.07,-402.6\"/>\n",
       "</g>\n",
       "<!-- 34 -->\n",
       "<g id=\"node41\" class=\"node\">\n",
       "<title>34</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"531\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"531\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">34</text>\n",
       "</g>\n",
       "<!-- 100383837&#45;&gt;34 -->\n",
       "<g id=\"edge47\" class=\"edge\">\n",
       "<title>100383837&#45;&gt;34</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M503.71,-432.05C507.98,-423.77 513.2,-413.62 517.92,-404.42\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"521.15,-405.8 522.61,-395.31 514.93,-402.6 521.15,-405.8\"/>\n",
       "</g>\n",
       "<!-- 28&#45;&gt;45 -->\n",
       "<g id=\"edge63\" class=\"edge\">\n",
       "<title>28&#45;&gt;45</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M497.12,-78.48C501.96,-76.22 507.12,-73.94 512,-72 555.74,-54.64 607.64,-38.51 640.93,-28.7\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"642.01,-32.03 650.63,-25.87 640.05,-25.31 642.01,-32.03\"/>\n",
       "</g>\n",
       "<!-- 30 -->\n",
       "<g id=\"node38\" class=\"node\">\n",
       "<title>30</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"459\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"459\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">30</text>\n",
       "</g>\n",
       "<!-- 30&#45;&gt;29 -->\n",
       "<g id=\"edge42\" class=\"edge\">\n",
       "<title>30&#45;&gt;29</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M459,-287.7C459,-279.98 459,-270.71 459,-262.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"462.5,-262.1 459,-252.1 455.5,-262.1 462.5,-262.1\"/>\n",
       "</g>\n",
       "<!-- 31&#45;&gt;30 -->\n",
       "<g id=\"edge43\" class=\"edge\">\n",
       "<title>31&#45;&gt;30</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M459,-359.7C459,-351.98 459,-342.71 459,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"462.5,-334.1 459,-324.1 455.5,-334.1 462.5,-334.1\"/>\n",
       "</g>\n",
       "<!-- 33 -->\n",
       "<g id=\"node40\" class=\"node\">\n",
       "<title>33</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"531\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"531\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">33</text>\n",
       "</g>\n",
       "<!-- 33&#45;&gt;32 -->\n",
       "<g id=\"edge45\" class=\"edge\">\n",
       "<title>33&#45;&gt;32</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M531,-287.7C531,-279.98 531,-270.71 531,-262.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"534.5,-262.1 531,-252.1 527.5,-262.1 534.5,-262.1\"/>\n",
       "</g>\n",
       "<!-- 34&#45;&gt;33 -->\n",
       "<g id=\"edge46\" class=\"edge\">\n",
       "<title>34&#45;&gt;33</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M531,-359.7C531,-351.98 531,-342.71 531,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"534.5,-334.1 531,-324.1 527.5,-334.1 534.5,-334.1\"/>\n",
       "</g>\n",
       "<!-- 100191673 -->\n",
       "<g id=\"node42\" class=\"node\">\n",
       "<title>100191673</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"746\" cy=\"-450\" rx=\"63.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"746\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">100191673</text>\n",
       "</g>\n",
       "<!-- 100191673&#45;&gt;35 -->\n",
       "<g id=\"edge48\" class=\"edge\">\n",
       "<title>100191673&#45;&gt;35</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M729.17,-432.41C719.54,-422.91 707.34,-410.88 696.95,-400.65\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"699.36,-398.1 689.78,-393.57 694.44,-403.08 699.36,-398.1\"/>\n",
       "</g>\n",
       "<!-- 37 -->\n",
       "<g id=\"node43\" class=\"node\">\n",
       "<title>37</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"747\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"747\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">37</text>\n",
       "</g>\n",
       "<!-- 100191673&#45;&gt;37 -->\n",
       "<g id=\"edge51\" class=\"edge\">\n",
       "<title>100191673&#45;&gt;37</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M746.25,-431.7C746.36,-423.98 746.49,-414.71 746.61,-406.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"750.11,-406.15 746.76,-396.1 743.11,-406.05 750.11,-406.15\"/>\n",
       "</g>\n",
       "<!-- 40 -->\n",
       "<g id=\"node47\" class=\"node\">\n",
       "<title>40</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"819\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"819\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">40</text>\n",
       "</g>\n",
       "<!-- 100191673&#45;&gt;40 -->\n",
       "<g id=\"edge54\" class=\"edge\">\n",
       "<title>100191673&#45;&gt;40</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M763.3,-432.41C773.34,-422.78 786.1,-410.55 796.86,-400.23\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"799.49,-402.56 804.29,-393.11 794.65,-397.5 799.49,-402.56\"/>\n",
       "</g>\n",
       "<!-- 37&#45;&gt;36 -->\n",
       "<g id=\"edge49\" class=\"edge\">\n",
       "<title>37&#45;&gt;36</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M747,-359.7C747,-351.98 747,-342.71 747,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"750.5,-334.1 747,-324.1 743.5,-334.1 750.5,-334.1\"/>\n",
       "</g>\n",
       "<!-- 41 -->\n",
       "<g id=\"node44\" class=\"node\">\n",
       "<title>41</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"891\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"891\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">41</text>\n",
       "</g>\n",
       "<!-- 41&#45;&gt;36 -->\n",
       "<g id=\"edge50\" class=\"edge\">\n",
       "<title>41&#45;&gt;36</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M869.75,-366.67C845.4,-354.83 805.28,-335.33 777.57,-321.86\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"778.86,-318.6 768.33,-317.37 775.8,-324.89 778.86,-318.6\"/>\n",
       "</g>\n",
       "<!-- 38 -->\n",
       "<g id=\"node45\" class=\"node\">\n",
       "<title>38</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"798\" cy=\"-162\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"798\" y=\"-158.3\" font-family=\"Times,serif\" font-size=\"14.00\">38</text>\n",
       "</g>\n",
       "<!-- 38&#45;&gt;45 -->\n",
       "<g id=\"edge64\" class=\"edge\">\n",
       "<title>38&#45;&gt;45</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M795.45,-143.85C791.82,-124.53 783.54,-93.07 766,-72 750.59,-53.49 726.94,-39.99 707.61,-31.26\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"708.82,-27.96 698.25,-27.23 706.05,-34.39 708.82,-27.96\"/>\n",
       "</g>\n",
       "<!-- 39 -->\n",
       "<g id=\"node46\" class=\"node\">\n",
       "<title>39</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"819\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"819\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">39</text>\n",
       "</g>\n",
       "<!-- 39&#45;&gt;38 -->\n",
       "<g id=\"edge52\" class=\"edge\">\n",
       "<title>39&#45;&gt;38</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M816.47,-287.87C812.89,-263.67 806.31,-219.21 802.05,-190.39\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"805.47,-189.57 800.54,-180.19 798.54,-190.59 805.47,-189.57\"/>\n",
       "</g>\n",
       "<!-- 40&#45;&gt;39 -->\n",
       "<g id=\"edge53\" class=\"edge\">\n",
       "<title>40&#45;&gt;39</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M819,-359.7C819,-351.98 819,-342.71 819,-334.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"822.5,-334.1 819,-324.1 815.5,-334.1 822.5,-334.1\"/>\n",
       "</g>\n",
       "<!-- 100275685 -->\n",
       "<g id=\"node48\" class=\"node\">\n",
       "<title>100275685</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"891\" cy=\"-450\" rx=\"63.89\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"891\" y=\"-446.3\" font-family=\"Times,serif\" font-size=\"14.00\">100275685</text>\n",
       "</g>\n",
       "<!-- 100275685&#45;&gt;41 -->\n",
       "<g id=\"edge55\" class=\"edge\">\n",
       "<title>100275685&#45;&gt;41</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M891,-431.7C891,-423.98 891,-414.71 891,-406.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"894.5,-406.1 891,-396.1 887.5,-406.1 894.5,-406.1\"/>\n",
       "</g>\n",
       "<!-- 44 -->\n",
       "<g id=\"node51\" class=\"node\">\n",
       "<title>44</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"963\" cy=\"-378\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"963\" y=\"-374.3\" font-family=\"Times,serif\" font-size=\"14.00\">44</text>\n",
       "</g>\n",
       "<!-- 100275685&#45;&gt;44 -->\n",
       "<g id=\"edge58\" class=\"edge\">\n",
       "<title>100275685&#45;&gt;44</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M908.06,-432.41C917.83,-422.91 930.21,-410.88 940.74,-400.65\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"943.28,-403.05 948.01,-393.57 938.4,-398.03 943.28,-403.05\"/>\n",
       "</g>\n",
       "<!-- 42 -->\n",
       "<g id=\"node49\" class=\"node\">\n",
       "<title>42</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"851\" cy=\"-90\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"851\" y=\"-86.3\" font-family=\"Times,serif\" font-size=\"14.00\">42</text>\n",
       "</g>\n",
       "<!-- 42&#45;&gt;45 -->\n",
       "<g id=\"edge65\" class=\"edge\">\n",
       "<title>42&#45;&gt;45</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M828.35,-79.99C797.46,-67.71 741.96,-45.63 706.88,-31.68\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"708.15,-28.42 697.56,-27.97 705.56,-34.92 708.15,-28.42\"/>\n",
       "</g>\n",
       "<!-- 43 -->\n",
       "<g id=\"node50\" class=\"node\">\n",
       "<title>43</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"927\" cy=\"-306\" rx=\"27\" ry=\"18\"/>\n",
       "<text text-anchor=\"middle\" x=\"927\" y=\"-302.3\" font-family=\"Times,serif\" font-size=\"14.00\">43</text>\n",
       "</g>\n",
       "<!-- 43&#45;&gt;42 -->\n",
       "<g id=\"edge56\" class=\"edge\">\n",
       "<title>43&#45;&gt;42</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M918.13,-288.59C912.76,-278.27 906,-264.57 901,-252 882.73,-206.06 866.8,-150.66 858.08,-118.27\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"861.36,-116.99 855.4,-108.22 854.59,-118.79 861.36,-116.99\"/>\n",
       "</g>\n",
       "<!-- 44&#45;&gt;43 -->\n",
       "<g id=\"edge57\" class=\"edge\">\n",
       "<title>44&#45;&gt;43</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M954.65,-360.76C950.29,-352.28 944.85,-341.71 939.96,-332.2\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"942.99,-330.44 935.3,-323.15 936.77,-333.64 942.99,-330.44\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x7f04b49eb750>"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def kegg_connections_digraph(kegg_connections, option = ''):\n",
    "    dot = ''\n",
    "    if option == '':\n",
    "        dot = Digraph()\n",
    "        for key in tqdm(kegg_connections.keys()):\n",
    "            dot.node(key)\n",
    "            for value in kegg_connections[key]:\n",
    "                # edge takes a head/tail whereas edges takes name pairs concatednated (A, B -> AB)in a list\n",
    "                dot.edge(value, key)\n",
    "\n",
    "    if option == 'number':\n",
    "        name_to_num_dict = dict(zip(list(kegg_connections.keys()),\n",
    "                                    [str(i) for i in range(len(list(kegg_connections.keys())))]))\n",
    "\n",
    "        temp = {}\n",
    "        for key in kegg_connections.keys():\n",
    "            temp[name_to_num_dict[key]] = [name_to_num_dict[e] if e in name_to_num_dict.keys() else e for e in kegg_connections[key]]\n",
    "\n",
    "        dot = Digraph()\n",
    "        for key in tqdm(temp.keys()):\n",
    "            dot.node(key)\n",
    "            for value in temp[key]:\n",
    "                # edge takes a head/tail whereas edges takes name pairs concatednated (A, B -> AB)in a list\n",
    "                dot.edge(value, key)                 \n",
    "                 \n",
    "    return dot\n",
    "\n",
    "kegg_connections_digraph(kegg_connections, option = 'number')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "597288ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # This is too big to render in full\n",
    "# dot = ''\n",
    "# if vis_dot_bool:\n",
    "#     dot = Digraph()\n",
    "#     for key in tqdm(kegg_connections.keys()):\n",
    "#         dot.node(key)\n",
    "#         for value in kegg_connections[key]:\n",
    "#             # edge takes a head/tail whereas edges takes name pairs concatednated (A, B -> AB)in a list\n",
    "#             dot.edge(value, key)    \n",
    "\n",
    "# dot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f1e6077",
   "metadata": {},
   "source": [
    "Version with the node names masked for size "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "e8ddd86c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dot = ''\n",
    "# if vis_dot_bool:\n",
    "#     name_to_num_dict = dict(zip(list(kegg_connections.keys()),\n",
    "#                                 [str(i) for i in range(len(list(kegg_connections.keys())))]))\n",
    "\n",
    "#     temp = {}\n",
    "#     for key in kegg_connections.keys():\n",
    "#         temp[name_to_num_dict[key]] = [name_to_num_dict[e] if e in name_to_num_dict.keys() else e for e in kegg_connections[key]]\n",
    "\n",
    "#     dot = Digraph()\n",
    "#     for key in tqdm(temp.keys()):\n",
    "#         dot.node(key)\n",
    "#         for value in temp[key]:\n",
    "#             # edge takes a head/tail whereas edges takes name pairs concatednated (A, B -> AB)in a list\n",
    "#             dot.edge(value, key)    \n",
    "\n",
    "#     # dot.render(directory=cache_path, view=True) \n",
    "# dot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1cb7324",
   "metadata": {},
   "source": [
    "### Setup to build the graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "301ac2d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6/6 [00:00<00:00, 44073.25it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "46"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_connections = kegg_connections_append_y_hat(\n",
    "    kegg_connections = kegg_connections_clean(\n",
    "        kegg_connections = kegg_connections_build(\n",
    "            n_genes = 6, kegg_gene_brite = kegg_gene_brite) ) )\n",
    "\n",
    "len(list(kegg_connections.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "bbf22b5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# kegg_connections\n",
    "\n",
    "def kegg_connections_find_in_out_nodes(kegg_connections):\n",
    "    # start by finding the top level -- all those keys which are theselves not values\n",
    "    res = find_uniq_keys_values(input_dict = kegg_connections)\n",
    "    all_keys = res['all_keys']\n",
    "    all_values = res['all_values']\n",
    "\n",
    "    # use the keys to find the input/outputs of the graph\n",
    "    output_nodes = [e for e in all_keys if e not in all_values]\n",
    "    input_nodes = [e for e in all_values if e not in all_keys]\n",
    "    return input_nodes, output_nodes\n",
    "\n",
    "\n",
    "input_nodes, output_nodes = kegg_connections_find_in_out_nodes(kegg_connections = kegg_connections)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "6e3cd18b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def kegg_connections_find_dependancy_order(kegg_connections, input_nodes):\n",
    "    # find the dependancies for run order from many dependancies to none\n",
    "    temp = kegg_connections.copy()\n",
    "\n",
    "    no_dependants = find_input_nodes(all_key_value_dict = find_uniq_keys_values(input_dict = temp))\n",
    "    # first pass. Same as the output nodes identified above\n",
    "    dependancy_order = []\n",
    "    # Then iterate\n",
    "    for ith in range(100): #TODO <- this should be set as a input parameter\n",
    "        top_nodes = find_top_nodes(all_key_value_dict = find_uniq_keys_values(input_dict = temp))\n",
    "        if top_nodes == []:\n",
    "            break\n",
    "        else:\n",
    "            dependancy_order += top_nodes    \n",
    "            # remove nodes from the graph that are at the 'top' level and haven't already been removed\n",
    "            for key in [e for e in dependancy_order if e in temp.keys()]:\n",
    "                temp.pop(key)\n",
    "\n",
    "    # dependancy_order\n",
    "                \n",
    "\n",
    "    # reverse to get the order that the nodes should be called\n",
    "    dependancy_order.reverse()\n",
    "    # dependancy_order\n",
    "\n",
    "\n",
    "\n",
    "    # Trying out new approach: add a node for the input data tha will only flatten the input.\n",
    "    dependancy_order = input_nodes+dependancy_order\n",
    "\n",
    "    for key in input_nodes:\n",
    "        kegg_connections[key] = [] #[key] # needs to contain itself so the model's `get_input_node()` function works \n",
    "                                # or that function needs to change.\n",
    "    \n",
    "    return kegg_connections, dependancy_order, no_dependants\n",
    "\n",
    "\n",
    "kegg_connections, dependancy_order, no_dependants = kegg_connections_find_dependancy_order(kegg_connections= kegg_connections, input_nodes= input_nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "78b6c996",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 6067/6067 [00:00<00:00, 944679.90it/s]\n"
     ]
    }
   ],
   "source": [
    "def kegg_gene_brite_build_node_lookup_dict(kegg_gene_brite):\n",
    "    # build a dict to go from the node names in `no_dependants` to the list index in `ACGT_gene_slice_list`\n",
    "    brite_node_to_list_idx_dict = {}\n",
    "    for i in tqdm(range(len(kegg_gene_brite))):\n",
    "        brite_node_to_list_idx_dict[str(kegg_gene_brite[i]['BRITE']['BRITE_PATHS'][0][-1])] = i\n",
    "    return brite_node_to_list_idx_dict\n",
    "\n",
    "brite_node_to_list_idx_dict = kegg_gene_brite_build_node_lookup_dict(kegg_gene_brite = kegg_gene_brite)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "005ba49d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_input_tensor_dict(ACGT_gene_slice_list, no_dependants):    \n",
    "    input_tensor_dict = {}\n",
    "    for e in no_dependants:\n",
    "        input_tensor_dict[e] = ACGT_gene_slice_list[brite_node_to_list_idx_dict[e]]\n",
    "    \n",
    "    return input_tensor_dict\n",
    "\n",
    "input_tensor_dict = build_input_tensor_dict(ACGT_gene_slice_list=ACGT_gene_slice_list, no_dependants=no_dependants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "d97df836",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 52/52 [00:00<00:00, 724597.37it/s]\n",
      "100%|██████████| 52/52 [00:00<00:00, 207915.93it/s]\n"
     ]
    }
   ],
   "source": [
    "def setup_network_param_dicts(\n",
    "        kegg_connections,  # for input sizes\n",
    "        input_tensor_dict, # for input sizes\n",
    "        dependancy_order,\n",
    "        default_output_size,\n",
    "        default_dropout_pr,\n",
    "        default_block_reps,\n",
    "    ):\n",
    "    # Figure out expected input/output shapes\n",
    "    #==NOTE! This assumes only dense connections!==\n",
    "\n",
    "    # This could be replaced by a sort of \"distance from output\" measure\n",
    "    output_size_dict = dict(zip(dependancy_order, \n",
    "                            [default_output_size for i in range(len(dependancy_order))]))\n",
    "    output_size_dict['y_hat'] = 1 \n",
    "\n",
    "\n",
    "    # Setup dropout % dictionary\n",
    "    dropout_pr_dict = dict(zip(dependancy_order, \n",
    "                            [default_dropout_pr for i in range(len(dependancy_order))]))\n",
    "    dropout_pr_dict['y_hat'] = 0 # not required, output node is purely linear without dropout\n",
    "\n",
    "\n",
    "    # Setup replicates of layers dictionary\n",
    "    block_rep_dict = dict(zip(dependancy_order, \n",
    "                            [default_block_reps for i in range(len(dependancy_order))]))\n",
    "    block_rep_dict['y_hat'] = 1 # not required, output node is purely linear. Not a linear block\n",
    "\n",
    "    # output_size_dict\n",
    "    # dropout_pr_dict\n",
    "    # block_rep_dict\n",
    "\n",
    "    # CHANNEL AWARE VERSION -----------------------------------------------------------------------------------\n",
    "    input_size_dict = kegg_connections.copy()\n",
    "\n",
    "    # use the expected output sizes from `output_size_dict` to fill in the non-data sizes\n",
    "    tensor_ndim = len(input_tensor_dict[list(input_tensor_dict.keys())[0]].shape)\n",
    "    for e in tqdm(input_size_dict.keys()):\n",
    "        # overwrite named connections with the output size of those connections\n",
    "        # if the entry is in no_dependants it's data so it's size needs to be grabbed from the input_tensor_dict\n",
    "        \n",
    "        # is there no channel dim? (major/minor allele)\n",
    "        if 2 == tensor_ndim:\n",
    "            input_size_dict[e] = [\n",
    "                list(input_tensor_dict[ee].shape)[-1] # <- NOTE! THIS ASSUMES ONLY DENSE CONNECTIONS (i.e. only the 1st dim is needed)\n",
    "                if ee in no_dependants\n",
    "                else output_size_dict[ee] for ee in input_size_dict[e]]\n",
    "        elif 3 == tensor_ndim: # There is a channel dim\n",
    "            input_size_dict[e] = [\n",
    "                (list(input_tensor_dict[ee].shape)[1]*list(input_tensor_dict[ee].shape)[2]) # <- NOTE! THIS ASSUMES ONLY DENSE CONNECTIONS (i.e. only the 1st dim is needed)  \n",
    "                if ee in no_dependants\n",
    "                else output_size_dict[ee] for ee in input_size_dict[e]]\n",
    "\n",
    "    # Now walk over entries and overwrite with the sum of the inputs\n",
    "    for e in tqdm(input_size_dict.keys()):\n",
    "        input_size_dict[e] = np.sum(input_size_dict[e])\n",
    "\n",
    "    return input_size_dict, output_size_dict, dropout_pr_dict, block_rep_dict\n",
    "\n",
    "input_size_dict, output_size_dict, dropout_pr_dict, block_rep_dict = setup_network_param_dicts(\n",
    "        kegg_connections = kegg_connections,\n",
    "        input_tensor_dict = input_tensor_dict,\n",
    "        dependancy_order = dependancy_order,\n",
    "        default_output_size = default_output_size,\n",
    "        default_dropout_pr = default_dropout_pr,\n",
    "        default_block_reps = default_block_reps,\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "3547d984",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 52/52 [00:00<00:00, 48869.33it/s]\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       " \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<!-- Generated by graphviz version 2.50.0 (0)\n",
       " -->\n",
       "<!-- Pages: 1 -->\n",
       "<svg width=\"1798pt\" height=\"600pt\"\n",
       " viewBox=\"0.00 0.00 1797.75 600.18\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">\n",
       "<g id=\"graph0\" class=\"graph\" transform=\"scale(1 1) rotate(0) translate(4 596.18)\">\n",
       "<polygon fill=\"white\" stroke=\"transparent\" points=\"-4,4 -4,-596.18 1793.75,-596.18 1793.75,4 -4,4\"/>\n",
       "<!-- KeggOrthology(Ko)[Br&#45;Zma00001] -->\n",
       "<g id=\"node1\" class=\"node\">\n",
       "<title>KeggOrthology(Ko)[Br&#45;Zma00001]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1083.38\" cy=\"-206.35\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1083.38\" y=\"-210.15\" font-family=\"Times,serif\" font-size=\"14.00\">in: 150</text>\n",
       "<text text-anchor=\"middle\" x=\"1083.38\" y=\"-195.15\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- y_hat -->\n",
       "<g id=\"node52\" class=\"node\">\n",
       "<title>y_hat</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1083.38\" cy=\"-26.87\" rx=\"46.84\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1083.38\" y=\"-30.67\" font-family=\"Times,serif\" font-size=\"14.00\">in: 350</text>\n",
       "<text text-anchor=\"middle\" x=\"1083.38\" y=\"-15.67\" font-family=\"Times,serif\" font-size=\"14.00\">out: 1</text>\n",
       "</g>\n",
       "<!-- KeggOrthology(Ko)[Br&#45;Zma00001]&#45;&gt;y_hat -->\n",
       "<g id=\"edge59\" class=\"edge\">\n",
       "<title>KeggOrthology(Ko)[Br&#45;Zma00001]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1083.38,-179.16C1083.38,-148.67 1083.38,-98.11 1083.38,-63.92\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1086.88,-63.88 1083.38,-53.88 1079.88,-63.88 1086.88,-63.88\"/>\n",
       "</g>\n",
       "<!-- 09180BriteHierarchies -->\n",
       "<g id=\"node2\" class=\"node\">\n",
       "<title>09180BriteHierarchies</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"970.38\" cy=\"-296.09\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"970.38\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 100</text>\n",
       "<text text-anchor=\"middle\" x=\"970.38\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 09180BriteHierarchies&#45;&gt;KeggOrthology(Ko)[Br&#45;Zma00001] -->\n",
       "<g id=\"edge1\" class=\"edge\">\n",
       "<title>09180BriteHierarchies&#45;&gt;KeggOrthology(Ko)[Br&#45;Zma00001]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M997.73,-273.85C1012.86,-262.1 1031.83,-247.38 1048.07,-234.77\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1050.52,-237.29 1056.27,-228.39 1046.23,-231.76 1050.52,-237.29\"/>\n",
       "</g>\n",
       "<!-- 09100Metabolism -->\n",
       "<g id=\"node3\" class=\"node\">\n",
       "<title>09100Metabolism</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1468.38\" cy=\"-296.09\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1468.38\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"1468.38\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 09100Metabolism&#45;&gt;KeggOrthology(Ko)[Br&#45;Zma00001] -->\n",
       "<g id=\"edge2\" class=\"edge\">\n",
       "<title>09100Metabolism&#45;&gt;KeggOrthology(Ko)[Br&#45;Zma00001]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1424.96,-285.2C1354.1,-269.05 1213.8,-237.07 1136.89,-219.55\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1137.46,-216.09 1126.93,-217.28 1135.9,-222.91 1137.46,-216.09\"/>\n",
       "</g>\n",
       "<!-- 09120GeneticInformationProcessing -->\n",
       "<g id=\"node4\" class=\"node\">\n",
       "<title>09120GeneticInformationProcessing</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1083.38\" cy=\"-296.09\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1083.38\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"1083.38\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 09120GeneticInformationProcessing&#45;&gt;KeggOrthology(Ko)[Br&#45;Zma00001] -->\n",
       "<g id=\"edge3\" class=\"edge\">\n",
       "<title>09120GeneticInformationProcessing&#45;&gt;KeggOrthology(Ko)[Br&#45;Zma00001]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1083.38,-268.77C1083.38,-260.8 1083.38,-251.91 1083.38,-243.42\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1086.88,-243.35 1083.38,-233.35 1079.88,-243.35 1086.88,-243.35\"/>\n",
       "</g>\n",
       "<!-- 09182ProteinFamilies&#45;GeneticInformationProcessing -->\n",
       "<g id=\"node5\" class=\"node\">\n",
       "<title>09182ProteinFamilies&#45;GeneticInformationProcessing</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"555.38\" cy=\"-385.83\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"555.38\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 150</text>\n",
       "<text text-anchor=\"middle\" x=\"555.38\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 09182ProteinFamilies&#45;GeneticInformationProcessing&#45;&gt;09180BriteHierarchies -->\n",
       "<g id=\"edge4\" class=\"edge\">\n",
       "<title>09182ProteinFamilies&#45;GeneticInformationProcessing&#45;&gt;09180BriteHierarchies</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M598.27,-374.41C619.58,-369.41 645.76,-363.52 669.38,-358.96 768.81,-339.74 796.09,-347.36 894.38,-322.96 903.42,-320.72 912.92,-317.88 922.02,-314.92\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"923.33,-318.17 931.7,-311.67 921.11,-311.53 923.33,-318.17\"/>\n",
       "</g>\n",
       "<!-- 09183ProteinFamilies&#45;SignalingAndCellularProcesses -->\n",
       "<g id=\"node6\" class=\"node\">\n",
       "<title>09183ProteinFamilies&#45;SignalingAndCellularProcesses</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1177.38\" cy=\"-385.83\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1177.38\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 100</text>\n",
       "<text text-anchor=\"middle\" x=\"1177.38\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 09183ProteinFamilies&#45;SignalingAndCellularProcesses&#45;&gt;09180BriteHierarchies -->\n",
       "<g id=\"edge5\" class=\"edge\">\n",
       "<title>09183ProteinFamilies&#45;SignalingAndCellularProcesses&#45;&gt;09180BriteHierarchies</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1141.33,-368.1C1134.4,-364.99 1127.19,-361.82 1120.38,-358.96 1079.51,-341.79 1068.24,-340.13 1027.38,-322.96 1023.54,-321.35 1019.58,-319.64 1015.62,-317.91\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1016.98,-314.68 1006.42,-313.82 1014.14,-321.08 1016.98,-314.68\"/>\n",
       "</g>\n",
       "<!-- 03016TransferRnaBiogenesis[Br&#45;Zma03016] -->\n",
       "<g id=\"node7\" class=\"node\">\n",
       "<title>03016TransferRnaBiogenesis[Br&#45;Zma03016]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"386.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"386.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 44</text>\n",
       "<text text-anchor=\"middle\" x=\"386.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 03016TransferRnaBiogenesis[Br&#45;Zma03016]&#45;&gt;09182ProteinFamilies&#45;GeneticInformationProcessing -->\n",
       "<g id=\"edge6\" class=\"edge\">\n",
       "<title>03016TransferRnaBiogenesis[Br&#45;Zma03016]&#45;&gt;09182ProteinFamilies&#45;GeneticInformationProcessing</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M420.58,-456.82C447.03,-443.08 483.83,-423.97 512.3,-409.19\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"513.96,-412.27 521.23,-404.56 510.74,-406.06 513.96,-412.27\"/>\n",
       "</g>\n",
       "<!-- 03036ChromosomeAndAssociatedProteins[Br&#45;Zma03036] -->\n",
       "<g id=\"node8\" class=\"node\">\n",
       "<title>03036ChromosomeAndAssociatedProteins[Br&#45;Zma03036]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"499.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"499.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 44</text>\n",
       "<text text-anchor=\"middle\" x=\"499.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 03036ChromosomeAndAssociatedProteins[Br&#45;Zma03036]&#45;&gt;09182ProteinFamilies&#45;GeneticInformationProcessing -->\n",
       "<g id=\"edge7\" class=\"edge\">\n",
       "<title>03036ChromosomeAndAssociatedProteins[Br&#45;Zma03036]&#45;&gt;09182ProteinFamilies&#45;GeneticInformationProcessing</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M514.98,-450.13C520.93,-440.8 527.83,-429.99 534.22,-419.98\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"537.24,-421.76 539.67,-411.44 531.34,-417.99 537.24,-421.76\"/>\n",
       "</g>\n",
       "<!-- 03019MessengerRnaBiogenesis[Br&#45;Zma03019] -->\n",
       "<g id=\"node9\" class=\"node\">\n",
       "<title>03019MessengerRnaBiogenesis[Br&#45;Zma03019]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"612.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"612.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 4</text>\n",
       "<text text-anchor=\"middle\" x=\"612.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 03019MessengerRnaBiogenesis[Br&#45;Zma03019]&#45;&gt;09182ProteinFamilies&#45;GeneticInformationProcessing -->\n",
       "<g id=\"edge8\" class=\"edge\">\n",
       "<title>03019MessengerRnaBiogenesis[Br&#45;Zma03019]&#45;&gt;09182ProteinFamilies&#45;GeneticInformationProcessing</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M596.5,-450.13C590.43,-440.8 583.42,-429.99 576.91,-419.98\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"579.75,-417.92 571.37,-411.44 573.88,-421.74 579.75,-417.92\"/>\n",
       "</g>\n",
       "<!-- 100278565 -->\n",
       "<g id=\"node10\" class=\"node\">\n",
       "<title>100278565</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"273.38\" cy=\"-565.31\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"273.38\" y=\"-569.11\" font-family=\"Times,serif\" font-size=\"14.00\">in: 0.0</text>\n",
       "<text text-anchor=\"middle\" x=\"273.38\" y=\"-554.11\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;03016TransferRnaBiogenesis[Br&#45;Zma03016] -->\n",
       "<g id=\"edge9\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;03016TransferRnaBiogenesis[Br&#45;Zma03016]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M300.73,-543.07C315.86,-531.32 334.83,-516.6 351.07,-503.99\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"353.52,-506.51 359.27,-497.61 349.23,-500.98 353.52,-506.51\"/>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;03036ChromosomeAndAssociatedProteins[Br&#45;Zma03036] -->\n",
       "<g id=\"edge10\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;03036ChromosomeAndAssociatedProteins[Br&#45;Zma03036]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M313.31,-550.64C347.76,-538.63 398.71,-520.31 442.38,-502.44 446.28,-500.84 450.31,-499.14 454.34,-497.39\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"455.94,-500.51 463.68,-493.27 453.11,-494.1 455.94,-500.51\"/>\n",
       "</g>\n",
       "<!-- 2.1.1.34Trna(Guanosine18&#45;2&#39;&#45;O)&#45;Methyltransferase -->\n",
       "<g id=\"node16\" class=\"node\">\n",
       "<title>2.1.1.34Trna(Guanosine18&#45;2&#39;&#45;O)&#45;Methyltransferase</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"47.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"47.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 44</text>\n",
       "<text text-anchor=\"middle\" x=\"47.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;2.1.1.34Trna(Guanosine18&#45;2&#39;&#45;O)&#45;Methyltransferase -->\n",
       "<g id=\"edge16\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;2.1.1.34Trna(Guanosine18&#45;2&#39;&#45;O)&#45;Methyltransferase</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M233.44,-550.64C198.99,-538.63 148.04,-520.31 104.38,-502.44 100.47,-500.84 96.44,-499.14 92.41,-497.39\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"93.64,-494.1 83.08,-493.27 90.81,-500.51 93.64,-494.1\"/>\n",
       "</g>\n",
       "<!-- Methyltransferases -->\n",
       "<g id=\"node23\" class=\"node\">\n",
       "<title>Methyltransferases</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"160.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"160.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 44</text>\n",
       "<text text-anchor=\"middle\" x=\"160.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;Methyltransferases -->\n",
       "<g id=\"edge23\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;Methyltransferases</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M246.02,-543.07C230.89,-531.32 211.92,-516.6 195.69,-503.99\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"197.52,-500.98 187.48,-497.61 193.23,-506.51 197.52,-500.98\"/>\n",
       "</g>\n",
       "<!-- Risc(Rna&#45;InducedSilencingComplex) -->\n",
       "<g id=\"node25\" class=\"node\">\n",
       "<title>Risc(Rna&#45;InducedSilencingComplex)</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"273.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"273.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 44</text>\n",
       "<text text-anchor=\"middle\" x=\"273.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100278565&#45;&gt;Risc(Rna&#45;InducedSilencingComplex) -->\n",
       "<g id=\"edge26\" class=\"edge\">\n",
       "<title>100278565&#45;&gt;Risc(Rna&#45;InducedSilencingComplex)</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M273.38,-537.99C273.38,-530.02 273.38,-521.13 273.38,-512.64\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"276.88,-512.57 273.38,-502.57 269.88,-512.57 276.88,-512.57\"/>\n",
       "</g>\n",
       "<!-- Enzymes[Br&#45;Zma01000] -->\n",
       "<g id=\"node11\" class=\"node\">\n",
       "<title>Enzymes[Br&#45;Zma01000]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1158.38\" cy=\"-116.61\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1158.38\" y=\"-120.41\" font-family=\"Times,serif\" font-size=\"14.00\">in: 100</text>\n",
       "<text text-anchor=\"middle\" x=\"1158.38\" y=\"-105.41\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- Enzymes[Br&#45;Zma01000]&#45;&gt;y_hat -->\n",
       "<g id=\"edge60\" class=\"edge\">\n",
       "<title>Enzymes[Br&#45;Zma01000]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1138.28,-92.1C1129.62,-81.97 1119.37,-69.98 1110.08,-59.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1112.69,-56.77 1103.53,-51.45 1107.37,-61.32 1112.69,-56.77\"/>\n",
       "</g>\n",
       "<!-- 2.Transferases -->\n",
       "<g id=\"node12\" class=\"node\">\n",
       "<title>2.Transferases</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"528.38\" cy=\"-206.35\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"528.38\" y=\"-210.15\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"528.38\" y=\"-195.15\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 2.Transferases&#45;&gt;Enzymes[Br&#45;Zma01000] -->\n",
       "<g id=\"edge11\" class=\"edge\">\n",
       "<title>2.Transferases&#45;&gt;Enzymes[Br&#45;Zma01000]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M574.35,-199.03C614.45,-193.59 674.22,-185.66 726.38,-179.48 876.4,-161.69 915.95,-171.62 1064.38,-143.48 1078.56,-140.79 1093.73,-137.02 1107.65,-133.19\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1108.86,-136.49 1117.53,-130.41 1106.96,-129.75 1108.86,-136.49\"/>\n",
       "</g>\n",
       "<!-- 7.Translocases -->\n",
       "<g id=\"node13\" class=\"node\">\n",
       "<title>7.Translocases</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1477.38\" cy=\"-206.35\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1477.38\" y=\"-210.15\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"1477.38\" y=\"-195.15\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 7.Translocases&#45;&gt;Enzymes[Br&#45;Zma01000] -->\n",
       "<g id=\"edge12\" class=\"edge\">\n",
       "<title>7.Translocases&#45;&gt;Enzymes[Br&#45;Zma01000]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1435.34,-193.79C1377.37,-177.84 1272.94,-149.12 1210.2,-131.86\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1211.07,-128.47 1200.5,-129.2 1209.21,-135.22 1211.07,-128.47\"/>\n",
       "</g>\n",
       "<!-- 2.1TransferringOne&#45;CarbonGroups -->\n",
       "<g id=\"node14\" class=\"node\">\n",
       "<title>2.1TransferringOne&#45;CarbonGroups</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"106.38\" cy=\"-296.09\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"106.38\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"106.38\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 2.1TransferringOne&#45;CarbonGroups&#45;&gt;2.Transferases -->\n",
       "<g id=\"edge13\" class=\"edge\">\n",
       "<title>2.1TransferringOne&#45;CarbonGroups&#45;&gt;2.Transferases</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M150.34,-285.95C228.14,-269.77 389.79,-236.16 474.28,-218.6\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"475.11,-222 484.19,-216.54 473.69,-215.15 475.11,-222\"/>\n",
       "</g>\n",
       "<!-- 2.1.1Methyltransferases -->\n",
       "<g id=\"node15\" class=\"node\">\n",
       "<title>2.1.1Methyltransferases</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"86.38\" cy=\"-385.83\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"86.38\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"86.38\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 2.1.1Methyltransferases&#45;&gt;2.1TransferringOne&#45;CarbonGroups -->\n",
       "<g id=\"edge14\" class=\"edge\">\n",
       "<title>2.1.1Methyltransferases&#45;&gt;2.1TransferringOne&#45;CarbonGroups</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M92.27,-358.98C94.14,-350.77 96.24,-341.54 98.24,-332.76\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"101.72,-333.28 100.53,-322.76 94.89,-331.73 101.72,-333.28\"/>\n",
       "</g>\n",
       "<!-- 2.1.1.34Trna(Guanosine18&#45;2&#39;&#45;O)&#45;Methyltransferase&#45;&gt;2.1.1Methyltransferases -->\n",
       "<g id=\"edge15\" class=\"edge\">\n",
       "<title>2.1.1.34Trna(Guanosine18&#45;2&#39;&#45;O)&#45;Methyltransferase&#45;&gt;2.1.1Methyltransferases</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M58.66,-449.19C62.5,-440.55 66.86,-430.74 70.96,-421.5\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"74.26,-422.71 75.12,-412.15 67.86,-419.87 74.26,-422.71\"/>\n",
       "</g>\n",
       "<!-- TransferRnaBiogenesis[Br&#45;Zma03016] -->\n",
       "<g id=\"node17\" class=\"node\">\n",
       "<title>TransferRnaBiogenesis[Br&#45;Zma03016]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"895.38\" cy=\"-116.61\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"895.38\" y=\"-120.41\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"895.38\" y=\"-105.41\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- TransferRnaBiogenesis[Br&#45;Zma03016]&#45;&gt;y_hat -->\n",
       "<g id=\"edge61\" class=\"edge\">\n",
       "<title>TransferRnaBiogenesis[Br&#45;Zma03016]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M931.18,-98.9C961.63,-84.69 1005.5,-64.21 1038.33,-48.89\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1040.12,-51.92 1047.71,-44.52 1037.16,-45.58 1040.12,-51.92\"/>\n",
       "</g>\n",
       "<!-- EukaryoticType -->\n",
       "<g id=\"node18\" class=\"node\">\n",
       "<title>EukaryoticType</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"782.38\" cy=\"-206.35\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"782.38\" y=\"-210.15\" font-family=\"Times,serif\" font-size=\"14.00\">in: 200</text>\n",
       "<text text-anchor=\"middle\" x=\"782.38\" y=\"-195.15\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- EukaryoticType&#45;&gt;TransferRnaBiogenesis[Br&#45;Zma03016] -->\n",
       "<g id=\"edge17\" class=\"edge\">\n",
       "<title>EukaryoticType&#45;&gt;TransferRnaBiogenesis[Br&#45;Zma03016]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M809.73,-184.11C824.86,-172.36 843.83,-157.64 860.07,-145.03\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"862.52,-147.55 868.27,-138.65 858.23,-142.02 862.52,-147.55\"/>\n",
       "</g>\n",
       "<!-- ChromosomeAndAssociatedProteins[Br&#45;Zma03036] -->\n",
       "<g id=\"node24\" class=\"node\">\n",
       "<title>ChromosomeAndAssociatedProteins[Br&#45;Zma03036]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1008.38\" cy=\"-116.61\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1008.38\" y=\"-120.41\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"1008.38\" y=\"-105.41\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- EukaryoticType&#45;&gt;ChromosomeAndAssociatedProteins[Br&#45;Zma03036] -->\n",
       "<g id=\"edge24\" class=\"edge\">\n",
       "<title>EukaryoticType&#45;&gt;ChromosomeAndAssociatedProteins[Br&#45;Zma03036]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M822.31,-191.68C856.76,-179.67 907.71,-161.35 951.38,-143.48 955.28,-141.88 959.31,-140.17 963.34,-138.43\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"964.94,-141.55 972.68,-134.31 962.11,-135.14 964.94,-141.55\"/>\n",
       "</g>\n",
       "<!-- MessengerRnaBiogenesis[Br&#45;Zma03019] -->\n",
       "<g id=\"node37\" class=\"node\">\n",
       "<title>MessengerRnaBiogenesis[Br&#45;Zma03019]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"782.38\" cy=\"-116.61\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"782.38\" y=\"-120.41\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"782.38\" y=\"-105.41\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- EukaryoticType&#45;&gt;MessengerRnaBiogenesis[Br&#45;Zma03019] -->\n",
       "<g id=\"edge41\" class=\"edge\">\n",
       "<title>EukaryoticType&#45;&gt;MessengerRnaBiogenesis[Br&#45;Zma03019]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M782.38,-179.03C782.38,-171.06 782.38,-162.17 782.38,-153.68\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"785.88,-153.61 782.38,-143.61 778.88,-153.61 785.88,-153.61\"/>\n",
       "</g>\n",
       "<!-- TrnaModificationFactors -->\n",
       "<g id=\"node19\" class=\"node\">\n",
       "<title>TrnaModificationFactors</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"379.38\" cy=\"-296.09\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"379.38\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"379.38\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- TrnaModificationFactors&#45;&gt;EukaryoticType -->\n",
       "<g id=\"edge18\" class=\"edge\">\n",
       "<title>TrnaModificationFactors&#45;&gt;EukaryoticType</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M423.27,-285.53C497.61,-269.35 648.05,-236.6 728.59,-219.06\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"729.51,-222.44 738.54,-216.9 728.02,-215.6 729.51,-222.44\"/>\n",
       "</g>\n",
       "<!-- GeneSilencing -->\n",
       "<g id=\"node20\" class=\"node\">\n",
       "<title>GeneSilencing</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"572.38\" cy=\"-296.09\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"572.38\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"572.38\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- GeneSilencing&#45;&gt;EukaryoticType -->\n",
       "<g id=\"edge19\" class=\"edge\">\n",
       "<title>GeneSilencing&#45;&gt;EukaryoticType</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M609.92,-279.4C645.05,-264.73 697.65,-242.75 735.51,-226.93\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"736.93,-230.13 744.8,-223.05 734.23,-223.67 736.93,-230.13\"/>\n",
       "</g>\n",
       "<!-- MrnaSurveillanceAndTransportFactors -->\n",
       "<g id=\"node21\" class=\"node\">\n",
       "<title>MrnaSurveillanceAndTransportFactors</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"725.38\" cy=\"-296.09\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"725.38\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"725.38\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- MrnaSurveillanceAndTransportFactors&#45;&gt;EukaryoticType -->\n",
       "<g id=\"edge20\" class=\"edge\">\n",
       "<title>MrnaSurveillanceAndTransportFactors&#45;&gt;EukaryoticType</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M741.25,-270.65C747.32,-261.32 754.34,-250.51 760.84,-240.5\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"763.87,-242.26 766.39,-231.96 758,-238.44 763.87,-242.26\"/>\n",
       "</g>\n",
       "<!-- MrnaDegradationFactors -->\n",
       "<g id=\"node22\" class=\"node\">\n",
       "<title>MrnaDegradationFactors</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"838.38\" cy=\"-296.09\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"838.38\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"838.38\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- MrnaDegradationFactors&#45;&gt;EukaryoticType -->\n",
       "<g id=\"edge21\" class=\"edge\">\n",
       "<title>MrnaDegradationFactors&#45;&gt;EukaryoticType</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M822.78,-270.65C816.82,-261.32 809.92,-250.51 803.53,-240.5\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"806.42,-238.51 798.09,-231.96 800.52,-242.28 806.42,-238.51\"/>\n",
       "</g>\n",
       "<!-- Methyltransferases&#45;&gt;TrnaModificationFactors -->\n",
       "<g id=\"edge22\" class=\"edge\">\n",
       "<title>Methyltransferases&#45;&gt;TrnaModificationFactors</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M186.79,-453.16C226.42,-421.05 301.26,-360.4 345.06,-324.9\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"347.3,-327.59 352.87,-318.57 342.89,-322.15 347.3,-327.59\"/>\n",
       "</g>\n",
       "<!-- ChromosomeAndAssociatedProteins[Br&#45;Zma03036]&#45;&gt;y_hat -->\n",
       "<g id=\"edge62\" class=\"edge\">\n",
       "<title>ChromosomeAndAssociatedProteins[Br&#45;Zma03036]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1028.47,-92.1C1037.13,-81.97 1047.38,-69.98 1056.67,-59.11\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1059.38,-61.32 1063.22,-51.45 1054.06,-56.77 1059.38,-61.32\"/>\n",
       "</g>\n",
       "<!-- Risc(Rna&#45;InducedSilencingComplex)&#45;&gt;GeneSilencing -->\n",
       "<g id=\"edge25\" class=\"edge\">\n",
       "<title>Risc(Rna&#45;InducedSilencingComplex)&#45;&gt;GeneSilencing</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M305.37,-455.58C359.97,-423.17 471.04,-357.24 531.29,-321.48\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"533.34,-324.33 540.16,-316.22 529.77,-318.31 533.34,-324.33\"/>\n",
       "</g>\n",
       "<!-- 09102EnergyMetabolism -->\n",
       "<g id=\"node26\" class=\"node\">\n",
       "<title>09102EnergyMetabolism</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1600.38\" cy=\"-385.83\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1600.38\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"1600.38\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 09102EnergyMetabolism&#45;&gt;09100Metabolism -->\n",
       "<g id=\"edge27\" class=\"edge\">\n",
       "<title>09102EnergyMetabolism&#45;&gt;09100Metabolism</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1570.42,-364.92C1551.65,-352.45 1527.27,-336.24 1507.05,-322.8\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1508.93,-319.85 1498.67,-317.22 1505.06,-325.68 1508.93,-319.85\"/>\n",
       "</g>\n",
       "<!-- 00190OxidativePhosphorylation -->\n",
       "<g id=\"node27\" class=\"node\">\n",
       "<title>00190OxidativePhosphorylation</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1629.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1629.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 12</text>\n",
       "<text text-anchor=\"middle\" x=\"1629.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 00190OxidativePhosphorylation&#45;&gt;09102EnergyMetabolism -->\n",
       "<g id=\"edge28\" class=\"edge\">\n",
       "<title>00190OxidativePhosphorylation&#45;&gt;09102EnergyMetabolism</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1620.83,-448.72C1618.09,-440.42 1615,-431.08 1612.07,-422.22\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1615.32,-420.89 1608.86,-412.5 1608.67,-423.09 1615.32,-420.89\"/>\n",
       "</g>\n",
       "<!-- 100383860 -->\n",
       "<g id=\"node28\" class=\"node\">\n",
       "<title>100383860</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1629.38\" cy=\"-565.31\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1629.38\" y=\"-569.11\" font-family=\"Times,serif\" font-size=\"14.00\">in: 0.0</text>\n",
       "<text text-anchor=\"middle\" x=\"1629.38\" y=\"-554.11\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100383860&#45;&gt;00190OxidativePhosphorylation -->\n",
       "<g id=\"edge29\" class=\"edge\">\n",
       "<title>100383860&#45;&gt;00190OxidativePhosphorylation</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1629.38,-537.99C1629.38,-530.02 1629.38,-521.13 1629.38,-512.64\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1632.88,-512.57 1629.38,-502.57 1625.88,-512.57 1632.88,-512.57\"/>\n",
       "</g>\n",
       "<!-- 7.1.2.1P&#45;TypeH+&#45;ExportingTransporter -->\n",
       "<g id=\"node32\" class=\"node\">\n",
       "<title>7.1.2.1P&#45;TypeH+&#45;ExportingTransporter</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1742.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1742.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 12</text>\n",
       "<text text-anchor=\"middle\" x=\"1742.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100383860&#45;&gt;7.1.2.1P&#45;TypeH+&#45;ExportingTransporter -->\n",
       "<g id=\"edge34\" class=\"edge\">\n",
       "<title>100383860&#45;&gt;7.1.2.1P&#45;TypeH+&#45;ExportingTransporter</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1656.73,-543.07C1671.86,-531.32 1690.83,-516.6 1707.07,-503.99\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1709.52,-506.51 1715.27,-497.61 1705.23,-500.98 1709.52,-506.51\"/>\n",
       "</g>\n",
       "<!-- 103630585 -->\n",
       "<g id=\"node29\" class=\"node\">\n",
       "<title>103630585</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1742.38\" cy=\"-565.31\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1742.38\" y=\"-569.11\" font-family=\"Times,serif\" font-size=\"14.00\">in: 0.0</text>\n",
       "<text text-anchor=\"middle\" x=\"1742.38\" y=\"-554.11\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 103630585&#45;&gt;00190OxidativePhosphorylation -->\n",
       "<g id=\"edge30\" class=\"edge\">\n",
       "<title>103630585&#45;&gt;00190OxidativePhosphorylation</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1715.02,-543.07C1699.89,-531.32 1680.92,-516.6 1664.69,-503.99\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1666.52,-500.98 1656.48,-497.61 1662.23,-506.51 1666.52,-500.98\"/>\n",
       "</g>\n",
       "<!-- 103630585&#45;&gt;7.1.2.1P&#45;TypeH+&#45;ExportingTransporter -->\n",
       "<g id=\"edge35\" class=\"edge\">\n",
       "<title>103630585&#45;&gt;7.1.2.1P&#45;TypeH+&#45;ExportingTransporter</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1742.38,-537.99C1742.38,-530.02 1742.38,-521.13 1742.38,-512.64\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1745.88,-512.57 1742.38,-502.57 1738.88,-512.57 1745.88,-512.57\"/>\n",
       "</g>\n",
       "<!-- 7.1CatalysingTheTranslocationOfHydrons -->\n",
       "<g id=\"node30\" class=\"node\">\n",
       "<title>7.1CatalysingTheTranslocationOfHydrons</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1605.38\" cy=\"-296.09\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1605.38\" y=\"-299.89\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"1605.38\" y=\"-284.89\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 7.1CatalysingTheTranslocationOfHydrons&#45;&gt;7.Translocases -->\n",
       "<g id=\"edge31\" class=\"edge\">\n",
       "<title>7.1CatalysingTheTranslocationOfHydrons&#45;&gt;7.Translocases</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1576.01,-274.96C1558.01,-262.62 1534.77,-246.69 1515.38,-233.4\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1517.28,-230.46 1507.05,-227.69 1513.32,-236.23 1517.28,-230.46\"/>\n",
       "</g>\n",
       "<!-- 7.1.2LinkedToTheHydrolysisOfANucleosideTriphosphate -->\n",
       "<g id=\"node31\" class=\"node\">\n",
       "<title>7.1.2LinkedToTheHydrolysisOfANucleosideTriphosphate</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1727.38\" cy=\"-385.83\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1727.38\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"1727.38\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 7.1.2LinkedToTheHydrolysisOfANucleosideTriphosphate&#45;&gt;7.1CatalysingTheTranslocationOfHydrons -->\n",
       "<g id=\"edge32\" class=\"edge\">\n",
       "<title>7.1.2LinkedToTheHydrolysisOfANucleosideTriphosphate&#45;&gt;7.1CatalysingTheTranslocationOfHydrons</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1698.77,-364.26C1682.07,-352.25 1660.8,-336.95 1642.77,-323.99\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1644.39,-320.83 1634.22,-317.84 1640.3,-326.52 1644.39,-320.83\"/>\n",
       "</g>\n",
       "<!-- 7.1.2.1P&#45;TypeH+&#45;ExportingTransporter&#45;&gt;7.1.2LinkedToTheHydrolysisOfANucleosideTriphosphate -->\n",
       "<g id=\"edge33\" class=\"edge\">\n",
       "<title>7.1.2.1P&#45;TypeH+&#45;ExportingTransporter&#45;&gt;7.1.2LinkedToTheHydrolysisOfANucleosideTriphosphate</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1737.96,-448.72C1736.57,-440.58 1735,-431.44 1733.51,-422.74\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1736.95,-422.07 1731.82,-412.8 1730.05,-423.25 1736.95,-422.07\"/>\n",
       "</g>\n",
       "<!-- 09123Folding,SortingAndDegradation -->\n",
       "<g id=\"node33\" class=\"node\">\n",
       "<title>09123Folding,SortingAndDegradation</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1064.38\" cy=\"-385.83\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1064.38\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 100</text>\n",
       "<text text-anchor=\"middle\" x=\"1064.38\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 09123Folding,SortingAndDegradation&#45;&gt;09120GeneticInformationProcessing -->\n",
       "<g id=\"edge36\" class=\"edge\">\n",
       "<title>09123Folding,SortingAndDegradation&#45;&gt;09120GeneticInformationProcessing</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1069.97,-358.98C1071.74,-350.84 1073.72,-341.7 1075.6,-333\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1079.06,-333.57 1077.75,-323.06 1072.22,-332.09 1079.06,-333.57\"/>\n",
       "</g>\n",
       "<!-- 03018RnaDegradation -->\n",
       "<g id=\"node34\" class=\"node\">\n",
       "<title>03018RnaDegradation</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"951.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"951.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 4</text>\n",
       "<text text-anchor=\"middle\" x=\"951.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 03018RnaDegradation&#45;&gt;09123Folding,SortingAndDegradation -->\n",
       "<g id=\"edge37\" class=\"edge\">\n",
       "<title>03018RnaDegradation&#45;&gt;09123Folding,SortingAndDegradation</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M978.73,-453.33C993.86,-441.58 1012.83,-426.86 1029.07,-414.25\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1031.52,-416.77 1037.27,-407.87 1027.23,-411.24 1031.52,-416.77\"/>\n",
       "</g>\n",
       "<!-- 03060ProteinExport -->\n",
       "<g id=\"node35\" class=\"node\">\n",
       "<title>03060ProteinExport</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1064.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1064.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 12</text>\n",
       "<text text-anchor=\"middle\" x=\"1064.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 03060ProteinExport&#45;&gt;09123Folding,SortingAndDegradation -->\n",
       "<g id=\"edge38\" class=\"edge\">\n",
       "<title>03060ProteinExport&#45;&gt;09123Folding,SortingAndDegradation</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1064.38,-448.25C1064.38,-440.28 1064.38,-431.39 1064.38,-422.9\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1067.88,-422.83 1064.38,-412.83 1060.88,-422.83 1067.88,-422.83\"/>\n",
       "</g>\n",
       "<!-- 100383837 -->\n",
       "<g id=\"node36\" class=\"node\">\n",
       "<title>100383837</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"781.38\" cy=\"-565.31\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"781.38\" y=\"-569.11\" font-family=\"Times,serif\" font-size=\"14.00\">in: 0.0</text>\n",
       "<text text-anchor=\"middle\" x=\"781.38\" y=\"-554.11\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100383837&#45;&gt;03019MessengerRnaBiogenesis[Br&#45;Zma03019] -->\n",
       "<g id=\"edge40\" class=\"edge\">\n",
       "<title>100383837&#45;&gt;03019MessengerRnaBiogenesis[Br&#45;Zma03019]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M747.18,-546.56C720.72,-532.82 683.92,-513.71 655.45,-498.93\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"657.01,-495.8 646.52,-494.3 653.79,-502.01 657.01,-495.8\"/>\n",
       "</g>\n",
       "<!-- 100383837&#45;&gt;03018RnaDegradation -->\n",
       "<g id=\"edge39\" class=\"edge\">\n",
       "<title>100383837&#45;&gt;03018RnaDegradation</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M815.78,-546.56C842.39,-532.82 879.41,-513.71 908.05,-498.93\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"909.74,-502 917.03,-494.3 906.53,-495.78 909.74,-502\"/>\n",
       "</g>\n",
       "<!-- P&#45;BodySpecificFactors -->\n",
       "<g id=\"node39\" class=\"node\">\n",
       "<title>P&#45;BodySpecificFactors</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"725.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"725.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 4</text>\n",
       "<text text-anchor=\"middle\" x=\"725.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100383837&#45;&gt;P&#45;BodySpecificFactors -->\n",
       "<g id=\"edge44\" class=\"edge\">\n",
       "<title>100383837&#45;&gt;P&#45;BodySpecificFactors</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M765.78,-539.87C759.82,-530.54 752.92,-519.73 746.53,-509.72\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"749.42,-507.73 741.09,-501.18 743.52,-511.5 749.42,-507.73\"/>\n",
       "</g>\n",
       "<!-- Ccr4&#45;NotComplex -->\n",
       "<g id=\"node41\" class=\"node\">\n",
       "<title>Ccr4&#45;NotComplex</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"838.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"838.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 4</text>\n",
       "<text text-anchor=\"middle\" x=\"838.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100383837&#45;&gt;Ccr4&#45;NotComplex -->\n",
       "<g id=\"edge47\" class=\"edge\">\n",
       "<title>100383837&#45;&gt;Ccr4&#45;NotComplex</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M797.25,-539.87C803.32,-530.54 810.34,-519.73 816.84,-509.72\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"819.87,-511.48 822.39,-501.18 814,-507.66 819.87,-511.48\"/>\n",
       "</g>\n",
       "<!-- MessengerRnaBiogenesis[Br&#45;Zma03019]&#45;&gt;y_hat -->\n",
       "<g id=\"edge63\" class=\"edge\">\n",
       "<title>MessengerRnaBiogenesis[Br&#45;Zma03019]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M817.55,-98.47C824.68,-95.33 832.2,-92.26 839.38,-89.74 903.42,-67.29 979.56,-49.31 1029.55,-38.64\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1030.43,-42.03 1039.49,-36.53 1028.98,-35.18 1030.43,-42.03\"/>\n",
       "</g>\n",
       "<!-- MrnaCycleFactors -->\n",
       "<g id=\"node38\" class=\"node\">\n",
       "<title>MrnaCycleFactors</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"725.38\" cy=\"-385.83\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"725.38\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"725.38\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- MrnaCycleFactors&#45;&gt;MrnaSurveillanceAndTransportFactors -->\n",
       "<g id=\"edge42\" class=\"edge\">\n",
       "<title>MrnaCycleFactors&#45;&gt;MrnaSurveillanceAndTransportFactors</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M725.38,-358.51C725.38,-350.54 725.38,-341.65 725.38,-333.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"728.88,-333.09 725.38,-323.09 721.88,-333.09 728.88,-333.09\"/>\n",
       "</g>\n",
       "<!-- P&#45;BodySpecificFactors&#45;&gt;MrnaCycleFactors -->\n",
       "<g id=\"edge43\" class=\"edge\">\n",
       "<title>P&#45;BodySpecificFactors&#45;&gt;MrnaCycleFactors</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M725.38,-448.25C725.38,-440.28 725.38,-431.39 725.38,-422.9\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"728.88,-422.83 725.38,-412.83 721.88,-422.83 728.88,-422.83\"/>\n",
       "</g>\n",
       "<!-- 3&#39;&#45;5&#39;Decay -->\n",
       "<g id=\"node40\" class=\"node\">\n",
       "<title>3&#39;&#45;5&#39;Decay</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"838.38\" cy=\"-385.83\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"838.38\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"838.38\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 3&#39;&#45;5&#39;Decay&#45;&gt;MrnaDegradationFactors -->\n",
       "<g id=\"edge45\" class=\"edge\">\n",
       "<title>3&#39;&#45;5&#39;Decay&#45;&gt;MrnaDegradationFactors</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M838.38,-358.51C838.38,-350.54 838.38,-341.65 838.38,-333.16\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"841.88,-333.09 838.38,-323.09 834.88,-333.09 841.88,-333.09\"/>\n",
       "</g>\n",
       "<!-- Ccr4&#45;NotComplex&#45;&gt;3&#39;&#45;5&#39;Decay -->\n",
       "<g id=\"edge46\" class=\"edge\">\n",
       "<title>Ccr4&#45;NotComplex&#45;&gt;3&#39;&#45;5&#39;Decay</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M838.38,-448.25C838.38,-440.28 838.38,-431.39 838.38,-422.9\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"841.88,-422.83 838.38,-412.83 834.88,-422.83 841.88,-422.83\"/>\n",
       "</g>\n",
       "<!-- 100191673 -->\n",
       "<g id=\"node42\" class=\"node\">\n",
       "<title>100191673</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1177.38\" cy=\"-565.31\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1177.38\" y=\"-569.11\" font-family=\"Times,serif\" font-size=\"14.00\">in: 0.0</text>\n",
       "<text text-anchor=\"middle\" x=\"1177.38\" y=\"-554.11\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100191673&#45;&gt;03060ProteinExport -->\n",
       "<g id=\"edge48\" class=\"edge\">\n",
       "<title>100191673&#45;&gt;03060ProteinExport</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1150.02,-543.07C1134.89,-531.32 1115.92,-516.6 1099.69,-503.99\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1101.52,-500.98 1091.48,-497.61 1097.23,-506.51 1101.52,-500.98\"/>\n",
       "</g>\n",
       "<!-- 02044SecretionSystem[Br&#45;Zma02044] -->\n",
       "<g id=\"node43\" class=\"node\">\n",
       "<title>02044SecretionSystem[Br&#45;Zma02044]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1177.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1177.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 12</text>\n",
       "<text text-anchor=\"middle\" x=\"1177.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100191673&#45;&gt;02044SecretionSystem[Br&#45;Zma02044] -->\n",
       "<g id=\"edge51\" class=\"edge\">\n",
       "<title>100191673&#45;&gt;02044SecretionSystem[Br&#45;Zma02044]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1177.38,-537.99C1177.38,-530.02 1177.38,-521.13 1177.38,-512.64\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1180.88,-512.57 1177.38,-502.57 1173.88,-512.57 1180.88,-512.57\"/>\n",
       "</g>\n",
       "<!-- EukaryoticSec&#45;SrpProtein -->\n",
       "<g id=\"node47\" class=\"node\">\n",
       "<title>EukaryoticSec&#45;SrpProtein</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1290.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1290.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 12</text>\n",
       "<text text-anchor=\"middle\" x=\"1290.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100191673&#45;&gt;EukaryoticSec&#45;SrpProtein -->\n",
       "<g id=\"edge54\" class=\"edge\">\n",
       "<title>100191673&#45;&gt;EukaryoticSec&#45;SrpProtein</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1204.73,-543.07C1219.86,-531.32 1238.83,-516.6 1255.07,-503.99\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1257.52,-506.51 1263.27,-497.61 1253.23,-500.98 1257.52,-506.51\"/>\n",
       "</g>\n",
       "<!-- 02044SecretionSystem[Br&#45;Zma02044]&#45;&gt;09183ProteinFamilies&#45;SignalingAndCellularProcesses -->\n",
       "<g id=\"edge49\" class=\"edge\">\n",
       "<title>02044SecretionSystem[Br&#45;Zma02044]&#45;&gt;09183ProteinFamilies&#45;SignalingAndCellularProcesses</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1177.38,-448.25C1177.38,-440.28 1177.38,-431.39 1177.38,-422.9\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1180.88,-422.83 1177.38,-412.83 1173.88,-422.83 1180.88,-422.83\"/>\n",
       "</g>\n",
       "<!-- 02000Transporters[Br&#45;Zma02000] -->\n",
       "<g id=\"node44\" class=\"node\">\n",
       "<title>02000Transporters[Br&#45;Zma02000]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1403.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1403.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 4</text>\n",
       "<text text-anchor=\"middle\" x=\"1403.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 02000Transporters[Br&#45;Zma02000]&#45;&gt;09183ProteinFamilies&#45;SignalingAndCellularProcesses -->\n",
       "<g id=\"edge50\" class=\"edge\">\n",
       "<title>02000Transporters[Br&#45;Zma02000]&#45;&gt;09183ProteinFamilies&#45;SignalingAndCellularProcesses</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1367.76,-457.65C1360.74,-454.53 1353.37,-451.4 1346.38,-448.7 1297.6,-429.86 1283.15,-431.54 1234.38,-412.7 1230.44,-411.18 1226.39,-409.53 1222.35,-407.82\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1223.55,-404.52 1212.99,-403.75 1220.77,-410.95 1223.55,-404.52\"/>\n",
       "</g>\n",
       "<!-- SecretionSystem[Br&#45;Zma02044] -->\n",
       "<g id=\"node45\" class=\"node\">\n",
       "<title>SecretionSystem[Br&#45;Zma02044]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1261.38\" cy=\"-206.35\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1261.38\" y=\"-210.15\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"1261.38\" y=\"-195.15\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- SecretionSystem[Br&#45;Zma02044]&#45;&gt;y_hat -->\n",
       "<g id=\"edge64\" class=\"edge\">\n",
       "<title>SecretionSystem[Br&#45;Zma02044]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1256.56,-179.39C1250.61,-153.78 1238.2,-114.92 1214.38,-89.74 1192.97,-67.12 1161.64,-51.91 1134.98,-42.23\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1135.99,-38.88 1125.4,-38.92 1133.7,-45.5 1135.99,-38.88\"/>\n",
       "</g>\n",
       "<!-- Sec(Secretion)System -->\n",
       "<g id=\"node46\" class=\"node\">\n",
       "<title>Sec(Secretion)System</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1290.38\" cy=\"-385.83\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1290.38\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"1290.38\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- Sec(Secretion)System&#45;&gt;SecretionSystem[Br&#45;Zma02044] -->\n",
       "<g id=\"edge52\" class=\"edge\">\n",
       "<title>Sec(Secretion)System&#45;&gt;SecretionSystem[Br&#45;Zma02044]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1286.16,-359.02C1281.18,-328.53 1272.86,-277.61 1267.25,-243.27\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1270.66,-242.49 1265.6,-233.19 1263.76,-243.62 1270.66,-242.49\"/>\n",
       "</g>\n",
       "<!-- EukaryoticSec&#45;SrpProtein&#45;&gt;Sec(Secretion)System -->\n",
       "<g id=\"edge53\" class=\"edge\">\n",
       "<title>EukaryoticSec&#45;SrpProtein&#45;&gt;Sec(Secretion)System</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1290.38,-448.25C1290.38,-440.28 1290.38,-431.39 1290.38,-422.9\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1293.88,-422.83 1290.38,-412.83 1286.88,-422.83 1293.88,-422.83\"/>\n",
       "</g>\n",
       "<!-- 100275685 -->\n",
       "<g id=\"node48\" class=\"node\">\n",
       "<title>100275685</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1459.38\" cy=\"-565.31\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1459.38\" y=\"-569.11\" font-family=\"Times,serif\" font-size=\"14.00\">in: 0.0</text>\n",
       "<text text-anchor=\"middle\" x=\"1459.38\" y=\"-554.11\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100275685&#45;&gt;02000Transporters[Br&#45;Zma02000] -->\n",
       "<g id=\"edge55\" class=\"edge\">\n",
       "<title>100275685&#45;&gt;02000Transporters[Br&#45;Zma02000]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1443.78,-539.87C1437.82,-530.54 1430.92,-519.73 1424.53,-509.72\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1427.42,-507.73 1419.09,-501.18 1421.52,-511.5 1427.42,-507.73\"/>\n",
       "</g>\n",
       "<!-- PoresIonChannels[Tc&#45;1] -->\n",
       "<g id=\"node51\" class=\"node\">\n",
       "<title>PoresIonChannels[Tc&#45;1]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1516.38\" cy=\"-475.57\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1516.38\" y=\"-479.37\" font-family=\"Times,serif\" font-size=\"14.00\">in: 4</text>\n",
       "<text text-anchor=\"middle\" x=\"1516.38\" y=\"-464.37\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- 100275685&#45;&gt;PoresIonChannels[Tc&#45;1] -->\n",
       "<g id=\"edge58\" class=\"edge\">\n",
       "<title>100275685&#45;&gt;PoresIonChannels[Tc&#45;1]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1475.25,-539.87C1481.32,-530.54 1488.34,-519.73 1494.84,-509.72\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1497.87,-511.48 1500.39,-501.18 1492,-507.66 1497.87,-511.48\"/>\n",
       "</g>\n",
       "<!-- Transporters[Br&#45;Zma02000] -->\n",
       "<g id=\"node49\" class=\"node\">\n",
       "<title>Transporters[Br&#45;Zma02000]</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1334.38\" cy=\"-116.61\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1334.38\" y=\"-120.41\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"1334.38\" y=\"-105.41\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- Transporters[Br&#45;Zma02000]&#45;&gt;y_hat -->\n",
       "<g id=\"edge65\" class=\"edge\">\n",
       "<title>Transporters[Br&#45;Zma02000]&#45;&gt;y_hat</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1294.99,-101.6C1284.11,-97.75 1272.29,-93.57 1261.38,-89.74 1217.78,-74.44 1168.12,-57.19 1132.34,-44.8\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1133.23,-41.41 1122.64,-41.44 1130.94,-48.02 1133.23,-41.41\"/>\n",
       "</g>\n",
       "<!-- OtherTransporters -->\n",
       "<g id=\"node50\" class=\"node\">\n",
       "<title>OtherTransporters</title>\n",
       "<ellipse fill=\"none\" stroke=\"black\" cx=\"1459.38\" cy=\"-385.83\" rx=\"47.25\" ry=\"26.74\"/>\n",
       "<text text-anchor=\"middle\" x=\"1459.38\" y=\"-389.63\" font-family=\"Times,serif\" font-size=\"14.00\">in: 50</text>\n",
       "<text text-anchor=\"middle\" x=\"1459.38\" y=\"-374.63\" font-family=\"Times,serif\" font-size=\"14.00\">out: 50</text>\n",
       "</g>\n",
       "<!-- OtherTransporters&#45;&gt;Transporters[Br&#45;Zma02000] -->\n",
       "<g id=\"edge56\" class=\"edge\">\n",
       "<title>OtherTransporters&#45;&gt;Transporters[Br&#45;Zma02000]</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1438.92,-361.25C1429.97,-350.15 1419.83,-336.41 1412.38,-322.96 1381.11,-266.58 1357.12,-195.11 1344.44,-153.07\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1347.73,-151.88 1341.53,-143.29 1341.02,-153.88 1347.73,-151.88\"/>\n",
       "</g>\n",
       "<!-- PoresIonChannels[Tc&#45;1]&#45;&gt;OtherTransporters -->\n",
       "<g id=\"edge57\" class=\"edge\">\n",
       "<title>PoresIonChannels[Tc&#45;1]&#45;&gt;OtherTransporters</title>\n",
       "<path fill=\"none\" stroke=\"black\" d=\"M1500.5,-450.13C1494.43,-440.8 1487.42,-429.99 1480.91,-419.98\"/>\n",
       "<polygon fill=\"black\" stroke=\"black\" points=\"1483.75,-417.92 1475.37,-411.44 1477.88,-421.74 1483.75,-417.92\"/>\n",
       "</g>\n",
       "</g>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<graphviz.graphs.Digraph at 0x7f05b35a3810>"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make in out size graph\n",
    "dot = ''\n",
    "dot = Digraph()\n",
    "for key in tqdm(kegg_connections.keys()):\n",
    "    key_label = 'in: '+str(input_size_dict[key])+'\\nout: '+str(output_size_dict[key])\n",
    "    dot.node(key, key_label)\n",
    "    for value in kegg_connections[key]:\n",
    "        # edge takes a head/tail whereas edges takes name pairs concatednated (A, B -> AB)in a list\n",
    "        dot.edge(value, key)    \n",
    "\n",
    "dot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "031878d4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca02856a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ebbc2a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d206d32d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41e9d1d1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "5d96f8dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # start by finding the top level -- all those keys which are theselves not values\n",
    "# res = find_uniq_keys_values(input_dict = kegg_connections)\n",
    "# all_keys = res['all_keys']\n",
    "# all_values = res['all_values']\n",
    "\n",
    "# # use the keys to find the input/outputs of the graph\n",
    "# output_nodes = [e for e in all_keys if e not in all_values]\n",
    "# input_nodes = [e for e in all_values if e not in all_keys]\n",
    "\n",
    "# # (output_nodes, input_nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "adc6c050",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # find the dependancies for run order from many dependancies to none\n",
    "# temp = kegg_connections.copy()\n",
    "\n",
    "# no_dependants = find_input_nodes(all_key_value_dict = find_uniq_keys_values(input_dict = temp))\n",
    "# # first pass. Same as the output nodes identified above\n",
    "# dependancy_order = []\n",
    "# # Then iterate\n",
    "# for ith in range(100): #TODO <- this should be set as a input parameter\n",
    "#     top_nodes = find_top_nodes(all_key_value_dict = find_uniq_keys_values(input_dict = temp))\n",
    "#     if top_nodes == []:\n",
    "#         break\n",
    "#     else:\n",
    "#         dependancy_order += top_nodes    \n",
    "#         # remove nodes from the graph that are at the 'top' level and haven't already been removed\n",
    "#         for key in [e for e in dependancy_order if e in temp.keys()]:\n",
    "#              temp.pop(key)\n",
    "\n",
    "# # dependancy_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "e8251172",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # reverse to get the order that the nodes should be called\n",
    "# dependancy_order.reverse()\n",
    "# # dependancy_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "272f38f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Trying out new approach: add a node for the input data tha will only flatten the input.\n",
    "# dependancy_order = input_nodes+dependancy_order\n",
    "\n",
    "# for key in input_nodes:\n",
    "#     kegg_connections[key] = [] #[key] # needs to contain itself so the model's `get_input_node()` function works \n",
    "#                                # or that function needs to change."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "53901bec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # build a dict to go from the node names in `no_dependants` to the list index in `ACGT_gene_slice_list`\n",
    "# brite_node_to_list_idx_dict = {}\n",
    "# for i in tqdm(range(len(kegg_gene_brite))):\n",
    "#     brite_node_to_list_idx_dict[str(kegg_gene_brite[i]['BRITE']['BRITE_PATHS'][0][-1])] = i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97426460",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14904e35",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52b49cbb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "e209df06",
   "metadata": {},
   "outputs": [],
   "source": [
    "# input_tensor_dict = {}\n",
    "# for e in no_dependants:\n",
    "#     input_tensor_dict[e] = ACGT_gene_slice_list[brite_node_to_list_idx_dict[e]]\n",
    "    \n",
    "# # input_tensor_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "7ca643b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Figure out expected input/output shapes\n",
    "# #==NOTE! This assumes only dense connections!==\n",
    "\n",
    "# # This could be replaced by a sort of \"distance from output\" measure\n",
    "# output_size_dict = dict(zip(dependancy_order, \n",
    "#                         [default_output_size for i in range(len(dependancy_order))]))\n",
    "# output_size_dict['y_hat'] = 1 \n",
    "# # output_size_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "b0e945e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Setup dropout % dictionary\n",
    "# dropout_pr_dict = dict(zip(dependancy_order, \n",
    "#                         [default_dropout_pr for i in range(len(dependancy_order))]))\n",
    "# dropout_pr_dict['y_hat'] = 0 # not required, output node is purely linear without dropout"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "8adedd40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Setup replicates of layers dictionary\n",
    "# block_rep_dict = dict(zip(dependancy_order, \n",
    "#                         [default_block_reps for i in range(len(dependancy_order))]))\n",
    "# block_rep_dict['y_hat'] = 1 # not required, output node is purely linear. Not a linear block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "4703e891",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # CHANNEL AWARE VERSION -----------------------------------------------------------------------------------\n",
    "# input_size_dict = kegg_connections.copy()\n",
    "\n",
    "# # use the expected output sizes from `output_size_dict` to fill in the non-data sizes\n",
    "# tensor_ndim = len(input_tensor_dict[list(input_tensor_dict.keys())[0]].shape)\n",
    "# for e in tqdm(input_size_dict.keys()):\n",
    "#     # overwrite named connections with the output size of those connections\n",
    "#     # if the entry is in no_dependants it's data so it's size needs to be grabbed from the input_tensor_dict\n",
    "    \n",
    "#     # is there no channel dim? (major/minor allele)\n",
    "#     if 2 == tensor_ndim:\n",
    "#         input_size_dict[e] = [\n",
    "#             list(input_tensor_dict[ee].shape)[-1] # <- NOTE! THIS ASSUMES ONLY DENSE CONNECTIONS (i.e. only the 1st dim is needed)\n",
    "#             if ee in no_dependants\n",
    "#             else output_size_dict[ee] for ee in input_size_dict[e]]\n",
    "#     elif 3 == tensor_ndim: # There is a channel dim\n",
    "#         input_size_dict[e] = [\n",
    "#             (list(input_tensor_dict[ee].shape)[1]*list(input_tensor_dict[ee].shape)[2]) # <- NOTE! THIS ASSUMES ONLY DENSE CONNECTIONS (i.e. only the 1st dim is needed)  \n",
    "#             if ee in no_dependants\n",
    "#             else output_size_dict[ee] for ee in input_size_dict[e]]\n",
    "\n",
    "# # Now walk over entries and overwrite with the sum of the inputs\n",
    "# for e in tqdm(input_size_dict.keys()):\n",
    "#     input_size_dict[e] = np.sum(input_size_dict[e])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "254e4664",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "df0c23fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# model = NeuralNetwork(example_dict = kegg_connections, \n",
    "#                       example_dict_input_size = input_size_dict,\n",
    "#                       example_dict_output_size = output_size_dict,\n",
    "#                       example_dict_dropout_pr= dropout_pr_dict,\n",
    "#                       example_block_rep_dict = block_rep_dict,\n",
    "#                       input_tensor_names = list(input_tensor_dict.keys()),\n",
    "#                       dependancy_order = dependancy_order) \n",
    "# model = model.to(device)\n",
    "# model(next(iter(training_dataloader))[1])\n",
    "# model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "07d9fb35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dot = ''\n",
    "# if vis_dot_bool:\n",
    "#     dot = Digraph()\n",
    "#     for key in tqdm(kegg_connections.keys()):\n",
    "#         key_label = 'in: '+str(input_size_dict[key])+'\\nout: '+str(output_size_dict[key])\n",
    "#         dot.node(key, key_label)\n",
    "#         for value in kegg_connections[key]:\n",
    "#             # edge takes a head/tail whereas edges takes name pairs concatednated (A, B -> AB)in a list\n",
    "#             dot.edge(value, key)    \n",
    "\n",
    "# dot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c9ebaea",
   "metadata": {},
   "source": [
    "## Set up DataSet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a550a369",
   "metadata": {},
   "source": [
    "Now we have\n",
    "\n",
    "- A dictionary with the connections: `example_dict`\n",
    "- The expected input sizes for each node: `example_dict_input_size`\n",
    "- A dictionary with the input tensors: `input_tensor_dict`\n",
    "- A list of the input tensors' names: `no_dependants` \n",
    "- A list with the order that each module should be called: `dependancy_order`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8e4e20e",
   "metadata": {},
   "source": [
    "To have a fair test of whether the model is working, I want to ensure there is information to learn in the dataset. To this end I'm using just two genotypes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "c5656cd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "x_list_temp = [torch.from_numpy(input_tensor_dict[key]).to(torch.float) for key in input_tensor_dict.keys()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "025f11ff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "c4b93408",
   "metadata": {},
   "outputs": [],
   "source": [
    "# YMat_cs = calc_cs(YMat[train_idx])\n",
    "# y_cs = apply_cs(YMat, YMat_cs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "529581bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# y_temp = torch.from_numpy(y_cs).to(torch.float)#[:, None]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "c7e82908",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "torch.Size([4926, 4, 1])"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_list_temp[0].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c21850c",
   "metadata": {},
   "source": [
    "## Set up NeuralNetwork, Data Loader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "0cb35a21",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Working version ====\n",
    "# Doesn't pass output node through relu\n",
    "class NeuralNetwork(nn.Module):\n",
    "    def __init__(self, \n",
    "                 example_dict, # contains the node (excluding input tensors)\n",
    "                 example_dict_input_size, # contains the input sizes (including the tensors)\n",
    "                 example_dict_output_size,\n",
    "                 example_dict_dropout_pr,\n",
    "                 example_block_rep_dict,\n",
    "                 input_tensor_names,\n",
    "                 dependancy_order\n",
    "                ):\n",
    "        super(NeuralNetwork, self).__init__()\n",
    "        def Linear_block(in_size, out_size, drop_pr, block_reps):\n",
    "            block_list = []\n",
    "            for i in range(block_reps):\n",
    "                if i == 0:\n",
    "                    block_list += [\n",
    "                        nn.Linear(in_size, out_size),\n",
    "                        nn.ReLU(),\n",
    "                        nn.Dropout(drop_pr)]\n",
    "                else:\n",
    "                    block_list += [\n",
    "                        nn.Linear(out_size, out_size),\n",
    "                        nn.ReLU(),\n",
    "                        nn.Dropout(drop_pr)]\n",
    "        \n",
    "            block = nn.ModuleList(block_list)\n",
    "            return(block)           \n",
    "        \n",
    "        # fill in the list in dependancy order. \n",
    "        layer_list = []\n",
    "        for key in dependancy_order:\n",
    "            if key in input_tensor_names:\n",
    "                layer_list += [\n",
    "                    nn.Flatten()\n",
    "                ]\n",
    "            elif key != 'y_hat':\n",
    "                layer_list += [\n",
    "                    Linear_block(in_size=example_dict_input_size[key], \n",
    "                                 out_size=example_dict_output_size[key], \n",
    "                                 drop_pr=example_dict_dropout_pr[key],\n",
    "                                 block_reps=example_block_rep_dict[key])\n",
    "                              ]\n",
    "            else:\n",
    "                layer_list += [\n",
    "                    nn.Linear(example_dict_input_size[key], \n",
    "                              example_dict_output_size[key])\n",
    "                              ]\n",
    "                \n",
    "\n",
    "        self.nn_layer_list = nn.ModuleList(layer_list)\n",
    "\n",
    "        # things for get_input_node in forward to work.\n",
    "        self.example_dict = example_dict\n",
    "        self.input_tensor_names = input_tensor_names\n",
    "        self.dependancy_order = dependancy_order\n",
    "        \n",
    "        self.input_tensor_lookup = dict(zip(input_tensor_names, \n",
    "                                            [i for i in range(len(input_tensor_names))]))\n",
    "        self.result_list = []\n",
    "        self.result_list_lookup = {}\n",
    "            \n",
    "\n",
    "    def forward(self, x):\n",
    "        # Note: x will be a list. input_tensor_lookup will contain the name: list index pairs.\n",
    "        # I use a dict instead of a list comprehension here because there could be an arbitrarily\n",
    "        # large number of inputs in the list. \n",
    "        def get_input_node(self, input_node, get_x):  \n",
    "#             print(input_node, self.result_list_lookup)\n",
    "            return(self.result_list[self.result_list_lookup[input_node]])\n",
    "        \n",
    "        # trying reinstantiating to get around inplace replacement issue.\n",
    "        self.result_list = []\n",
    "        self.result_list_lookup = {}\n",
    "        for key in self.dependancy_order:\n",
    "            input_nodes = self.example_dict[key]\n",
    "            nn_layer_list_idx = [i for i in range(len(dependancy_order)) if dependancy_order[i]==key][0]\n",
    "            \n",
    "            self.result_list_lookup[key] = len(self.result_list_lookup)                \n",
    "            if key in self.input_tensor_names: # If the input node is an input (flatten) layer\n",
    "                self.result_list = self.result_list + [self.nn_layer_list[nn_layer_list_idx](\n",
    "                    x[self.input_tensor_lookup[key]]\n",
    "                ).clone()]\n",
    "\n",
    "            elif key != 'y_hat':\n",
    "                # refactored to handle module lists (even if module list contains only one entry)\n",
    "                out = torch.concat(\n",
    "                    [get_input_node(self, input_node = e, get_x = x) for e in input_nodes], \n",
    "                    -1)\n",
    "            \n",
    "                for module in self.nn_layer_list[nn_layer_list_idx]:\n",
    "                    out = module(out)\n",
    "        \n",
    "                self.result_list = self.result_list + [out] \n",
    "            \n",
    "            else:\n",
    "                self.result_list = self.result_list + [self.nn_layer_list[nn_layer_list_idx](torch.concat(\n",
    "                    [get_input_node(self, input_node = e, get_x = x) for e in input_nodes], \n",
    "                    -1)).clone()]            \n",
    "\n",
    "        return self.result_list[self.result_list_lookup['y_hat']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "a3061d85",
   "metadata": {},
   "outputs": [],
   "source": [
    "# model = NeuralNetwork(example_dict = kegg_connections, \n",
    "#                       example_dict_input_size = input_size_dict,\n",
    "#                       example_dict_output_size = output_size_dict,\n",
    "#                       example_dict_dropout_pr= dropout_pr_dict,\n",
    "#                       example_block_rep_dict = block_rep_dict,\n",
    "#                       input_tensor_names = list(input_tensor_dict.keys()),\n",
    "#                       dependancy_order = dependancy_order) \n",
    "# model = model.to(device)\n",
    "# model(next(iter(training_dataloader))[1])\n",
    "# model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "55b56467",
   "metadata": {},
   "outputs": [],
   "source": [
    "class ListDataset(Dataset): # for any G containing matix with many (phno) to one (geno)\n",
    "    def __init__(self, \n",
    "                 y, \n",
    "                 x_list,\n",
    "                 obs_idxs, # this is a list of the indexes used. It allows us to pass in smaller \n",
    "                           # tensors and then get the right genotype\n",
    "                 obs_geno_lookup,\n",
    "                 transform = None, target_transform = None,\n",
    "                 **kwargs \n",
    "                ):\n",
    "        self.device = device\n",
    "        self.y = y \n",
    "        self.x_list = x_list\n",
    "        self.obs_idxs = obs_idxs\n",
    "        self.obs_geno_lookup = obs_geno_lookup\n",
    "        self.transform = transform\n",
    "        self.target_transform = target_transform    \n",
    "        \n",
    "    def __len__(self):\n",
    "        return len(self.y)\n",
    "    \n",
    "    def __getitem__(self, idx):\n",
    "        y_idx =self.y[idx]\n",
    "        \n",
    "        new_idx = self.obs_idxs[idx]\n",
    "        idx_geno = self.obs_geno_lookup[new_idx, 1]\n",
    "        x_idx =[x[idx_geno, ] for x in self.x_list] \n",
    "        \n",
    "        if self.target_transform:\n",
    "            y_idx = self.transform(y_idx)\n",
    "            x_idx = [self.transform(x) for x in x_idx]\n",
    "            \n",
    "        return y_idx, x_idx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "cb4ead9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "training_dataloader = DataLoader(ListDataset(\n",
    "        y =               X.get('YMat',ops_string='cs filter:val:train asarray from_numpy float cuda:0')[:, None],\n",
    "        x_list = [e.to('cuda') for e in x_list_temp],\n",
    "        obs_idxs =        X.get('val:train',       ops_string='   asarray from_numpy      '), \n",
    "        obs_geno_lookup = X.get('obs_geno_lookup', ops_string='   asarray from_numpy      ')\n",
    "    ),\n",
    "    batch_size = batch_size,\n",
    "    shuffle = True\n",
    ")\n",
    "\n",
    "\n",
    "validation_dataloader = DataLoader(ListDataset(\n",
    "        y =               X.get('YMat',ops_string='cs filter:val:test asarray from_numpy float cuda:0')[:, None],\n",
    "        x_list = [e.to('cuda') for e in x_list_temp],\n",
    "        obs_idxs =        X.get('val:test',       ops_string='   asarray from_numpy      '), \n",
    "        obs_geno_lookup = X.get('obs_geno_lookup', ops_string='   asarray from_numpy      ')\n",
    "    ),\n",
    "    batch_size = batch_size,\n",
    "    shuffle = True\n",
    ")\n",
    "# next(iter(training_dataloader))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a872acf1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9216d0e1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "c3b34f22",
   "metadata": {},
   "outputs": [],
   "source": [
    "# training_dataloader = DataLoader(ListDataset(\n",
    "#         y = y_temp[train_idx][:, None].to('cuda'),\n",
    "#         x_list = [e.to('cuda') for e in x_list_temp],\n",
    "#         obs_idxs = train_idx, \n",
    "#         obs_geno_lookup = obs_geno_lookup\n",
    "#     ),\n",
    "#     batch_size = dataloader_batch_size,\n",
    "#     shuffle = True\n",
    "# )\n",
    "\n",
    "# validation_dataloader = DataLoader(ListDataset(\n",
    "#         y = y_temp[test_idx][:, None].to('cuda'),\n",
    "#         x_list = [e.to('cuda') for e in x_list_temp],\n",
    "#         obs_idxs = test_idx, \n",
    "#         obs_geno_lookup = obs_geno_lookup\n",
    "#     ),\n",
    "#     batch_size = dataloader_batch_size,\n",
    "#     shuffle = True\n",
    "# )\n",
    "\n",
    "# # next(iter(training_dataloader))[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "996c1c29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# next(iter(training_dataloader))[1].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "0db51906",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "NeuralNetwork(\n",
       "  (nn_layer_list): ModuleList(\n",
       "    (0-5): 6 x Flatten(start_dim=1, end_dim=-1)\n",
       "    (6-7): 2 x ModuleList(\n",
       "      (0): Linear(in_features=4, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (8): ModuleList(\n",
       "      (0): Linear(in_features=12, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (9): ModuleList(\n",
       "      (0): Linear(in_features=44, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (10): ModuleList(\n",
       "      (0): Linear(in_features=4, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (11-12): 2 x ModuleList(\n",
       "      (0): Linear(in_features=12, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (13-14): 2 x ModuleList(\n",
       "      (0): Linear(in_features=50, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (15-16): 2 x ModuleList(\n",
       "      (0): Linear(in_features=4, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (17): ModuleList(\n",
       "      (0): Linear(in_features=50, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (18): ModuleList(\n",
       "      (0): Linear(in_features=12, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (19-20): 2 x ModuleList(\n",
       "      (0): Linear(in_features=44, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (21): ModuleList(\n",
       "      (0): Linear(in_features=50, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (22-23): 2 x ModuleList(\n",
       "      (0): Linear(in_features=44, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (24): ModuleList(\n",
       "      (0): Linear(in_features=4, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (25): ModuleList(\n",
       "      (0): Linear(in_features=12, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (26): ModuleList(\n",
       "      (0): Linear(in_features=100, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (27-28): 2 x ModuleList(\n",
       "      (0): Linear(in_features=50, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (29): ModuleList(\n",
       "      (0): Linear(in_features=100, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (30-34): 5 x ModuleList(\n",
       "      (0): Linear(in_features=50, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (35): ModuleList(\n",
       "      (0): Linear(in_features=150, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (36-40): 5 x ModuleList(\n",
       "      (0): Linear(in_features=50, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (41): ModuleList(\n",
       "      (0): Linear(in_features=200, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (42): ModuleList(\n",
       "      (0): Linear(in_features=50, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (43): ModuleList(\n",
       "      (0): Linear(in_features=100, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (44-48): 5 x ModuleList(\n",
       "      (0): Linear(in_features=50, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (49): ModuleList(\n",
       "      (0): Linear(in_features=100, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (50): ModuleList(\n",
       "      (0): Linear(in_features=150, out_features=50, bias=True)\n",
       "      (1): ReLU()\n",
       "      (2): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (51): Linear(in_features=350, out_features=1, bias=True)\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = NeuralNetwork(example_dict = kegg_connections, \n",
    "                      example_dict_input_size = input_size_dict,\n",
    "                      example_dict_output_size = output_size_dict,\n",
    "                      example_dict_dropout_pr= dropout_pr_dict,\n",
    "                      example_block_rep_dict = block_rep_dict,\n",
    "                      input_tensor_names = list(input_tensor_dict.keys()),\n",
    "                      dependancy_order = dependancy_order)\n",
    "\n",
    "\n",
    "model.to('cuda')\n",
    "# LSUV_(model, data = next(iter(training_dataloader))[1])\n",
    "# model(next(iter(training_dataloader))[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "640349cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Applying orthogonal init (zero init if dim < 2) to params in 46 module(s).\n",
      "Applying LSUV to 46 module(s) (up to 10 iters per module):\n",
      "Module  0 after  2 itr(s) | Mean:  0.107 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module  1 after  2 itr(s) | Mean:  0.096 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module  2 after  2 itr(s) | Mean: -0.039 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module  3 after  2 itr(s) | Mean:  0.192 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module  4 after  2 itr(s) | Mean: -0.002 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module  5 after  2 itr(s) | Mean: -0.108 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module  6 after  2 itr(s) | Mean: -0.001 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module  7 after  2 itr(s) | Mean:  0.133 | Std: 0.989 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module  8 after  2 itr(s) | Mean: -0.023 | Std: 1.008 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module  9 after  2 itr(s) | Mean:  0.126 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 10 after  2 itr(s) | Mean:  0.016 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 11 after  2 itr(s) | Mean: -0.089 | Std: 1.017 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 12 after  2 itr(s) | Mean: -0.013 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 13 after  2 itr(s) | Mean:  0.112 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 14 after  2 itr(s) | Mean:  0.074 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 15 after  2 itr(s) | Mean:  0.001 | Std: 0.993 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 16 after  2 itr(s) | Mean: -0.030 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 17 after  2 itr(s) | Mean:  0.222 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 18 after  2 itr(s) | Mean: -0.119 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 19 after  2 itr(s) | Mean: -0.099 | Std: 1.000 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 20 after  2 itr(s) | Mean:  0.046 | Std: 1.012 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 21 after  2 itr(s) | Mean:  0.194 | Std: 1.032 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 22 after  2 itr(s) | Mean: -0.096 | Std: 1.028 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 23 after  2 itr(s) | Mean: -0.113 | Std: 0.970 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 24 after  2 itr(s) | Mean: -0.036 | Std: 1.050 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 25 after  2 itr(s) | Mean:  0.020 | Std: 1.033 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 26 after  2 itr(s) | Mean: -0.015 | Std: 0.983 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 27 after  2 itr(s) | Mean: -0.111 | Std: 0.998 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 28 after  2 itr(s) | Mean:  0.077 | Std: 0.979 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 29 after  2 itr(s) | Mean:  0.156 | Std: 1.003 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 30 after  2 itr(s) | Mean: -0.061 | Std: 0.955 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 31 after  2 itr(s) | Mean: -0.146 | Std: 0.988 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 32 after  2 itr(s) | Mean:  0.191 | Std: 1.012 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 33 after  2 itr(s) | Mean:  0.055 | Std: 0.960 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 34 after  2 itr(s) | Mean: -0.170 | Std: 1.029 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 35 after  2 itr(s) | Mean: -0.095 | Std: 1.046 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 36 after  2 itr(s) | Mean: -0.122 | Std: 0.986 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 37 after  2 itr(s) | Mean:  0.162 | Std: 1.013 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 38 after  2 itr(s) | Mean: -0.046 | Std: 0.985 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 39 after  2 itr(s) | Mean: -0.068 | Std: 1.008 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 40 after  2 itr(s) | Mean:  0.009 | Std: 0.962 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 41 after  2 itr(s) | Mean:  0.085 | Std: 0.943 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 42 after  2 itr(s) | Mean: -0.032 | Std: 1.011 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 43 after  2 itr(s) | Mean: -0.090 | Std: 0.970 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 44 after  2 itr(s) | Mean: -0.095 | Std: 1.030 | <class 'torch.nn.modules.linear.Linear'>\n",
      "Module 45 after  2 itr(s) | Mean:  0.243 | Std: 1.083 | <class 'torch.nn.modules.linear.Linear'>\n"
     ]
    }
   ],
   "source": [
    "LSUV_(model, data = next(iter(training_dataloader))[1])\n",
    "# model(next(iter(training_dataloader))[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "ad25d25c",
   "metadata": {},
   "outputs": [],
   "source": [
    "class plVNN(pl.LightningModule):\n",
    "    def __init__(self, mod):\n",
    "        super().__init__()\n",
    "        self.mod = mod\n",
    "    def training_step(self, batch, batch_idx):\n",
    "        y_i, xs_i = batch\n",
    "#         pred, out = self.mod(xs_i)\n",
    "        pred = self.mod(xs_i)\n",
    "        loss = F.mse_loss(pred, y_i)\n",
    "        self.log(\"train_loss\", loss)\n",
    "        \n",
    "#         with torch.no_grad():\n",
    "#             weight_list=[(name, param) for name, param in model.named_parameters() if name.split('.')[-1] == 'weight']\n",
    "#             for l in weight_list:\n",
    "#                 self.log((\"train_mean\"+l[0]), l[1].mean())\n",
    "#                 self.log((\"train_std\"+l[0]), l[1].std())        \n",
    "        return(loss)\n",
    "        \n",
    "    def validation_step(self, batch, batch_idx):\n",
    "        y_i, xs_i = batch\n",
    "#         pred, out = self.mod(xs_i)\n",
    "        pred = self.mod(xs_i)\n",
    "        loss = F.mse_loss(pred, y_i)\n",
    "        self.log('val_loss', loss)        \n",
    "     \n",
    "    def configure_optimizers(self, **kwargs):\n",
    "        optimizer = torch.optim.Adam(self.parameters(), **kwargs)\n",
    "        return optimizer    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "2116f9e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# 2 epochs took 58 minutes so a 3 day weekend would be \n",
    "# 24*3*2=144 epochs\n",
    "# max_epoch = 150"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "199df0ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kickd/miniconda3/envs/fastai/lib/python3.11/site-packages/lightning/fabric/plugins/environments/slurm.py:191: The `srun` command is available on your system but is not used. HINT: If your intention is to run Lightning on SLURM, prepend your python command with `srun` like so: srun python /home/kickd/miniconda3/envs/fastai/lib/python3.11/si ...\n",
      "GPU available: True (cuda), used: True\n",
      "TPU available: False, using: 0 TPU cores\n",
      "IPU available: False, using: 0 IPUs\n",
      "HPU available: False, using: 0 HPUs\n",
      "You are using a CUDA device ('NVIDIA GeForce RTX 4090') that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision('medium' | 'high')` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision\n",
      "LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]\n",
      "\n",
      "  | Name | Type          | Params\n",
      "---------------------------------------\n",
      "0 | mod  | NeuralNetwork | 117 K \n",
      "---------------------------------------\n",
      "117 K     Trainable params\n",
      "0         Non-trainable params\n",
      "117 K     Total params\n",
      "0.471     Total estimated model params size (MB)\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8c2f715b644a42fba602490e645f1a52",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Sanity Checking: |          | 0/? [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/kickd/miniconda3/envs/fastai/lib/python3.11/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:492: Your `val_dataloader`'s sampler has shuffling enabled, it is strongly recommended that you turn shuffling off for val/test dataloaders.\n",
      "/home/kickd/miniconda3/envs/fastai/lib/python3.11/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:441: The 'val_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=31` in the `DataLoader` to improve performance.\n",
      "/home/kickd/miniconda3/envs/fastai/lib/python3.11/site-packages/lightning/pytorch/trainer/connectors/data_connector.py:441: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=31` in the `DataLoader` to improve performance.\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9ff703afd2f24a2a8f87701acd836c5d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Training: |          | 0/? [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "VNN = plVNN(model)\n",
    "optimizer = VNN.configure_optimizers()\n",
    "\n",
    "logger = TensorBoardLogger(\"tb_logs\", name=\"vnn-02.31-TESTING-REMOVE-ME\")\n",
    "trainer = pl.Trainer(max_epochs=max_epoch, logger=logger)\n",
    "\n",
    "trainer.fit(model=VNN, train_dataloaders=training_dataloader, val_dataloaders=validation_dataloader)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5e34e76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# torch.save(VNN.mod, cache_path+'vnn'+'.pt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cc7f440",
   "metadata": {},
   "outputs": [],
   "source": [
    "# torch.save(VNN, cache_path+'vnn'+'.pt')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
